{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from mutil.alemutdf import get_all_sample_mut_df, get_gene_mut_count_mat, get_multi_exp_max_freq_mut_df, get_mut_type_avg_frac_across_class_df\n",
    "from mutil.mut import is_coding_mut, get_original_nuc_mut_range\n",
    "from mutil.metadata import get_condition_val_dict, get_condition_field_val_set\n",
    "from mutil.genome import get_feature_hit_set, is_overlap, get_promoter_range_from_RegulonDB_df_row, NON_K12_EXP_L "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(585, 14)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./data/1_df.pkl\")\n",
    "display(all_muts_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Renaming the \"Gene\" column to \"mutation target annotation\" because this is more accurate.\n",
    "all_muts_df = all_muts_df.rename(index=str, columns={\"Gene\": \"mutation target annotation\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build genetic feature annotation\n",
    "Need to build genetic feature annotations before genomic features annotations since the genetic annotations are used to fill gaps with unknown genomic targets."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## K12 specific genetic annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>GENE_POSLEFT</th>\n",
       "      <th>GENE_POSRIGHT</th>\n",
       "      <th>GENE_STRAND</th>\n",
       "      <th>GENE_SEQUENCE</th>\n",
       "      <th>GC_CONTENT</th>\n",
       "      <th>CRI_SCORE</th>\n",
       "      <th>GENE_NOTE</th>\n",
       "      <th>GENE_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>GENE_TYPE</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr</td>\n",
       "      <td>4265782.0</td>\n",
       "      <td>4266861.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...</td>\n",
       "      <td>55.93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4265782, 4266861)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>modB</td>\n",
       "      <td>795862.0</td>\n",
       "      <td>796551.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...</td>\n",
       "      <td>54.06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(795862, 796551)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120000003</td>\n",
       "      <td>cysZ</td>\n",
       "      <td>2531463.0</td>\n",
       "      <td>2532224.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...</td>\n",
       "      <td>50.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2531463, 2532224)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120000004</td>\n",
       "      <td>dfp</td>\n",
       "      <td>3812731.0</td>\n",
       "      <td>3813951.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...</td>\n",
       "      <td>53.64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3812731, 3813951)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120000005</td>\n",
       "      <td>dcuB</td>\n",
       "      <td>4347404.0</td>\n",
       "      <td>4348744.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...</td>\n",
       "      <td>52.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4347404, 4348744)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GENE_ID GENE_NAME  GENE_POSLEFT  GENE_POSRIGHT GENE_STRAND  \\\n",
       "0  ECK120000001       alr     4265782.0      4266861.0     forward   \n",
       "1  ECK120000002      modB      795862.0       796551.0     forward   \n",
       "2  ECK120000003      cysZ     2531463.0      2532224.0     forward   \n",
       "3  ECK120000004       dfp     3812731.0      3813951.0     forward   \n",
       "4  ECK120000005      dcuB     4347404.0      4348744.0     reverse   \n",
       "\n",
       "                                       GENE_SEQUENCE  GC_CONTENT  CRI_SCORE  \\\n",
       "0  ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...       55.93        NaN   \n",
       "1  ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...       54.06        NaN   \n",
       "2  ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...       50.13        NaN   \n",
       "3  ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...       53.64        NaN   \n",
       "4  ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...       52.27        NaN   \n",
       "\n",
       "  GENE_NOTE  GENE_INTERNAL_COMMENT KEY_ID_ORG GENE_TYPE               range  \n",
       "0       NaN                    NaN      ECK12       NaN  (4265782, 4266861)  \n",
       "1       NaN                    NaN      ECK12       NaN    (795862, 796551)  \n",
       "2       NaN                    NaN      ECK12       NaN  (2531463, 2532224)  \n",
       "3       NaN                    NaN      ECK12       NaN  (3812731, 3813951)  \n",
       "4       NaN                    NaN      ECK12       NaN  (4347404, 4348744)  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df = pd.read_pickle(\"./data/gene_df.pkl\")\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13957</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>M599L (ATG→TTG)</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(13957, 13957)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28175</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>W295R (TGG→AGG)</td>\n",
       "      <td>rihC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(28175, 28175)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101342</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>T193I (ACC→ATC)</td>\n",
       "      <td>murC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(101342, 101342)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145691</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A204V (GCC→GTC)</td>\n",
       "      <td>yadE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(145691, 145691)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171072</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→G</td>\n",
       "      <td>K166K (AAA→AAG)</td>\n",
       "      <td>fhuD</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(171072, 171072)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "1392  GLU    3    412        2         1       1.0     13957           SNP   \n",
       "1428  GLU    3    412        2         1       1.0     28175           SNP   \n",
       "1393  GLU    3    412        2         1       1.0    101342           SNP   \n",
       "1394  GLU    3    412        2         1       1.0    145691           SNP   \n",
       "1429  GLU    3    412        2         1       1.0    171072           SNP   \n",
       "\n",
       "     Sequence Change          Details mutation target annotation  \\\n",
       "1392             A→T  M599L (ATG→TTG)                       dnaK   \n",
       "1428             T→A  W295R (TGG→AGG)                       rihC   \n",
       "1393             C→T  T193I (ACC→ATC)                       murC   \n",
       "1394             C→T  A204V (GCC→GTC)                       yadE   \n",
       "1429             A→G  K166K (AAA→AAG)                       fhuD   \n",
       "\n",
       "     Reference Seq     sample  coding             range  \n",
       "1392     NC_000913  3 412 2 1    True    (13957, 13957)  \n",
       "1428     NC_000913  3 412 2 1    True    (28175, 28175)  \n",
       "1393     NC_000913  3 412 2 1    True  (101342, 101342)  \n",
       "1394     NC_000913  3 412 2 1    True  (145691, 145691)  \n",
       "1429     NC_000913  3 412 2 1    True  (171072, 171072)  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_muts_df[\"range\"] = all_muts_df.apply(get_original_nuc_mut_range, axis=1)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13957</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>M599L (ATG→TTG)</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(13957, 13957)</td>\n",
       "      <td>{ECK120000235}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28175</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>W295R (TGG→AGG)</td>\n",
       "      <td>rihC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(28175, 28175)</td>\n",
       "      <td>{ECK120001070}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101342</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>T193I (ACC→ATC)</td>\n",
       "      <td>murC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(101342, 101342)</td>\n",
       "      <td>{ECK120000612}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145691</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A204V (GCC→GTC)</td>\n",
       "      <td>yadE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(145691, 145691)</td>\n",
       "      <td>{ECK120001687}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171072</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→G</td>\n",
       "      <td>K166K (AAA→AAG)</td>\n",
       "      <td>fhuD</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(171072, 171072)</td>\n",
       "      <td>{ECK120000299}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "1392  GLU    3    412        2         1       1.0     13957           SNP   \n",
       "1428  GLU    3    412        2         1       1.0     28175           SNP   \n",
       "1393  GLU    3    412        2         1       1.0    101342           SNP   \n",
       "1394  GLU    3    412        2         1       1.0    145691           SNP   \n",
       "1429  GLU    3    412        2         1       1.0    171072           SNP   \n",
       "\n",
       "     Sequence Change          Details mutation target annotation  \\\n",
       "1392             A→T  M599L (ATG→TTG)                       dnaK   \n",
       "1428             T→A  W295R (TGG→AGG)                       rihC   \n",
       "1393             C→T  T193I (ACC→ATC)                       murC   \n",
       "1394             C→T  A204V (GCC→GTC)                       yadE   \n",
       "1429             A→G  K166K (AAA→AAG)                       fhuD   \n",
       "\n",
       "     Reference Seq     sample  coding             range gene RegulonDB ID  \n",
       "1392     NC_000913  3 412 2 1    True    (13957, 13957)    {ECK120000235}  \n",
       "1428     NC_000913  3 412 2 1    True    (28175, 28175)    {ECK120001070}  \n",
       "1393     NC_000913  3 412 2 1    True  (101342, 101342)    {ECK120000612}  \n",
       "1394     NC_000913  3 412 2 1    True  (145691, 145691)    {ECK120001687}  \n",
       "1429     NC_000913  3 412 2 1    True  (171072, 171072)    {ECK120000299}  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_muts_df[\"gene RegulonDB ID\"] = all_muts_df.apply(\n",
    "    lambda r: get_feature_hit_set(\n",
    "        r[\"range\"], gene_df, \"range\", \"GENE_ID\") if r.exp not in NON_K12_EXP_L else set(), axis=1\n",
    ")  # Maybe instead just have a list of K12 experiments\n",
    "\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>RegulonDB ID</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>thrL/thrA</td>\n",
       "      <td>ECK120001251/ECK120000987</td>\n",
       "      <td>(256, 336)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>thrA/thrB</td>\n",
       "      <td>ECK120000987/ECK120000988</td>\n",
       "      <td>(2800, 2800)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>thrC/yaaX</td>\n",
       "      <td>ECK120000989/ECK120002701</td>\n",
       "      <td>(5021, 5233)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>yaaX/yaaA</td>\n",
       "      <td>ECK120002701/ECK120000009</td>\n",
       "      <td>(5531, 5682)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>yaaA/yaaJ</td>\n",
       "      <td>ECK120000009/ECK120001508</td>\n",
       "      <td>(6460, 6528)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        name               RegulonDB ID         range\n",
       "0  thrL/thrA  ECK120001251/ECK120000987    (256, 336)\n",
       "1  thrA/thrB  ECK120000987/ECK120000988  (2800, 2800)\n",
       "2  thrC/yaaX  ECK120000989/ECK120002701  (5021, 5233)\n",
       "3  yaaX/yaaA  ECK120002701/ECK120000009  (5531, 5682)\n",
       "4  yaaA/yaaJ  ECK120000009/ECK120001508  (6460, 6528)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenic_df = pd.read_pickle(\"./data/intergenic_region_df.pkl\")\n",
    "intergenic_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13957</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>M599L (ATG→TTG)</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(13957, 13957)</td>\n",
       "      <td>{ECK120000235}</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28175</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>W295R (TGG→AGG)</td>\n",
       "      <td>rihC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(28175, 28175)</td>\n",
       "      <td>{ECK120001070}</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101342</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>T193I (ACC→ATC)</td>\n",
       "      <td>murC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(101342, 101342)</td>\n",
       "      <td>{ECK120000612}</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145691</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A204V (GCC→GTC)</td>\n",
       "      <td>yadE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(145691, 145691)</td>\n",
       "      <td>{ECK120001687}</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171072</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→G</td>\n",
       "      <td>K166K (AAA→AAG)</td>\n",
       "      <td>fhuD</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(171072, 171072)</td>\n",
       "      <td>{ECK120000299}</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "1392  GLU    3    412        2         1       1.0     13957           SNP   \n",
       "1428  GLU    3    412        2         1       1.0     28175           SNP   \n",
       "1393  GLU    3    412        2         1       1.0    101342           SNP   \n",
       "1394  GLU    3    412        2         1       1.0    145691           SNP   \n",
       "1429  GLU    3    412        2         1       1.0    171072           SNP   \n",
       "\n",
       "     Sequence Change          Details mutation target annotation  \\\n",
       "1392             A→T  M599L (ATG→TTG)                       dnaK   \n",
       "1428             T→A  W295R (TGG→AGG)                       rihC   \n",
       "1393             C→T  T193I (ACC→ATC)                       murC   \n",
       "1394             C→T  A204V (GCC→GTC)                       yadE   \n",
       "1429             A→G  K166K (AAA→AAG)                       fhuD   \n",
       "\n",
       "     Reference Seq     sample  coding             range gene RegulonDB ID  \\\n",
       "1392     NC_000913  3 412 2 1    True    (13957, 13957)    {ECK120000235}   \n",
       "1428     NC_000913  3 412 2 1    True    (28175, 28175)    {ECK120001070}   \n",
       "1393     NC_000913  3 412 2 1    True  (101342, 101342)    {ECK120000612}   \n",
       "1394     NC_000913  3 412 2 1    True  (145691, 145691)    {ECK120001687}   \n",
       "1429     NC_000913  3 412 2 1    True  (171072, 171072)    {ECK120000299}   \n",
       "\n",
       "                                       genetic features  \n",
       "1392  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  \n",
       "1428  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  \n",
       "1393  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  \n",
       "1394  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  \n",
       "1429  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def get_feature_range(feat_id, feat_type):\n",
    "    r = ()\n",
    "    feat_id_col = 0\n",
    "    range_col = \"range\"\n",
    "    feat_id_col = \"GENE_ID\"\n",
    "    feat_type_df = gene_df\n",
    "\n",
    "    df = feat_type_df[feat_type_df[feat_id_col] == feat_id]\n",
    "    if not df.empty:\n",
    "        r = df.iloc[0][range_col]  # Assuming all rows describe the same range.\n",
    "    l = [int(i) for i in r]\n",
    "    return tuple(l)\n",
    "\n",
    "\n",
    "# This will include all implicit intergenic regions, where get_genomic_features(...) does not.\n",
    "# TODO: this may not handle wrapping around the genome from the last gene to the first gene. Write test for this.\n",
    "def get_ordered_genetic_targets(genetic_targets):\n",
    "    ordered_genetic_targets = sorted(\n",
    "        genetic_targets, key=lambda k: k[\"range\"][0])\n",
    "    return ordered_genetic_targets\n",
    "\n",
    "\n",
    "test_genetic_targets = [\n",
    "    {'RegulonDB ID': \"ECK120000003\",\n",
    "     \"range\": (2531463, 2532224)},\n",
    "    {'RegulonDB ID': \"ECK120000001\",\n",
    "     \"range\": (4265782, 4266861)},\n",
    "    {'RegulonDB ID': \"ECK120000002\",\n",
    "     \"range\": (795862, 796551)}]\n",
    "assert(get_ordered_genetic_targets(test_genetic_targets) ==\n",
    "       [{'RegulonDB ID': 'ECK120000002', 'range': (795862, 796551)},\n",
    "        {'RegulonDB ID': 'ECK120000003', 'range': (2531463, 2532224)},\n",
    "        {'RegulonDB ID': 'ECK120000001', 'range': (4265782, 4266861)}])\n",
    "\n",
    "\n",
    "#Simply returning the first gene name in the second column that has the RegulonDB ID in the first column.\n",
    "def get_gene_name(RegulonDB_ID, debug=False):\n",
    "    if debug:\n",
    "        display(RegulonDB_ID, gene_df[gene_df[\"GENE_ID\"] == RegulonDB_ID])\n",
    "    return gene_df[gene_df[\"GENE_ID\"] == RegulonDB_ID].iloc[0][\"GENE_NAME\"]\n",
    "\n",
    "\n",
    "def get_genetic_feat_JSON(mut_row):\n",
    "    l = []\n",
    "    for gene_id in mut_row[\"gene RegulonDB ID\"]:\n",
    "        l.append({\n",
    "            # IDs are unique; assuming only be one entry per DF.\n",
    "            \"name\": get_gene_name(gene_id),\n",
    "            \"RegulonDB ID\": gene_id,\n",
    "            \"range\": get_feature_range(gene_id, \"gene RegulonDB ID\"),\n",
    "            \"feature type\": \"gene\"\n",
    "        })\n",
    "\n",
    "    intergenic_regions = get_feature_hit_set(mut_row[\"range\"], intergenic_df, \"range\", \"RegulonDB ID\")\n",
    "    for intergenic_region in intergenic_regions:\n",
    "        l.append({\n",
    "            \"RegulonDB ID\": intergenic_region,\n",
    "            \"name\": str(get_gene_name(intergenic_region.split('/')[0])) + '/' + str(get_gene_name(intergenic_region.split('/')[1])),\n",
    "            \"range\": intergenic_df[intergenic_df[\"RegulonDB ID\"] == intergenic_region].iloc[0][\"range\"],\n",
    "            \"feature type\": \"unknown\"\n",
    "        })\n",
    "\n",
    "    return l\n",
    "\n",
    "\n",
    "all_muts_df[\"genetic features\"] = all_muts_df.apply(lambda r: get_genetic_feat_JSON(r) if r.exp not in NON_K12_EXP_L else [], axis=1)\n",
    "display(all_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_muts_df.to_pickle(\"./data/1_5_df.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
