{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from mutil.gene import get_coding_genetic_target_len_d, get_intergenic_len_d\n",
    "from mutil.genome import get_feature_hit_set, is_overlap, get_promoter_range_from_RegulonDB_df_row\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(585, 23)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13957</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>M599L (ATG→TTG)</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(13957, 13957)</td>\n",
       "      <td>{ECK120000235}</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28175</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>W295R (TGG→AGG)</td>\n",
       "      <td>rihC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(28175, 28175)</td>\n",
       "      <td>{ECK120001070}</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101342</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>T193I (ACC→ATC)</td>\n",
       "      <td>murC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(101342, 101342)</td>\n",
       "      <td>{ECK120000612}</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145691</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A204V (GCC→GTC)</td>\n",
       "      <td>yadE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(145691, 145691)</td>\n",
       "      <td>{ECK120001687}</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171072</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→G</td>\n",
       "      <td>K166K (AAA→AAG)</td>\n",
       "      <td>fhuD</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(171072, 171072)</td>\n",
       "      <td>{ECK120000299}</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "1392  GLU    3    412        2         1       1.0     13957           SNP   \n",
       "1428  GLU    3    412        2         1       1.0     28175           SNP   \n",
       "1393  GLU    3    412        2         1       1.0    101342           SNP   \n",
       "1394  GLU    3    412        2         1       1.0    145691           SNP   \n",
       "1429  GLU    3    412        2         1       1.0    171072           SNP   \n",
       "\n",
       "     Sequence Change          Details mutation target annotation  \\\n",
       "1392             A→T  M599L (ATG→TTG)                       dnaK   \n",
       "1428             T→A  W295R (TGG→AGG)                       rihC   \n",
       "1393             C→T  T193I (ACC→ATC)                       murC   \n",
       "1394             C→T  A204V (GCC→GTC)                       yadE   \n",
       "1429             A→G  K166K (AAA→AAG)                       fhuD   \n",
       "\n",
       "     Reference Seq     sample  coding             range gene RegulonDB ID  \\\n",
       "1392     NC_000913  3 412 2 1    True    (13957, 13957)    {ECK120000235}   \n",
       "1428     NC_000913  3 412 2 1    True    (28175, 28175)    {ECK120001070}   \n",
       "1393     NC_000913  3 412 2 1    True  (101342, 101342)    {ECK120000612}   \n",
       "1394     NC_000913  3 412 2 1    True  (145691, 145691)    {ECK120001687}   \n",
       "1429     NC_000913  3 412 2 1    True  (171072, 171072)    {ECK120000299}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1392  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1428  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  False       False   \n",
       "1393  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  False       False   \n",
       "1394  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  False       False   \n",
       "1429  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  False       False   \n",
       "\n",
       "     TFBS promoter RBS terminator  \n",
       "1392   {}       {}  {}         {}  \n",
       "1428   {}       {}  {}         {}  \n",
       "1393   {}       {}  {}         {}  \n",
       "1394   {}       {}  {}         {}  \n",
       "1429   {}       {}  {}         {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./data/2_df.pkl\")\n",
    "display(all_muts_df.shape, all_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add component ranges\n",
    "Need to find genetic features first to serve as defaults for when genomic features can't be identified."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>GENE_POSLEFT</th>\n",
       "      <th>GENE_POSRIGHT</th>\n",
       "      <th>GENE_STRAND</th>\n",
       "      <th>GENE_SEQUENCE</th>\n",
       "      <th>GC_CONTENT</th>\n",
       "      <th>CRI_SCORE</th>\n",
       "      <th>GENE_NOTE</th>\n",
       "      <th>GENE_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>GENE_TYPE</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr</td>\n",
       "      <td>4265782.0</td>\n",
       "      <td>4266861.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...</td>\n",
       "      <td>55.93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4265782.0, 4266861.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>modB</td>\n",
       "      <td>795862.0</td>\n",
       "      <td>796551.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...</td>\n",
       "      <td>54.06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(795862.0, 796551.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120000003</td>\n",
       "      <td>cysZ</td>\n",
       "      <td>2531463.0</td>\n",
       "      <td>2532224.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...</td>\n",
       "      <td>50.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2531463.0, 2532224.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120000004</td>\n",
       "      <td>dfp</td>\n",
       "      <td>3812731.0</td>\n",
       "      <td>3813951.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...</td>\n",
       "      <td>53.64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3812731.0, 3813951.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120000005</td>\n",
       "      <td>dcuB</td>\n",
       "      <td>4347404.0</td>\n",
       "      <td>4348744.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...</td>\n",
       "      <td>52.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4347404.0, 4348744.0)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GENE_ID GENE_NAME  GENE_POSLEFT  GENE_POSRIGHT GENE_STRAND  \\\n",
       "0  ECK120000001       alr     4265782.0      4266861.0     forward   \n",
       "1  ECK120000002      modB      795862.0       796551.0     forward   \n",
       "2  ECK120000003      cysZ     2531463.0      2532224.0     forward   \n",
       "3  ECK120000004       dfp     3812731.0      3813951.0     forward   \n",
       "4  ECK120000005      dcuB     4347404.0      4348744.0     reverse   \n",
       "\n",
       "                                       GENE_SEQUENCE  GC_CONTENT  CRI_SCORE  \\\n",
       "0  ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...       55.93        NaN   \n",
       "1  ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...       54.06        NaN   \n",
       "2  ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...       50.13        NaN   \n",
       "3  ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...       53.64        NaN   \n",
       "4  ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...       52.27        NaN   \n",
       "\n",
       "  GENE_NOTE  GENE_INTERNAL_COMMENT KEY_ID_ORG GENE_TYPE  \\\n",
       "0       NaN                    NaN      ECK12       NaN   \n",
       "1       NaN                    NaN      ECK12       NaN   \n",
       "2       NaN                    NaN      ECK12       NaN   \n",
       "3       NaN                    NaN      ECK12       NaN   \n",
       "4       NaN                    NaN      ECK12       NaN   \n",
       "\n",
       "                    range  \n",
       "0  (4265782.0, 4266861.0)  \n",
       "1    (795862.0, 796551.0)  \n",
       "2  (2531463.0, 2532224.0)  \n",
       "3  (3812731.0, 3813951.0)  \n",
       "4  (4347404.0, 4348744.0)  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df = pd.read_pickle(\"./data/gene_df.pkl\")\n",
    "gene_df[\"range\"] = gene_df.apply(lambda row: (row[\"GENE_POSLEFT\"], row[\"GENE_POSRIGHT\"]), axis=1)\n",
    "TU_df = pd.read_pickle(\"./data/TU_df.pkl\")\n",
    "operon_df = pd.read_pickle(\"./data/operon_df.pkl\")\n",
    "\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRANSCRIPTION_UNIT_ID</th>\n",
       "      <th>NUMTU</th>\n",
       "      <th>TU_POSLEFT</th>\n",
       "      <th>TU_POSRIGHT</th>\n",
       "      <th>TU_TYPE</th>\n",
       "      <th>TU_OBJECT_CLASS</th>\n",
       "      <th>TU_OBJECT_ID</th>\n",
       "      <th>TU_OBJECT_NAME</th>\n",
       "      <th>TU_OBJECT_POSLEFT</th>\n",
       "      <th>TU_OBJECT_POSRIGHT</th>\n",
       "      <th>TU_OBJECT_STRAND</th>\n",
       "      <th>TU_OBJECT_COLORCLASS</th>\n",
       "      <th>TU_OBJECT_DESCRIPTION</th>\n",
       "      <th>TU_OBJECT_SIGMA</th>\n",
       "      <th>TU_OBJECT_EVIDENCE</th>\n",
       "      <th>TU_OBJECT_RI_TYPE</th>\n",
       "      <th>TU_OBJECT_TYPE</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>OPERON_ID</th>\n",
       "      <th>OPERON_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>PM</td>\n",
       "      <td>ECK120009851</td>\n",
       "      <td>astCp1</td>\n",
       "      <td>1832013</td>\n",
       "      <td>1832013</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sigma70</td>\n",
       "      <td>Human inference of promoter position</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003528</td>\n",
       "      <td>astE</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1826923</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003529</td>\n",
       "      <td>astB</td>\n",
       "      <td>1826916</td>\n",
       "      <td>1828259</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003532</td>\n",
       "      <td>astC</td>\n",
       "      <td>1830762</td>\n",
       "      <td>1831982</td>\n",
       "      <td>R</td>\n",
       "      <td>0,0,255</td>\n",
       "      <td>nitrogen metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003530</td>\n",
       "      <td>astD</td>\n",
       "      <td>1828256</td>\n",
       "      <td>1829734</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003531</td>\n",
       "      <td>astA</td>\n",
       "      <td>1829731</td>\n",
       "      <td>1830765</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>PM</td>\n",
       "      <td>ECK120009851</td>\n",
       "      <td>astCp1</td>\n",
       "      <td>1832013</td>\n",
       "      <td>1832013</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sigma70</td>\n",
       "      <td>Human inference of promoter position</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003528</td>\n",
       "      <td>astE</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1826923</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003529</td>\n",
       "      <td>astB</td>\n",
       "      <td>1826916</td>\n",
       "      <td>1828259</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003532</td>\n",
       "      <td>astC</td>\n",
       "      <td>1830762</td>\n",
       "      <td>1831982</td>\n",
       "      <td>R</td>\n",
       "      <td>0,0,255</td>\n",
       "      <td>nitrogen metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK120014360</td>\n",
       "      <td>astCADBE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  TRANSCRIPTION_UNIT_ID  NUMTU  TU_POSLEFT  TU_POSRIGHT TU_TYPE  \\\n",
       "0          ECK120008913      3     1825955      1832013       H   \n",
       "1          ECK120008913      3     1825955      1832013       H   \n",
       "2          ECK120008913      3     1825955      1832013       H   \n",
       "3          ECK120008913      3     1825955      1832013       H   \n",
       "4          ECK120008913      3     1825955      1832013       H   \n",
       "5          ECK120008913      3     1825955      1832013       H   \n",
       "6          ECK120008913      3     1825955      1832013       H   \n",
       "7          ECK120008913      3     1825955      1832013       H   \n",
       "8          ECK120008913      3     1825955      1832013       H   \n",
       "9          ECK120008913      3     1825955      1832013       H   \n",
       "\n",
       "  TU_OBJECT_CLASS  TU_OBJECT_ID TU_OBJECT_NAME  TU_OBJECT_POSLEFT  \\\n",
       "0              PM  ECK120009851         astCp1            1832013   \n",
       "1              GN  ECK120003528           astE            1825955   \n",
       "2              GN  ECK120003529           astB            1826916   \n",
       "3              GN  ECK120003532           astC            1830762   \n",
       "4              GN  ECK120003530           astD            1828256   \n",
       "5              GN  ECK120003531           astA            1829731   \n",
       "6              PM  ECK120009851         astCp1            1832013   \n",
       "7              GN  ECK120003528           astE            1825955   \n",
       "8              GN  ECK120003529           astB            1826916   \n",
       "9              GN  ECK120003532           astC            1830762   \n",
       "\n",
       "   TU_OBJECT_POSRIGHT TU_OBJECT_STRAND TU_OBJECT_COLORCLASS  \\\n",
       "0             1832013                R                  NaN   \n",
       "1             1826923                R           51,153,255   \n",
       "2             1828259                R           51,153,255   \n",
       "3             1831982                R              0,0,255   \n",
       "4             1829734                R           51,153,255   \n",
       "5             1830765                R           51,153,255   \n",
       "6             1832013                R                  NaN   \n",
       "7             1826923                R           51,153,255   \n",
       "8             1828259                R           51,153,255   \n",
       "9             1831982                R              0,0,255   \n",
       "\n",
       "  TU_OBJECT_DESCRIPTION TU_OBJECT_SIGMA                    TU_OBJECT_EVIDENCE  \\\n",
       "0                   NaN         Sigma70  Human inference of promoter position   \n",
       "1           amino acids             NaN                                   NaN   \n",
       "2           amino acids             NaN                                   NaN   \n",
       "3   nitrogen metabolism             NaN                                   NaN   \n",
       "4           amino acids             NaN                                   NaN   \n",
       "5           amino acids             NaN                                   NaN   \n",
       "6                   NaN         Sigma70  Human inference of promoter position   \n",
       "7           amino acids             NaN                                   NaN   \n",
       "8           amino acids             NaN                                   NaN   \n",
       "9   nitrogen metabolism             NaN                                   NaN   \n",
       "\n",
       "  TU_OBJECT_RI_TYPE TU_OBJECT_TYPE  EVIDENCE     OPERON_ID OPERON_NAME  \n",
       "0               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "1               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "2               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "3               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "4               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "5               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "6               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "7               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "8               NaN      predicted       NaN  ECK120014360    astCADBE  \n",
       "9               NaN      predicted       NaN  ECK120014360    astCADBE  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tu_objects_df = pd.read_pickle(\"./data/TU_objects_df.pkl\")\n",
    "\n",
    "def get_operon_id(TU_ID, TU_df):\n",
    "    operon_id = \"\"\n",
    "    df = TU_df[TU_df[\"TRANSCRIPTION_UNIT_ID\"]==TU_ID]\n",
    "    if len(df) > 0:\n",
    "        operon_id = list(df[\"OPERON_ID\"])[0]\n",
    "    return operon_id\n",
    "\n",
    "\n",
    "def get_operon_name_from_operon_ID(operon_ID, operon_df):\n",
    "    operon_name = \"unknown\"\n",
    "    df = operon_df[operon_df[\"OPERON_ID\"]==operon_ID]\n",
    "    if len(df) > 0:\n",
    "        operon_name = list(df[\"OPERON_NAME\"])[0]\n",
    "    return operon_name\n",
    "\n",
    "\n",
    "tu_objects_df[\"OPERON_ID\"] = tu_objects_df[\"TRANSCRIPTION_UNIT_ID\"].apply(get_operon_id, args=[TU_df])\n",
    "tu_objects_df[\"OPERON_NAME\"] = tu_objects_df[\"OPERON_ID\"].apply(get_operon_name_from_operon_ID, args=[operon_df])\n",
    "\n",
    "tu_objects_df[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3562, 15)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK125258528</td>\n",
       "      <td>485709</td>\n",
       "      <td>485732</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ECK120033472</td>\n",
       "      <td>acrAB</td>\n",
       "      <td>-</td>\n",
       "      <td>acrAp</td>\n",
       "      <td>-22.5</td>\n",
       "      <td>gcgttagattTACATACATTTGTGAATGTATGTAccatagcacg</td>\n",
       "      <td>[BCE|W|Binding of cellular extracts],[GEA|W|Ge...</td>\n",
       "      <td>Strong</td>\n",
       "      <td>(485709, 485732)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK125258528</td>\n",
       "      <td>485709</td>\n",
       "      <td>485732</td>\n",
       "      <td>forward</td>\n",
       "      <td>ECK125134945</td>\n",
       "      <td>acrR</td>\n",
       "      <td>-</td>\n",
       "      <td>acrRp</td>\n",
       "      <td>22.5</td>\n",
       "      <td>cgtgctatggTACATACATTCACAAATGTATGTAaatctaacgc</td>\n",
       "      <td>[BCE|W|Binding of cellular extracts],[GEA|W|Ge...</td>\n",
       "      <td>Strong</td>\n",
       "      <td>(485709, 485732)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK125202663</td>\n",
       "      <td>1619048</td>\n",
       "      <td>1619058</td>\n",
       "      <td>forward</td>\n",
       "      <td>ECK125202664</td>\n",
       "      <td>marRAB</td>\n",
       "      <td>-</td>\n",
       "      <td>marRp</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>catcggtcaaTTCATTCATTtgacttatac</td>\n",
       "      <td>[GEA|W|Gene expression analysis],[BPP|S|Bindin...</td>\n",
       "      <td>Strong</td>\n",
       "      <td>(1619048, 1619058)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK125242724</td>\n",
       "      <td>1978422</td>\n",
       "      <td>1978432</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ECK125242725</td>\n",
       "      <td>flhDC</td>\n",
       "      <td>-</td>\n",
       "      <td>flhDp</td>\n",
       "      <td>-31.5</td>\n",
       "      <td>tcactacacgCACATACAACggaggggggc</td>\n",
       "      <td>[GEA|W|Gene expression analysis],[HIBSCS|W|Hum...</td>\n",
       "      <td>Weak</td>\n",
       "      <td>(1978422, 1978432)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK120035040</td>\n",
       "      <td>2313112</td>\n",
       "      <td>2313135</td>\n",
       "      <td>forward</td>\n",
       "      <td>ECK120035041</td>\n",
       "      <td>micF</td>\n",
       "      <td>-</td>\n",
       "      <td>micFp</td>\n",
       "      <td>41.0</td>\n",
       "      <td>atttattaccGTCATTCATTTCTGAATGTCTGTTtacccctatt</td>\n",
       "      <td>[AIBSCS|W|Automated inference based on similar...</td>\n",
       "      <td>Weak</td>\n",
       "      <td>(2313112, 2313135)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0     1             2        3        4        5             6  \\\n",
       "0  ECK120015994  AcrR  ECK125258528   485709   485732  reverse  ECK120033472   \n",
       "1  ECK120015994  AcrR  ECK125258528   485709   485732  forward  ECK125134945   \n",
       "2  ECK120015994  AcrR  ECK125202663  1619048  1619058  forward  ECK125202664   \n",
       "3  ECK120015994  AcrR  ECK125242724  1978422  1978432  reverse  ECK125242725   \n",
       "4  ECK120015994  AcrR  ECK120035040  2313112  2313135  forward  ECK120035041   \n",
       "\n",
       "        7  8      9    10                                            11  \\\n",
       "0   acrAB  -  acrAp -22.5  gcgttagattTACATACATTTGTGAATGTATGTAccatagcacg   \n",
       "1    acrR  -  acrRp  22.5  cgtgctatggTACATACATTCACAAATGTATGTAaatctaacgc   \n",
       "2  marRAB  -  marRp -40.5                catcggtcaaTTCATTCATTtgacttatac   \n",
       "3   flhDC  -  flhDp -31.5                tcactacacgCACATACAACggaggggggc   \n",
       "4    micF  -  micFp  41.0  atttattaccGTCATTCATTTCTGAATGTCTGTTtacccctatt   \n",
       "\n",
       "                                                  12      13  \\\n",
       "0  [BCE|W|Binding of cellular extracts],[GEA|W|Ge...  Strong   \n",
       "1  [BCE|W|Binding of cellular extracts],[GEA|W|Ge...  Strong   \n",
       "2  [GEA|W|Gene expression analysis],[BPP|S|Bindin...  Strong   \n",
       "3  [GEA|W|Gene expression analysis],[HIBSCS|W|Hum...    Weak   \n",
       "4  [AIBSCS|W|Automated inference based on similar...    Weak   \n",
       "\n",
       "                range  \n",
       "0    (485709, 485732)  \n",
       "1    (485709, 485732)  \n",
       "2  (1619048, 1619058)  \n",
       "3  (1978422, 1978432)  \n",
       "4  (2313112, 2313135)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# tfbs_df = pd.read_csv(\"./data/RegulonDBwebsite10/BindingSiteSet.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "tfbs_df = pd.read_pickle(\"./data/TFBS_df.pkl\")\n",
    "\n",
    "\n",
    "def get_TFBS_range(tfbs_df_row):\n",
    "    r = ()\n",
    "    if not pd.isna(tfbs_df_row[3]) and not pd.isna(tfbs_df_row[4]):\n",
    "        r = (int(tfbs_df_row[3]), int(tfbs_df_row[4])) \n",
    "    return r\n",
    "\n",
    "\n",
    "tfbs_df[\"range\"] = tfbs_df.apply(get_TFBS_range, axis=1)\n",
    "display(tfbs_df.shape, tfbs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120009842</td>\n",
       "      <td>galRp</td>\n",
       "      <td>forward</td>\n",
       "      <td>2976569.0</td>\n",
       "      <td>Sigma70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tcccacgatgaaaacacgccaccccttgaaccaacgggcgttttcc...</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2976509, 2976589)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120009843</td>\n",
       "      <td>lpxLp</td>\n",
       "      <td>reverse</td>\n",
       "      <td>1116709.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gcggcatgatatagcaattatcgataattaacatccacacatttta...</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(1116689, 1116769)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120009844</td>\n",
       "      <td>yceAp</td>\n",
       "      <td>forward</td>\n",
       "      <td>1116772.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gcaaatgtagcgtaaaatgtgtggatgttaattatcgataattgct...</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(1116712, 1116792)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120009845</td>\n",
       "      <td>mraZp</td>\n",
       "      <td>forward</td>\n",
       "      <td>89596.0</td>\n",
       "      <td>Sigma70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tatgccttgtgactggcttgacaagcttttcctcagctccgtaaac...</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>The contribution of the mraZp promoter to the ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(89536, 89616)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120009846</td>\n",
       "      <td>sohBp1</td>\n",
       "      <td>forward</td>\n",
       "      <td>1329284.0</td>\n",
       "      <td>Sigma70, Sigma38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aaatggatactttgtcatactttcgctgcaataacatctctgcgag...</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>We assigned a putative transcription start sit...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(1329224, 1329304)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0       1        2          3                 4   5   6   7   8  \\\n",
       "0  ECK120009842   galRp  forward  2976569.0           Sigma70 NaN NaN NaN NaN   \n",
       "1  ECK120009843   lpxLp  reverse  1116709.0               NaN NaN NaN NaN NaN   \n",
       "2  ECK120009844   yceAp  forward  1116772.0               NaN NaN NaN NaN NaN   \n",
       "3  ECK120009845   mraZp  forward    89596.0           Sigma70 NaN NaN NaN NaN   \n",
       "4  ECK120009846  sohBp1  forward  1329284.0  Sigma70, Sigma38 NaN NaN NaN NaN   \n",
       "\n",
       "                                                   9     10  \\\n",
       "0  tcccacgatgaaaacacgccaccccttgaaccaacgggcgttttcc...  ECK12   \n",
       "1  gcggcatgatatagcaattatcgataattaacatccacacatttta...  ECK12   \n",
       "2  gcaaatgtagcgtaaaatgtgtggatgttaattatcgataattgct...  ECK12   \n",
       "3  tatgccttgtgactggcttgacaagcttttcctcagctccgtaaac...  ECK12   \n",
       "4  aaatggatactttgtcatactttcgctgcaataacatctctgcgag...  ECK12   \n",
       "\n",
       "                                                  11  12               range  \n",
       "0                                                NaN NaN  (2976509, 2976589)  \n",
       "1                                                NaN NaN  (1116689, 1116769)  \n",
       "2                                                NaN NaN  (1116712, 1116792)  \n",
       "3  The contribution of the mraZp promoter to the ... NaN      (89536, 89616)  \n",
       "4  We assigned a putative transcription start sit... NaN  (1329224, 1329304)  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# promoter_df = pd.read_csv(\"./data/RegulonDB10/promoter.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "promoter_df = pd.read_pickle(\"./data/promoter_df.pkl\")\n",
    "promoter_df[\"range\"] = promoter_df.apply(get_promoter_range_from_RegulonDB_df_row, axis=1)\n",
    "promoter_df.head()\n",
    "\n",
    "\n",
    "# def get_TSS_range_from_row(promoter_df_row):\n",
    "#     r = ()\n",
    "#     promoter_seq_str = promoter_df_row[9]\n",
    "#     if not pd.isna(promoter_seq_str):\n",
    "#         TSS_genome_pos = int(promoter_df_row[3])\n",
    "#         r = (TSS_genome_pos, TSS_genome_pos)\n",
    "#     return r\n",
    "\n",
    "# promoter_df[\"TSS range\"] = promoter_df.apply(get_TSS_range_from_row, axis=1)\n",
    "# promoter_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RegulonDB ID</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK125143526</td>\n",
       "      <td>(200, 311)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK125143530</td>\n",
       "      <td>(4979, 5078)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK125143534</td>\n",
       "      <td>(14134, 14155)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK125143536</td>\n",
       "      <td>(21166, 21255)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK125143540</td>\n",
       "      <td>(20912, 20982)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   RegulonDB ID           range\n",
       "0  ECK125143526      (200, 311)\n",
       "1  ECK125143530    (4979, 5078)\n",
       "2  ECK125143534  (14134, 14155)\n",
       "3  ECK125143536  (21166, 21255)\n",
       "4  ECK125143540  (20912, 20982)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "att_term_df = pd.read_pickle(\"./data/att_term_df.pkl\")\n",
    "att_term_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120010779</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2738912</td>\n",
       "      <td>2738940</td>\n",
       "      <td>rho-independent</td>\n",
       "      <td>ctgatgaaaaGGTGCCGGATGATGTGAATCATCCGGCACtggattatta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(2738912, 2738940)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120010780</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2684075</td>\n",
       "      <td>2684093</td>\n",
       "      <td>rho-independent</td>\n",
       "      <td>taacgtagaaAGGCTTCCCGAAGGAAGCCttgatgatca</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(2684075, 2684093)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120010781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2311610</td>\n",
       "      <td>2311624</td>\n",
       "      <td>rho-independent</td>\n",
       "      <td>caatgaaaaaAGGGCCCGCAGGCCCtttgttcgat</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(2311610, 2311624)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120010782</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1159325</td>\n",
       "      <td>1159346</td>\n",
       "      <td>rho-independent</td>\n",
       "      <td>tggggagactAAGGCAGCCAGATGGCTGCCTTttttacaggt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(1159325, 1159346)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120010783</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1113532</td>\n",
       "      <td>1113560</td>\n",
       "      <td>rho-independent</td>\n",
       "      <td>acgagccaatAAAAATACCGGCGTTATGCCGGTATTTTTttacgaaaga</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(1113532, 1113560)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0   1        2        3                4  \\\n",
       "0  ECK120010779 NaN  2738912  2738940  rho-independent   \n",
       "1  ECK120010780 NaN  2684075  2684093  rho-independent   \n",
       "2  ECK120010781 NaN  2311610  2311624  rho-independent   \n",
       "3  ECK120010782 NaN  1159325  1159346  rho-independent   \n",
       "4  ECK120010783 NaN  1113532  1113560  rho-independent   \n",
       "\n",
       "                                                   5    6   7      8  \\\n",
       "0  ctgatgaaaaGGTGCCGGATGATGTGAATCATCCGGCACtggattatta  NaN NaN  ECK12   \n",
       "1            taacgtagaaAGGCTTCCCGAAGGAAGCCttgatgatca  NaN NaN  ECK12   \n",
       "2                caatgaaaaaAGGGCCCGCAGGCCCtttgttcgat  NaN NaN  ECK12   \n",
       "3         tggggagactAAGGCAGCCAGATGGCTGCCTTttttacaggt  NaN NaN  ECK12   \n",
       "4  acgagccaatAAAAATACCGGCGTTATGCCGGTATTTTTttacgaaaga  NaN NaN  ECK12   \n",
       "\n",
       "                range  \n",
       "0  (2738912, 2738940)  \n",
       "1  (2684075, 2684093)  \n",
       "2  (2311610, 2311624)  \n",
       "3  (1159325, 1159346)  \n",
       "4  (1113532, 1113560)  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# terminator_df = pd.read_csv(\"./data/RegulonDB10/terminator.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "terminator_df = pd.read_pickle(\"./data/term_df.pkl\")\n",
    "terminator_df[\"range\"] = terminator_df.apply(lambda row: (row[2], row[3]), axis=1)\n",
    "terminator_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SHINE_DALGARNO_ID</th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>SHINE_DALGARNO_DIST_GENE</th>\n",
       "      <th>SHINE_DALGARNO_POSLEFT</th>\n",
       "      <th>SHINE_DALGARNO_POSRIGHT</th>\n",
       "      <th>SHINE_DALGARNO_SEQUENCE</th>\n",
       "      <th>SHINE_DALGARNO_NOTE</th>\n",
       "      <th>SD_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120014181</td>\n",
       "      <td>ECK120000266</td>\n",
       "      <td>-11</td>\n",
       "      <td>3151252</td>\n",
       "      <td>3151257</td>\n",
       "      <td>aaattacgcgCAGGATaatatccGAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3151252, 3151257)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120014182</td>\n",
       "      <td>ECK120000265</td>\n",
       "      <td>-9</td>\n",
       "      <td>3151991</td>\n",
       "      <td>3151996</td>\n",
       "      <td>acttgcgtccTGGAGAtacacAGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3151991, 3151996)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120014183</td>\n",
       "      <td>ECK120000496</td>\n",
       "      <td>-11</td>\n",
       "      <td>3957829</td>\n",
       "      <td>3957834</td>\n",
       "      <td>acgtcaacatCGAGGGctgtcccTGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3957829, 3957834)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120014184</td>\n",
       "      <td>ECK120000488</td>\n",
       "      <td>-10</td>\n",
       "      <td>3957957</td>\n",
       "      <td>3957962</td>\n",
       "      <td>cacaacatcaCGAGGAatcaccATG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3957957, 3957962)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120014185</td>\n",
       "      <td>ECK120001215</td>\n",
       "      <td>-8</td>\n",
       "      <td>3469859</td>\n",
       "      <td>3469864</td>\n",
       "      <td>tttacgtcacAAGGGAttatAAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3469859, 3469864)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SHINE_DALGARNO_ID       GENE_ID  SHINE_DALGARNO_DIST_GENE  \\\n",
       "0      ECK120014181  ECK120000266                       -11   \n",
       "1      ECK120014182  ECK120000265                        -9   \n",
       "2      ECK120014183  ECK120000496                       -11   \n",
       "3      ECK120014184  ECK120000488                       -10   \n",
       "4      ECK120014185  ECK120001215                        -8   \n",
       "\n",
       "   SHINE_DALGARNO_POSLEFT  SHINE_DALGARNO_POSRIGHT  \\\n",
       "0                 3151252                  3151257   \n",
       "1                 3151991                  3151996   \n",
       "2                 3957829                  3957834   \n",
       "3                 3957957                  3957962   \n",
       "4                 3469859                  3469864   \n",
       "\n",
       "      SHINE_DALGARNO_SEQUENCE  SHINE_DALGARNO_NOTE  SD_INTERNAL_COMMENT  \\\n",
       "0  aaattacgcgCAGGATaatatccGAT                  NaN                  NaN   \n",
       "1    acttgcgtccTGGAGAtacacAGT                  NaN                  NaN   \n",
       "2  acgtcaacatCGAGGGctgtcccTGT                  NaN                  NaN   \n",
       "3   cacaacatcaCGAGGAatcaccATG                  NaN                  NaN   \n",
       "4     tttacgtcacAAGGGAttatAAT                  NaN                  NaN   \n",
       "\n",
       "  KEY_ID_ORG               range  \n",
       "0      ECK12  (3151252, 3151257)  \n",
       "1      ECK12  (3151991, 3151996)  \n",
       "2      ECK12  (3957829, 3957834)  \n",
       "3      ECK12  (3957957, 3957962)  \n",
       "4      ECK12  (3469859, 3469864)  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RBS_df = pd.read_pickle(\"./data/RBS_df.pkl\")\n",
    "RBS_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13957</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>M599L (ATG→TTG)</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(13957, 13957)</td>\n",
       "      <td>{ECK120000235}</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28175</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>W295R (TGG→AGG)</td>\n",
       "      <td>rihC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(28175, 28175)</td>\n",
       "      <td>{ECK120001070}</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101342</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>T193I (ACC→ATC)</td>\n",
       "      <td>murC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(101342, 101342)</td>\n",
       "      <td>{ECK120000612}</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145691</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A204V (GCC→GTC)</td>\n",
       "      <td>yadE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(145691, 145691)</td>\n",
       "      <td>{ECK120001687}</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171072</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→G</td>\n",
       "      <td>K166K (AAA→AAG)</td>\n",
       "      <td>fhuD</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(171072, 171072)</td>\n",
       "      <td>{ECK120000299}</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "1392  GLU    3    412        2         1       1.0     13957           SNP   \n",
       "1428  GLU    3    412        2         1       1.0     28175           SNP   \n",
       "1393  GLU    3    412        2         1       1.0    101342           SNP   \n",
       "1394  GLU    3    412        2         1       1.0    145691           SNP   \n",
       "1429  GLU    3    412        2         1       1.0    171072           SNP   \n",
       "\n",
       "     Sequence Change          Details mutation target annotation  \\\n",
       "1392             A→T  M599L (ATG→TTG)                       dnaK   \n",
       "1428             T→A  W295R (TGG→AGG)                       rihC   \n",
       "1393             C→T  T193I (ACC→ATC)                       murC   \n",
       "1394             C→T  A204V (GCC→GTC)                       yadE   \n",
       "1429             A→G  K166K (AAA→AAG)                       fhuD   \n",
       "\n",
       "     Reference Seq     sample  coding             range gene RegulonDB ID  \\\n",
       "1392     NC_000913  3 412 2 1    True    (13957, 13957)    {ECK120000235}   \n",
       "1428     NC_000913  3 412 2 1    True    (28175, 28175)    {ECK120001070}   \n",
       "1393     NC_000913  3 412 2 1    True  (101342, 101342)    {ECK120000612}   \n",
       "1394     NC_000913  3 412 2 1    True  (145691, 145691)    {ECK120001687}   \n",
       "1429     NC_000913  3 412 2 1    True  (171072, 171072)    {ECK120000299}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1392  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1428  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  False       False   \n",
       "1393  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  False       False   \n",
       "1394  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  False       False   \n",
       "1429  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  False       False   \n",
       "\n",
       "     TFBS promoter RBS terminator  genetic  \n",
       "1392   {}       {}  {}         {}     True  \n",
       "1428   {}       {}  {}         {}     True  \n",
       "1393   {}       {}  {}         {}     True  \n",
       "1394   {}       {}  {}         {}     True  \n",
       "1429   {}       {}  {}         {}     True  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from mutil.mut import is_genetic_mut\n",
    "all_muts_df[\"genetic\"] = all_muts_df[\"Details\"].apply(is_genetic_mut)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build component annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13957</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>M599L (ATG→TTG)</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(13957, 13957)</td>\n",
       "      <td>{ECK120000235}</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28175</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>W295R (TGG→AGG)</td>\n",
       "      <td>rihC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(28175, 28175)</td>\n",
       "      <td>{ECK120001070}</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101342</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>T193I (ACC→ATC)</td>\n",
       "      <td>murC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(101342, 101342)</td>\n",
       "      <td>{ECK120000612}</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145691</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A204V (GCC→GTC)</td>\n",
       "      <td>yadE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(145691, 145691)</td>\n",
       "      <td>{ECK120001687}</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171072</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→G</td>\n",
       "      <td>K166K (AAA→AAG)</td>\n",
       "      <td>fhuD</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(171072, 171072)</td>\n",
       "      <td>{ECK120000299}</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "1392  GLU    3    412        2         1       1.0     13957           SNP   \n",
       "1428  GLU    3    412        2         1       1.0     28175           SNP   \n",
       "1393  GLU    3    412        2         1       1.0    101342           SNP   \n",
       "1394  GLU    3    412        2         1       1.0    145691           SNP   \n",
       "1429  GLU    3    412        2         1       1.0    171072           SNP   \n",
       "\n",
       "     Sequence Change          Details mutation target annotation  \\\n",
       "1392             A→T  M599L (ATG→TTG)                       dnaK   \n",
       "1428             T→A  W295R (TGG→AGG)                       rihC   \n",
       "1393             C→T  T193I (ACC→ATC)                       murC   \n",
       "1394             C→T  A204V (GCC→GTC)                       yadE   \n",
       "1429             A→G  K166K (AAA→AAG)                       fhuD   \n",
       "\n",
       "     Reference Seq     sample  coding             range gene RegulonDB ID  \\\n",
       "1392     NC_000913  3 412 2 1    True    (13957, 13957)    {ECK120000235}   \n",
       "1428     NC_000913  3 412 2 1    True    (28175, 28175)    {ECK120001070}   \n",
       "1393     NC_000913  3 412 2 1    True  (101342, 101342)    {ECK120000612}   \n",
       "1394     NC_000913  3 412 2 1    True  (145691, 145691)    {ECK120001687}   \n",
       "1429     NC_000913  3 412 2 1    True  (171072, 171072)    {ECK120000299}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1392  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1428  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  False       False   \n",
       "1393  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  False       False   \n",
       "1394  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  False       False   \n",
       "1429  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  False       False   \n",
       "\n",
       "     TFBS promoter RBS terminator  genetic  \\\n",
       "1392   {}       {}  {}         {}     True   \n",
       "1428   {}       {}  {}         {}     True   \n",
       "1393   {}       {}  {}         {}     True   \n",
       "1394   {}       {}  {}         {}     True   \n",
       "1429   {}       {}  {}         {}     True   \n",
       "\n",
       "                                       genomic features  \n",
       "1392  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  \n",
       "1428  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  \n",
       "1393  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  \n",
       "1394  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  \n",
       "1429  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NONCODING_FEATURE_COLUMNS = [\n",
    "    \"TFBS\",\n",
    "    \"promoter\",\n",
    "#     \"TSS\",\n",
    "    \"RBS\",\n",
    "    \"terminator\",\n",
    "#     \"attenuator terminator\"\n",
    "]\n",
    "\n",
    "\n",
    "CODING_FEATURE_COLUMNS = [\"gene RegulonDB ID\"]\n",
    "\n",
    "\n",
    "FEATURE_COLUMNS = NONCODING_FEATURE_COLUMNS + CODING_FEATURE_COLUMNS\n",
    "\n",
    "\n",
    "def get_gene_name(RegulonDB_ID, debug=False):\n",
    "    if debug:\n",
    "        display(RegulonDB_ID, gene_df[gene_df[\"GENE_ID\"] == RegulonDB_ID])\n",
    "    return gene_df[gene_df[\"GENE_ID\"] == RegulonDB_ID].iloc[0][\"GENE_NAME\"]  #Simply returning the first gene name in the second column that has the RegulonDB ID in the first column.\n",
    "\n",
    "\n",
    "def get_feature_range(feat_id, feat_type):\n",
    "    r = ()\n",
    "    # default parameters\n",
    "    feat_id_col = 0\n",
    "    range_col = \"range\"\n",
    "    if feat_type == \"gene RegulonDB ID\":\n",
    "        feat_id_col = \"GENE_ID\"\n",
    "        feat_type_df = gene_df\n",
    "    elif feat_type == \"TFBS\":\n",
    "        feat_id_col = 2\n",
    "        feat_type_df = tfbs_df\n",
    "    elif feat_type == \"promoter\":\n",
    "        feat_type_df = promoter_df\n",
    "#     elif feat_type == \"TSS\":\n",
    "#         feat_type_df = promoter_df\n",
    "#         range_col = \"TSS range\"\n",
    "    elif feat_type == \"RBS\":\n",
    "        feat_id_col = \"SHINE_DALGARNO_ID\"\n",
    "        feat_type_df = RBS_df\n",
    "    elif feat_type == \"terminator\":\n",
    "        feat_type_df = terminator_df\n",
    "    elif feat_type == \"attenuator terminator\":\n",
    "        feat_id_col = \"RegulonDB ID\"\n",
    "        feat_type_df = att_term_df\n",
    "    \n",
    "    df = feat_type_df[feat_type_df[feat_id_col] == feat_id]\n",
    "    if not df.empty:\n",
    "        r = df.iloc[0][range_col]  # Assuming all rows describe the same range.\n",
    "    l = [int(i) for i in r]\n",
    "    return tuple(l)\n",
    "\n",
    "\n",
    "def _get_operon_name(TU_object_ID):\n",
    "    operon_name = \"unknown\"\n",
    "    df = tu_objects_df[tu_objects_df[\"TU_OBJECT_ID\"] == TU_object_ID]\n",
    "    if len(df) > 0:\n",
    "        # non-coding features can be associated to more than one TU.\n",
    "        operon_name_set = set([])\n",
    "        for _, row in df.iterrows():\n",
    "            operon_name_set.add(row[\"OPERON_NAME\"])\n",
    "        operon_name = ' '.join(operon_name_set)\n",
    "    return operon_name\n",
    "\n",
    "\n",
    "def get_genomic_feature_name(feat_id, feat_type):\n",
    "    name = \"\"\n",
    "    if feat_type == \"gene RegulonDB ID\":\n",
    "        name = get_gene_name(feat_id)\n",
    "    elif feat_type == \"TFBS\":  # TFBS can be assoicated to more than one TU.\n",
    "        for _, row in tfbs_df[tfbs_df[2]==feat_id].iterrows():\n",
    "            name += row[7] + \" \"\n",
    "        name += \"TFBS\"\n",
    "    elif feat_type == \"promoter\":\n",
    "        name = promoter_df[promoter_df[0]==feat_id].iloc[0][1]\n",
    "#     elif feat_type == \"TSS\":\n",
    "#         name = promoter_df[promoter_df[0]==feat_id].iloc[0][1] + \" TSS\"\n",
    "    elif feat_type == \"RBS\":\n",
    "        gene_id = RBS_df[RBS_df[\"SHINE_DALGARNO_ID\"]==feat_id].iloc[0][1]\n",
    "        name = get_gene_name(gene_id) + \" RBS\"\n",
    "    elif feat_type == \"terminator\":\n",
    "        name = _get_operon_name(feat_id) + \" \" + \"terminator\"\n",
    "    elif feat_type == \"attenuator terminator\":\n",
    "        name = _get_operon_name(feat_id) + \" \" + \"attenuator terminator\"\n",
    "    return name\n",
    "\n",
    "\n",
    "def get_genomic_features(mut_row, ignore_RegDB_feats=False):\n",
    "    annots = []\n",
    "    if not ignore_RegDB_feats:\n",
    "        for feat_col in FEATURE_COLUMNS:\n",
    "            for feat_id in mut_row[feat_col]:\n",
    "                d = {\n",
    "                    \"name\": get_genomic_feature_name(feat_id, feat_col),\n",
    "                    \"RegulonDB ID\": feat_id,\n",
    "                    \"range\": get_feature_range(feat_id, feat_col),\n",
    "                    \"genetic\": False if feat_col in NONCODING_FEATURE_COLUMNS else True,  # could be pseudogene which is non-coding though still genetic.\n",
    "                    \"feature type\": \"gene\" if feat_col == \"gene RegulonDB ID\" else feat_col,\n",
    "                    \"operon\": _get_operon_name(feat_id)\n",
    "                }\n",
    "                annots.append(d)\n",
    "                \n",
    "    if len(annots) == 0:\n",
    "        annots = mut_row[\"genetic features\"]  # Just use what's in the genetic features col.\n",
    "        for a in annots:\n",
    "            a[\"genetic\"] = False if '/' in a[\"RegulonDB ID\"] else True\n",
    "            # Currently not getting operon here for integenic. Rather only getting it within the operon NB, where it's used for intergenic.\n",
    "        \n",
    "    return annots\n",
    "\n",
    "all_muts_df[\"genomic features\"] = all_muts_df.apply(lambda r: get_genomic_features(r), axis=1)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define if features are found to overlap\n",
    "for _, r in all_muts_df.iterrows():\n",
    "    if len(r[\"genomic features\"]) > 1:\n",
    "        for i in range(0, len(r[\"genomic features\"])):\n",
    "            feat_d1 = r[\"genomic features\"][i]\n",
    "            for j in range(i+1, len(r[\"genomic features\"])):\n",
    "                feat_d2 = r[\"genomic features\"][j]                \n",
    "                if is_overlap(feat_d1[\"range\"], feat_d2[\"range\"]):\n",
    "                    if \"overlap\" not in feat_d1.keys(): feat_d1[\"overlap\"] = set()\n",
    "                    feat_d1[\"overlap\"].add(feat_d2[\"name\"])\n",
    "                    if \"overlap\" not in feat_d2.keys(): feat_d2[\"overlap\"] = set()\n",
    "                    feat_d2[\"overlap\"].add(feat_d1[\"name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove terminators that overlap with attenuator-terminators\n",
    "# The following requires all indexes to be unique\n",
    "all_muts_df = all_muts_df.reset_index()\n",
    "\n",
    "for i, r in all_muts_df.iterrows():\n",
    "    if len(r[\"genomic features\"]) > 1:\n",
    "        idxs_to_remove = []\n",
    "        for k in range(0, len(r[\"genomic features\"])):\n",
    "            # finding a mutated attenuator-terminator that also overlaps with a mutated terminator from the same mutation.\n",
    "            feat_d = r[\"genomic features\"][k]\n",
    "            if \"overlap\" in feat_d.keys() \\\n",
    "            and feat_d[\"feature type\"] == \"terminator\" \\\n",
    "            and feat_d[\"operon\"]+' '+\"attenuator terminator\" in feat_d[\"overlap\"]:\n",
    "                idxs_to_remove.append(k)\n",
    "        if len(idxs_to_remove) > 0:\n",
    "            # removing the overlapping attenuator terminator\n",
    "            feats = [r[\"genomic features\"][j] for j in range(0, len(r[\"genomic features\"])) if j not in idxs_to_remove]\n",
    "            all_muts_df.at[i, \"genomic features\"]=feats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # test code to check if any features mutated within ALEdb are completely overlapped by a gene.\n",
    "\n",
    "\n",
    "# gene_df = pd.read_csv(\n",
    "#     \"./data/RegulonDB10/gene.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "# gene_df.columns = [\n",
    "#     \"GENE_ID\",\n",
    "#     \"GENE_NAME\",\n",
    "#     \"GENE_POSLEFT\",\n",
    "#     \"GENE_POSRIGHT\",\n",
    "#     \"GENE_STRAND\",\n",
    "#     \"GENE_SEQUENCE\",\n",
    "#     \"GC_CONTENT\",\n",
    "#     \"CRI_SCORE\",\n",
    "#     \"GENE_NOTE\",\n",
    "#     \"GENE_INTERNAL_COMMENT\",\n",
    "#     \"KEY_ID_ORG\",\n",
    "#     \"GENE_TYPE\"\n",
    "# ]\n",
    "# display(len(gene_df), gene_df.head())\n",
    "\n",
    "\n",
    "# for i, mut in all_muts_df.iterrows():\n",
    "#     for geno_feat_d in mut[\"genomic features\"]:\n",
    "#         if not geno_feat_d[\"genetic\"]:\n",
    "#             for j, gene in gene_df.iterrows():\n",
    "#                 if not pd.isna(gene.GENE_POSLEFT) and not pd.isna(gene.GENE_POSRIGHT):\n",
    "#                     if (geno_feat_d[\"range\"][0] >= gene.GENE_POSLEFT) and (geno_feat_d[\"range\"][1] <=gene.GENE_POSRIGHT):\n",
    "#                         display(geno_feat_d, gene, \"##########################################\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_muts_df.to_pickle(\"./data/2_2_df.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
