{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same feature tables are constantly being rebuilt in this pipeline, where some can take a substantial amount of time to build and annotate. Should only have to build these tables once (into dataframes) and just import them where needed.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SHINE_DALGARNO_ID</th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>SHINE_DALGARNO_DIST_GENE</th>\n",
       "      <th>SHINE_DALGARNO_POSLEFT</th>\n",
       "      <th>SHINE_DALGARNO_POSRIGHT</th>\n",
       "      <th>SHINE_DALGARNO_SEQUENCE</th>\n",
       "      <th>SHINE_DALGARNO_NOTE</th>\n",
       "      <th>SD_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120014181</td>\n",
       "      <td>ECK120000266</td>\n",
       "      <td>-11</td>\n",
       "      <td>3151252</td>\n",
       "      <td>3151257</td>\n",
       "      <td>aaattacgcgCAGGATaatatccGAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3151252, 3151257)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120014182</td>\n",
       "      <td>ECK120000265</td>\n",
       "      <td>-9</td>\n",
       "      <td>3151991</td>\n",
       "      <td>3151996</td>\n",
       "      <td>acttgcgtccTGGAGAtacacAGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3151991, 3151996)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120014183</td>\n",
       "      <td>ECK120000496</td>\n",
       "      <td>-11</td>\n",
       "      <td>3957829</td>\n",
       "      <td>3957834</td>\n",
       "      <td>acgtcaacatCGAGGGctgtcccTGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3957829, 3957834)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120014184</td>\n",
       "      <td>ECK120000488</td>\n",
       "      <td>-10</td>\n",
       "      <td>3957957</td>\n",
       "      <td>3957962</td>\n",
       "      <td>cacaacatcaCGAGGAatcaccATG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3957957, 3957962)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120014185</td>\n",
       "      <td>ECK120001215</td>\n",
       "      <td>-8</td>\n",
       "      <td>3469859</td>\n",
       "      <td>3469864</td>\n",
       "      <td>tttacgtcacAAGGGAttatAAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3469859, 3469864)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SHINE_DALGARNO_ID       GENE_ID  SHINE_DALGARNO_DIST_GENE  \\\n",
       "0      ECK120014181  ECK120000266                       -11   \n",
       "1      ECK120014182  ECK120000265                        -9   \n",
       "2      ECK120014183  ECK120000496                       -11   \n",
       "3      ECK120014184  ECK120000488                       -10   \n",
       "4      ECK120014185  ECK120001215                        -8   \n",
       "\n",
       "   SHINE_DALGARNO_POSLEFT  SHINE_DALGARNO_POSRIGHT  \\\n",
       "0                 3151252                  3151257   \n",
       "1                 3151991                  3151996   \n",
       "2                 3957829                  3957834   \n",
       "3                 3957957                  3957962   \n",
       "4                 3469859                  3469864   \n",
       "\n",
       "      SHINE_DALGARNO_SEQUENCE  SHINE_DALGARNO_NOTE  SD_INTERNAL_COMMENT  \\\n",
       "0  aaattacgcgCAGGATaatatccGAT                  NaN                  NaN   \n",
       "1    acttgcgtccTGGAGAtacacAGT                  NaN                  NaN   \n",
       "2  acgtcaacatCGAGGGctgtcccTGT                  NaN                  NaN   \n",
       "3   cacaacatcaCGAGGAatcaccATG                  NaN                  NaN   \n",
       "4     tttacgtcacAAGGGAttatAAT                  NaN                  NaN   \n",
       "\n",
       "  KEY_ID_ORG               range  \n",
       "0      ECK12  (3151252, 3151257)  \n",
       "1      ECK12  (3151991, 3151996)  \n",
       "2      ECK12  (3957829, 3957834)  \n",
       "3      ECK12  (3957957, 3957962)  \n",
       "4      ECK12  (3469859, 3469864)  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RBS_df = pd.read_csv(\n",
    "    \"./data/RegulonDB10/shine_dalgarno.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "RBS_df.columns = [\n",
    "    \"SHINE_DALGARNO_ID\",\n",
    "    \"GENE_ID\",\n",
    "    \"SHINE_DALGARNO_DIST_GENE\",\n",
    "    \"SHINE_DALGARNO_POSLEFT\",\n",
    "    \"SHINE_DALGARNO_POSRIGHT\",\n",
    "    \"SHINE_DALGARNO_SEQUENCE\",\n",
    "    \"SHINE_DALGARNO_NOTE\",\n",
    "    \"SD_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\",\n",
    "]\n",
    "\n",
    "\n",
    "RBS_df[\"range\"] = RBS_df.apply(\n",
    "    lambda row: (row[\"SHINE_DALGARNO_POSLEFT\"], row[\"SHINE_DALGARNO_POSRIGHT\"]), axis=1)\n",
    "RBS_df.to_pickle(\"./data/RBS_df.pkl\")\n",
    "RBS_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OPERON_ID</th>\n",
       "      <th>OPERON_NAME</th>\n",
       "      <th>FIRSTGENEPOSLEFT</th>\n",
       "      <th>LASTGENEPOSRIGHT</th>\n",
       "      <th>REGULATIONPOSLEFT</th>\n",
       "      <th>REGULATIONPOSRIGHT</th>\n",
       "      <th>OPERON_STRAND</th>\n",
       "      <th>OPERON_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120011191</td>\n",
       "      <td>rob</td>\n",
       "      <td>4634441</td>\n",
       "      <td>4635310</td>\n",
       "      <td>4634441</td>\n",
       "      <td>4635382</td>\n",
       "      <td>reverse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(4634441, 4635382)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120011203</td>\n",
       "      <td>argP</td>\n",
       "      <td>3059753</td>\n",
       "      <td>3060646</td>\n",
       "      <td>3059684</td>\n",
       "      <td>3060680</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3059684, 3060680)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120011329</td>\n",
       "      <td>lrhA</td>\n",
       "      <td>2405703</td>\n",
       "      <td>2406641</td>\n",
       "      <td>2405703</td>\n",
       "      <td>2406866</td>\n",
       "      <td>reverse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(2405703, 2406866)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120011490</td>\n",
       "      <td>rtcR</td>\n",
       "      <td>3558268</td>\n",
       "      <td>3559866</td>\n",
       "      <td>3558268</td>\n",
       "      <td>3559866</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3558268, 3559866)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120011629</td>\n",
       "      <td>lrp</td>\n",
       "      <td>932595</td>\n",
       "      <td>933089</td>\n",
       "      <td>932265</td>\n",
       "      <td>933089</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(932265, 933089)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      OPERON_ID OPERON_NAME  FIRSTGENEPOSLEFT  LASTGENEPOSRIGHT  \\\n",
       "0  ECK120011191         rob           4634441           4635310   \n",
       "1  ECK120011203        argP           3059753           3060646   \n",
       "2  ECK120011329        lrhA           2405703           2406641   \n",
       "3  ECK120011490        rtcR           3558268           3559866   \n",
       "4  ECK120011629         lrp            932595            933089   \n",
       "\n",
       "   REGULATIONPOSLEFT  REGULATIONPOSRIGHT OPERON_STRAND  \\\n",
       "0            4634441             4635382       reverse   \n",
       "1            3059684             3060680       forward   \n",
       "2            2405703             2406866       reverse   \n",
       "3            3558268             3559866       forward   \n",
       "4             932265              933089       forward   \n",
       "\n",
       "   OPERON_INTERNAL_COMMENT KEY_ID_ORG               range  \n",
       "0                      NaN      ECK12  (4634441, 4635382)  \n",
       "1                      NaN      ECK12  (3059684, 3060680)  \n",
       "2                      NaN      ECK12  (2405703, 2406866)  \n",
       "3                      NaN      ECK12  (3558268, 3559866)  \n",
       "4                      NaN      ECK12    (932265, 933089)  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)\n",
    "operon_df.to_pickle(\"./data/operon_df.pkl\")\n",
    "operon_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4375748"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# need to get total \"unknown\" operon category length\n",
    "# follow the same approach as with total \"unknown\" COG length (get_COG_lengths NB)\n",
    "# Could be faster\n",
    "all_op_pos_set = set()\n",
    "for i, r in operon_df.iterrows():\n",
    "    op_pos = set(range(r[\"REGULATIONPOSLEFT\"], r[\"REGULATIONPOSRIGHT\"] + 1))\n",
    "    all_op_pos_set = all_op_pos_set.union(op_pos)\n",
    "display(len(all_op_pos_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>operon description</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>unknown</td>\n",
       "      <td>265905</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  operon description  length\n",
       "0            unknown  265905"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unknown_op_pos_set = set(range(1 ,4641653 + 1)).difference(all_op_pos_set)\n",
    "unknown_op_nuc_len = len(unknown_op_pos_set)\n",
    "op_len_df = pd.DataFrame(\n",
    "    data={\"operon description\": [\"unknown\"], \"length\": [unknown_op_nuc_len]},\n",
    "    columns=[\"operon description\", \"length\"],\n",
    ")\n",
    "op_len_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "op_len_df.to_pickle(\"./data/operon_length_df.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# attenuator terminator\n",
    "The point of this notebook is to build those tables ASW-1216"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A_TERMINATOR_ID</th>\n",
       "      <th>A_TERMINATOR_TYPE</th>\n",
       "      <th>A_TERMINATOR_POSLEFT</th>\n",
       "      <th>A_TERMINATOR_POSRIGHT</th>\n",
       "      <th>A_TERMINATOR_ENERGY</th>\n",
       "      <th>A_TERMINATOR_SEQUENCE</th>\n",
       "      <th>A_TERMINATOR_ATTENUATOR_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK125143527</td>\n",
       "      <td>terminator</td>\n",
       "      <td>276</td>\n",
       "      <td>311</td>\n",
       "      <td>-20.00</td>\n",
       "      <td>aacacagaaaAAAGCCCGCACCTGACAGTGCGGGCTTTTTTTTTCg...</td>\n",
       "      <td>ECK125143526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK125143528</td>\n",
       "      <td>anti-terminator</td>\n",
       "      <td>244</td>\n",
       "      <td>288</td>\n",
       "      <td>-19.41</td>\n",
       "      <td>acaggtaacgGTGCGGGCTGACGCGTACAGGAAACACAGAAAAAAG...</td>\n",
       "      <td>ECK125143526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK125143529</td>\n",
       "      <td>anti-anti-terminator</td>\n",
       "      <td>200</td>\n",
       "      <td>255</td>\n",
       "      <td>-14.50</td>\n",
       "      <td>tgaaacgcatTAGCACCACCATTACCACCACCATCACCATTACCAC...</td>\n",
       "      <td>ECK125143526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK125143531</td>\n",
       "      <td>terminator</td>\n",
       "      <td>5042</td>\n",
       "      <td>5078</td>\n",
       "      <td>-19.30</td>\n",
       "      <td>atctcaatcaGGCCGGGTTTGCTTTTATGCAGCCCGGCTTTTTTAT...</td>\n",
       "      <td>ECK125143530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK125143532</td>\n",
       "      <td>anti-terminator</td>\n",
       "      <td>4979</td>\n",
       "      <td>5051</td>\n",
       "      <td>-11.90</td>\n",
       "      <td>ctgcccgccgATTTTGCTGCGTTGCGTAAATTGATGATGAATCATC...</td>\n",
       "      <td>ECK125143530</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  A_TERMINATOR_ID     A_TERMINATOR_TYPE  A_TERMINATOR_POSLEFT  \\\n",
       "0    ECK125143527            terminator                   276   \n",
       "1    ECK125143528       anti-terminator                   244   \n",
       "2    ECK125143529  anti-anti-terminator                   200   \n",
       "3    ECK125143531            terminator                  5042   \n",
       "4    ECK125143532       anti-terminator                  4979   \n",
       "\n",
       "   A_TERMINATOR_POSRIGHT  A_TERMINATOR_ENERGY  \\\n",
       "0                    311               -20.00   \n",
       "1                    288               -19.41   \n",
       "2                    255               -14.50   \n",
       "3                   5078               -19.30   \n",
       "4                   5051               -11.90   \n",
       "\n",
       "                               A_TERMINATOR_SEQUENCE  \\\n",
       "0  aacacagaaaAAAGCCCGCACCTGACAGTGCGGGCTTTTTTTTTCg...   \n",
       "1  acaggtaacgGTGCGGGCTGACGCGTACAGGAAACACAGAAAAAAG...   \n",
       "2  tgaaacgcatTAGCACCACCATTACCACCACCATCACCATTACCAC...   \n",
       "3  atctcaatcaGGCCGGGTTTGCTTTTATGCAGCCCGGCTTTTTTAT...   \n",
       "4  ctgcccgccgATTTTGCTGCGTTGCGTAAATTGATGATGAATCATC...   \n",
       "\n",
       "  A_TERMINATOR_ATTENUATOR_ID  \n",
       "0               ECK125143526  \n",
       "1               ECK125143526  \n",
       "2               ECK125143526  \n",
       "3               ECK125143530  \n",
       "4               ECK125143530  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "att_term_df = pd.read_csv(\n",
    "    \"./data/RegulonDB10/attenuator_terminator.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "att_term_df.columns = [\n",
    "    \"A_TERMINATOR_ID\",\n",
    "    \"A_TERMINATOR_TYPE\",\n",
    "    \"A_TERMINATOR_POSLEFT\",\n",
    "    \"A_TERMINATOR_POSRIGHT\",\n",
    "    \"A_TERMINATOR_ENERGY\",\n",
    "    \"A_TERMINATOR_SEQUENCE\",\n",
    "    \"A_TERMINATOR_ATTENUATOR_ID\"\n",
    "]\n",
    "# att_term_df[\"range\"] = att_term_df.apply(lambda row: (row.A_TERMINATOR_POSLEFT, row.A_TERMINATOR_POSRIGHT), axis=1)\n",
    "att_term_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RegulonDB ID</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK125143526</td>\n",
       "      <td>(200, 311)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK125143530</td>\n",
       "      <td>(4979, 5078)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK125143534</td>\n",
       "      <td>(14134, 14155)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK125143536</td>\n",
       "      <td>(21166, 21255)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK125143540</td>\n",
       "      <td>(20912, 20982)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   RegulonDB ID           range\n",
       "0  ECK125143526      (200, 311)\n",
       "1  ECK125143530    (4979, 5078)\n",
       "2  ECK125143534  (14134, 14155)\n",
       "3  ECK125143536  (21166, 21255)\n",
       "4  ECK125143540  (20912, 20982)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_att_term_df = pd.DataFrame()\n",
    "for att_term_id, r in att_term_df.groupby([\"A_TERMINATOR_ATTENUATOR_ID\"]):\n",
    "    range_l = list(r.A_TERMINATOR_POSLEFT) + list(r.A_TERMINATOR_POSRIGHT)\n",
    "    combo_att_term_range = (min(range_l), max(range_l))\n",
    "    new_att_term_df = new_att_term_df.append({\"RegulonDB ID\": att_term_id, \"range\": combo_att_term_range}, ignore_index=True)\n",
    "new_att_term_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_att_term_df.to_pickle(\"./data/att_term_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "terminator_df = pd.read_csv(\"./data/RegulonDB10/terminator.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "terminator_df.to_pickle(\"./data/term_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "TU_df = pd.read_csv(\"./data/RegulonDB10/transcription_unit.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "TU_df.columns = [\n",
    "    \"TRANSCRIPTION_UNIT_ID\",\n",
    "    \"PROMOTER_ID\",\n",
    "    \"TRANSCRIPTION_UNIT_NAME\",\n",
    "    \"OPERON_ID\",\n",
    "    \"TRANSCRIPTION_UNIT_NOTE\",\n",
    "    \"TU_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "TU_df.to_pickle(\"./data/TU_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df = pd.read_csv(\n",
    "    \"./data/RegulonDB10/gene.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "gene_df.columns = [\n",
    "    \"GENE_ID\",\n",
    "    \"GENE_NAME\",\n",
    "    \"GENE_POSLEFT\",\n",
    "    \"GENE_POSRIGHT\",\n",
    "    \"GENE_STRAND\",\n",
    "    \"GENE_SEQUENCE\",\n",
    "    \"GC_CONTENT\",\n",
    "    \"CRI_SCORE\",\n",
    "    \"GENE_NOTE\",\n",
    "    \"GENE_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\",\n",
    "    \"GENE_TYPE\"\n",
    "]\n",
    "\n",
    "def get_gene_range(row):\n",
    "    r = ()\n",
    "    if not pd.isna(row[\"GENE_POSLEFT\"]) and not pd.isna(row[\"GENE_POSRIGHT\"]):\n",
    "        r = (int(row[\"GENE_POSLEFT\"]), int(row[\"GENE_POSRIGHT\"])) \n",
    "    return r\n",
    "\n",
    "gene_df[\"range\"] = gene_df.apply(lambda r: get_gene_range(r), axis=1)\n",
    "gene_df.to_pickle(\"./data/gene_df.pkl\")\n",
    "\n",
    "\n",
    "df = gene_df.copy()\n",
    "df = df.sort_values(by=[\"GENE_POSLEFT\"])  # Ordering the genes according to their left-most position.\n",
    "df = df.reset_index()\n",
    "\n",
    "intergenic_df = pd.DataFrame(columns=[\"name\", \"RegulonDB ID\", \"range\"])\n",
    "\n",
    "\n",
    "gene_idex_iterator = iter(range(0, len(df) - 1))  # Don't want to iterate to the last one since the gene will be considered in i-1\n",
    "for i in gene_idex_iterator:\n",
    "    # check if there even exists an intergenic region between genes.\n",
    "    if (df.loc[i + 1][\"GENE_POSLEFT\"] - df.loc[i][\"GENE_POSRIGHT\"]) > 1:\n",
    "        intergenic_df = intergenic_df.append(\n",
    "            {\"name\": str(df.loc[i][\"GENE_NAME\"]) + '/' + str(df.loc[i + 1][\"GENE_NAME\"]),\n",
    "             \"RegulonDB ID\": df.loc[i][\"GENE_ID\"] + '/' + df.loc[i + 1][\"GENE_ID\"],\n",
    "             \"range\": (int(df.loc[i][\"GENE_POSRIGHT\"] + 1), int(df.loc[i + 1][\"GENE_POSLEFT\"] - 1))},\n",
    "            ignore_index=True)\n",
    "    else:\n",
    "        # Checks if needs to skip next gene, otherwise will iterate into it.\n",
    "        if (df.loc[i + 1][\"GENE_POSRIGHT\"] - df.loc[i][\"GENE_POSRIGHT\"]) <= 0:\n",
    "            next(gene_idex_iterator)  # skips the next gene since is resides completely within the current\n",
    "\n",
    "intergenic_df.to_pickle(\"./data/intergenic_region_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "TU_objects_df = pd.read_csv(\"./data/RegulonDB10/tu_objects_tmp.txt\",\n",
    "                            sep=\"\\t\",\n",
    "                            comment='#',\n",
    "                            header=None,\n",
    "                           )\n",
    "\n",
    "TU_objects_df.columns = [\n",
    "\"TRANSCRIPTION_UNIT_ID\",\n",
    "\"NUMTU\",\n",
    "\"TU_POSLEFT\",\n",
    "\"TU_POSRIGHT\",\n",
    "\"TU_TYPE\",\n",
    "\"TU_OBJECT_CLASS\",\n",
    "\"TU_OBJECT_ID\",\n",
    "\"TU_OBJECT_NAME\",\n",
    "\"TU_OBJECT_POSLEFT\",\n",
    "\"TU_OBJECT_POSRIGHT\",\n",
    "\"TU_OBJECT_STRAND\",\n",
    "\"TU_OBJECT_COLORCLASS\",\n",
    "\"TU_OBJECT_DESCRIPTION\",\n",
    "\"TU_OBJECT_SIGMA\",\n",
    "\"TU_OBJECT_EVIDENCE\",\n",
    "\"TU_OBJECT_RI_TYPE\",\n",
    "\"TU_OBJECT_TYPE\",\n",
    "\"EVIDENCE\"\n",
    "]\n",
    "\n",
    "TU_objects_df.to_pickle(\"./data/TU_objects_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBJECT_ID</th>\n",
       "      <th>OBJECT_SYNONYM_NAME</th>\n",
       "      <th>OS_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>EG10001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>ECK4045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>b4053</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>b0764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      OBJECT_ID OBJECT_SYNONYM_NAME  OS_INTERNAL_COMMENT KEY_ID_ORG\n",
       "0  ECK120000001             EG10001                  NaN      ECK12\n",
       "1  ECK120000001             ECK4045                  NaN      ECK12\n",
       "2  ECK120000001               b4053                  NaN      ECK12\n",
       "3  ECK120000001                alr5                  NaN      ECK12\n",
       "4  ECK120000002               b0764                  NaN      ECK12"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_synonym_df = pd.read_csv(\n",
    "    \"./data/RegulonDB10/object_synonym.txt\",\n",
    "    sep=\"\\t\",\n",
    "    comment='#',\n",
    "    header=None,\n",
    "    quoting=3\n",
    ")\n",
    "gene_synonym_df.columns = [\"OBJECT_ID\", \"OBJECT_SYNONYM_NAME\", \"OS_INTERNAL_COMMENT\", \"KEY_ID_ORG\"]\n",
    "gene_synonym_df.to_pickle(\"./data/gene_synonym_df.pkl\")\n",
    "gene_synonym_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq_locus_tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>ec_description</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b0002</td>\n",
       "      <td>thrA</td>\n",
       "      <td>Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...</td>\n",
       "      <td>1.1.1.3</td>\n",
       "      <td>Homoserine dehydrogenase</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b0002</td>\n",
       "      <td>thrA</td>\n",
       "      <td>Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...</td>\n",
       "      <td>1.1.1.3</td>\n",
       "      <td>Homoserine dehydrogenase</td>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b0002</td>\n",
       "      <td>thrA</td>\n",
       "      <td>Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...</td>\n",
       "      <td>1.1.1.3</td>\n",
       "      <td>Homoserine dehydrogenase</td>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0002</td>\n",
       "      <td>thrA</td>\n",
       "      <td>Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...</td>\n",
       "      <td>2.7.2.4</td>\n",
       "      <td>Aspartate kinase</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b0002</td>\n",
       "      <td>thrA</td>\n",
       "      <td>Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...</td>\n",
       "      <td>2.7.2.4</td>\n",
       "      <td>Aspartate kinase</td>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refseq_locus_tag  gene                                            product  \\\n",
       "0            b0002  thrA  Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...   \n",
       "1            b0002  thrA  Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...   \n",
       "2            b0002  thrA  Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...   \n",
       "3            b0002  thrA  Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...   \n",
       "4            b0002  thrA  Aspartokinase (EC 2.7.2.4) / Homoserine dehydr...   \n",
       "\n",
       "  ec_number            ec_description  \\\n",
       "0   1.1.1.3  Homoserine dehydrogenase   \n",
       "1   1.1.1.3  Homoserine dehydrogenase   \n",
       "2   1.1.1.3  Homoserine dehydrogenase   \n",
       "3   2.7.2.4          Aspartate kinase   \n",
       "4   2.7.2.4          Aspartate kinase   \n",
       "\n",
       "                               pathway_name  \n",
       "0  Glycine, serine and threonine metabolism  \n",
       "1        Cysteine and methionine metabolism  \n",
       "2                       Lysine biosynthesis  \n",
       "3  Glycine, serine and threonine metabolism  \n",
       "4        Cysteine and methionine metabolism  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pathway_df = pd.read_csv(\n",
    "    \"./data/pathways/511145.12.PATRIC.pathway.tab\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "# Some of the entries have different entries for columsn in ec_description and ec_number and therefore can make duplicate entries when associating pathways to genes\n",
    "gene_pathway_df = gene_pathway_df.drop(columns=[\n",
    "    'genome_id',\n",
    "    'genome_name',\n",
    "    'patric_id',\n",
    "    'alt_locus_tag',\n",
    "    #  'gene',\n",
    "#     'product',\n",
    "#     'ec_number',\n",
    "#     'ec_description',\n",
    "    'pathway_id'\n",
    "])\n",
    "# gene_pathway_df = gene_pathway_df.drop_duplicates()\n",
    "gene_pathway_df.to_pickle(\"./data/gene_pathway_df.pkl\")\n",
    "gene_pathway_df.head()\n",
    "\n",
    "# display(gene_pathway_df[gene_pathway_df[\"refseq_locus_tag\"]==\"b3650\"], gene_pathway_df[gene_pathway_df[\"gene\"]==\"spoT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3441"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>COG category primary</th>\n",
       "      <th>COG class</th>\n",
       "      <th>COG description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>thrA</td>\n",
       "      <td>b0002</td>\n",
       "      <td>E</td>\n",
       "      <td>METABOLISM</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>metL</td>\n",
       "      <td>b3940</td>\n",
       "      <td>E</td>\n",
       "      <td>METABOLISM</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lysC</td>\n",
       "      <td>b4024</td>\n",
       "      <td>E</td>\n",
       "      <td>METABOLISM</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>thrB</td>\n",
       "      <td>b0003</td>\n",
       "      <td>E</td>\n",
       "      <td>METABOLISM</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>thrC</td>\n",
       "      <td>b0004</td>\n",
       "      <td>E</td>\n",
       "      <td>METABOLISM</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  locus COG category primary   COG class  \\\n",
       "0  thrA  b0002                    E  METABOLISM   \n",
       "1  metL  b3940                    E  METABOLISM   \n",
       "4  lysC  b4024                    E  METABOLISM   \n",
       "5  thrB  b0003                    E  METABOLISM   \n",
       "6  thrC  b0004                    E  METABOLISM   \n",
       "\n",
       "                       COG description  \n",
       "0  Amino acid transport and metabolism  \n",
       "1  Amino acid transport and metabolism  \n",
       "4  Amino acid transport and metabolism  \n",
       "5  Amino acid transport and metabolism  \n",
       "6  Amino acid transport and metabolism  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "COG_df = pd.read_csv(\"./data/cogs_ecoli_mg1655.csv\", encoding = \"ISO-8859-1\")\n",
    "COG_df = COG_df.drop([\"GI\", \"COG ID\", \"COG name\", \"COG category\"], axis=1).drop_duplicates()\n",
    "COG_df = COG_df[~(COG_df[\"COG description\"]==\"Function unknown\")]  # We want to ignore the \"Function unknown\" COG entries because we ignore it and it's mutation for the statistical work and mutflow visualizations\n",
    "COG_df.to_pickle(\"./data/COG_df.pkl\")\n",
    "display(len(COG_df), COG_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3562, 14)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK125258528</td>\n",
       "      <td>485709</td>\n",
       "      <td>485732</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ECK120033472</td>\n",
       "      <td>acrAB</td>\n",
       "      <td>-</td>\n",
       "      <td>acrAp</td>\n",
       "      <td>-22.5</td>\n",
       "      <td>gcgttagattTACATACATTTGTGAATGTATGTAccatagcacg</td>\n",
       "      <td>[BCE|W|Binding of cellular extracts],[GEA|W|Ge...</td>\n",
       "      <td>Strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK125258528</td>\n",
       "      <td>485709</td>\n",
       "      <td>485732</td>\n",
       "      <td>forward</td>\n",
       "      <td>ECK125134945</td>\n",
       "      <td>acrR</td>\n",
       "      <td>-</td>\n",
       "      <td>acrRp</td>\n",
       "      <td>22.5</td>\n",
       "      <td>cgtgctatggTACATACATTCACAAATGTATGTAaatctaacgc</td>\n",
       "      <td>[BCE|W|Binding of cellular extracts],[GEA|W|Ge...</td>\n",
       "      <td>Strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK125202663</td>\n",
       "      <td>1619048</td>\n",
       "      <td>1619058</td>\n",
       "      <td>forward</td>\n",
       "      <td>ECK125202664</td>\n",
       "      <td>marRAB</td>\n",
       "      <td>-</td>\n",
       "      <td>marRp</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>catcggtcaaTTCATTCATTtgacttatac</td>\n",
       "      <td>[GEA|W|Gene expression analysis],[BPP|S|Bindin...</td>\n",
       "      <td>Strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK125242724</td>\n",
       "      <td>1978422</td>\n",
       "      <td>1978432</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ECK125242725</td>\n",
       "      <td>flhDC</td>\n",
       "      <td>-</td>\n",
       "      <td>flhDp</td>\n",
       "      <td>-31.5</td>\n",
       "      <td>tcactacacgCACATACAACggaggggggc</td>\n",
       "      <td>[GEA|W|Gene expression analysis],[HIBSCS|W|Hum...</td>\n",
       "      <td>Weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120015994</td>\n",
       "      <td>AcrR</td>\n",
       "      <td>ECK120035040</td>\n",
       "      <td>2313112</td>\n",
       "      <td>2313135</td>\n",
       "      <td>forward</td>\n",
       "      <td>ECK120035041</td>\n",
       "      <td>micF</td>\n",
       "      <td>-</td>\n",
       "      <td>micFp</td>\n",
       "      <td>41.0</td>\n",
       "      <td>atttattaccGTCATTCATTTCTGAATGTCTGTTtacccctatt</td>\n",
       "      <td>[AIBSCS|W|Automated inference based on similar...</td>\n",
       "      <td>Weak</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0     1             2        3        4        5             6   \\\n",
       "0  ECK120015994  AcrR  ECK125258528   485709   485732  reverse  ECK120033472   \n",
       "1  ECK120015994  AcrR  ECK125258528   485709   485732  forward  ECK125134945   \n",
       "2  ECK120015994  AcrR  ECK125202663  1619048  1619058  forward  ECK125202664   \n",
       "3  ECK120015994  AcrR  ECK125242724  1978422  1978432  reverse  ECK125242725   \n",
       "4  ECK120015994  AcrR  ECK120035040  2313112  2313135  forward  ECK120035041   \n",
       "\n",
       "       7  8      9     10                                            11  \\\n",
       "0   acrAB  -  acrAp -22.5  gcgttagattTACATACATTTGTGAATGTATGTAccatagcacg   \n",
       "1    acrR  -  acrRp  22.5  cgtgctatggTACATACATTCACAAATGTATGTAaatctaacgc   \n",
       "2  marRAB  -  marRp -40.5                catcggtcaaTTCATTCATTtgacttatac   \n",
       "3   flhDC  -  flhDp -31.5                tcactacacgCACATACAACggaggggggc   \n",
       "4    micF  -  micFp  41.0  atttattaccGTCATTCATTTCTGAATGTCTGTTtacccctatt   \n",
       "\n",
       "                                                  12      13  \n",
       "0  [BCE|W|Binding of cellular extracts],[GEA|W|Ge...  Strong  \n",
       "1  [BCE|W|Binding of cellular extracts],[GEA|W|Ge...  Strong  \n",
       "2  [GEA|W|Gene expression analysis],[BPP|S|Bindin...  Strong  \n",
       "3  [GEA|W|Gene expression analysis],[HIBSCS|W|Hum...    Weak  \n",
       "4  [AIBSCS|W|Automated inference based on similar...    Weak  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tfbs_df = pd.read_csv(\"./data/RegulonDBwebsite10/BindingSiteSet.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "\n",
    "def get_TFBS_range(tfbs_df_row):\n",
    "    r = ()\n",
    "    if not pd.isna(tfbs_df_row[3]) and not pd.isna(tfbs_df_row[4]):\n",
    "        r = (int(tfbs_df_row[3]), int(tfbs_df_row[4])) \n",
    "    return r\n",
    "\n",
    "tfbs_df.to_pickle(\"./data/TFBS_df.pkl\")\n",
    "display(tfbs_df.shape, tfbs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "promoter_df = pd.read_csv(\"./data/RegulonDB10/promoter.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "promoter_df.to_pickle(\"./data/promoter_df.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
