{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('display.max_rows', 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "585"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1392</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13957</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>M599L (ATG→TTG)</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(13957, 13957)</td>\n",
       "      <td>{ECK120000235}</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1428</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28175</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>W295R (TGG→AGG)</td>\n",
       "      <td>rihC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(28175, 28175)</td>\n",
       "      <td>{ECK120001070}</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1393</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101342</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>T193I (ACC→ATC)</td>\n",
       "      <td>murC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(101342, 101342)</td>\n",
       "      <td>{ECK120000612}</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1394</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145691</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A204V (GCC→GTC)</td>\n",
       "      <td>yadE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(145691, 145691)</td>\n",
       "      <td>{ECK120001687}</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1429</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171072</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→G</td>\n",
       "      <td>K166K (AAA→AAG)</td>\n",
       "      <td>fhuD</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(171072, 171072)</td>\n",
       "      <td>{ECK120000299}</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0  1392  GLU    3    412        2         1       1.0     13957           SNP   \n",
       "1  1428  GLU    3    412        2         1       1.0     28175           SNP   \n",
       "2  1393  GLU    3    412        2         1       1.0    101342           SNP   \n",
       "3  1394  GLU    3    412        2         1       1.0    145691           SNP   \n",
       "4  1429  GLU    3    412        2         1       1.0    171072           SNP   \n",
       "\n",
       "  Sequence Change          Details mutation target annotation Reference Seq  \\\n",
       "0             A→T  M599L (ATG→TTG)                       dnaK     NC_000913   \n",
       "1             T→A  W295R (TGG→AGG)                       rihC     NC_000913   \n",
       "2             C→T  T193I (ACC→ATC)                       murC     NC_000913   \n",
       "3             C→T  A204V (GCC→GTC)                       yadE     NC_000913   \n",
       "4             A→G  K166K (AAA→AAG)                       fhuD     NC_000913   \n",
       "\n",
       "      sample  coding             range gene RegulonDB ID  \\\n",
       "0  3 412 2 1    True    (13957, 13957)    {ECK120000235}   \n",
       "1  3 412 2 1    True    (28175, 28175)    {ECK120001070}   \n",
       "2  3 412 2 1    True  (101342, 101342)    {ECK120000612}   \n",
       "3  3 412 2 1    True  (145691, 145691)    {ECK120001687}   \n",
       "4  3 412 2 1    True  (171072, 171072)    {ECK120000299}   \n",
       "\n",
       "                                    genetic features   oriC  pseudogene TFBS  \\\n",
       "0  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  False       False   {}   \n",
       "1  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  False       False   {}   \n",
       "2  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  False       False   {}   \n",
       "3  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  False       False   {}   \n",
       "4  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  False       False   {}   \n",
       "\n",
       "  promoter RBS terminator  genetic  \\\n",
       "0       {}  {}         {}     True   \n",
       "1       {}  {}         {}     True   \n",
       "2       {}  {}         {}     True   \n",
       "3       {}  {}         {}     True   \n",
       "4       {}  {}         {}     True   \n",
       "\n",
       "                                    genomic features  \n",
       "0  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  \n",
       "1  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  \n",
       "2  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  \n",
       "3  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  \n",
       "4  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./data/2_2_df.pkl\")\n",
    "display(len(all_muts_df), all_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G203G (GGC→GGT)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756515.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GLU</td>\n",
       "      <td>412.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>1435</td>\n",
       "      <td>2.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756515, 756515)</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G103C (GGC→TGC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756213.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>164.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>812</td>\n",
       "      <td>8.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756213, 756213)</td>\n",
       "      <td>3 164 8 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D377D (GAT→GAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>757037.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TOL_hexamethylenediamine</td>\n",
       "      <td>50.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>1051</td>\n",
       "      <td>3.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(757037, 757037)</td>\n",
       "      <td>4 50 3 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>W65R (TGG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756099.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>840</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756099, 756099)</td>\n",
       "      <td>5 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q82* (CAG→TAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756150.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>44.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>852</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756150, 756150)</td>\n",
       "      <td>6 44 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>L327R (CTG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756886.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→G</td>\n",
       "      <td>{}</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>56.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>893</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756886, 756886)</td>\n",
       "      <td>7 56 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H354Q (CAC→CAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756968.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→G</td>\n",
       "      <td>{}</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756968, 756968)</td>\n",
       "      <td>8 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type  Position RBS Reference Seq Sequence Change  \\\n",
       "0  G203G (GGC→GGT)           SNP  756515.0  {}     NC_000913             C→T   \n",
       "1  G103C (GGC→TGC)           SNP  756213.0  {}           NaN             G→T   \n",
       "2  D377D (GAT→GAC)           SNP  757037.0  {}     NC_000913             T→C   \n",
       "3   W65R (TGG→CGG)           SNP  756099.0  {}     NC_000913             T→C   \n",
       "4   Q82* (CAG→TAG)           SNP  756150.0  {}     NC_000913             C→T   \n",
       "5  L327R (CTG→CGG)           SNP  756886.0  {}     NC_000913             T→G   \n",
       "6  H354Q (CAC→CAG)           SNP  756968.0  {}     NC_000913             C→G   \n",
       "\n",
       "  TFBS  ale  coding                       exp  flask gene RegulonDB ID  \\\n",
       "0   {}  3.0     1.0                       GLU  412.0    {ECK120000920}   \n",
       "1   {}  3.0     1.0               SSW_GLU_XYL  164.0    {ECK120000920}   \n",
       "2   {}  4.0     1.0  TOL_hexamethylenediamine   50.0    {ECK120000920}   \n",
       "3   {}  5.0     1.0                  ndh-cyoB   59.0    {ECK120000920}   \n",
       "4   {}  6.0     1.0                  ndh-cyoB   44.0    {ECK120000920}   \n",
       "5   {}  7.0     1.0                  ndh-cyoB   56.0    {ECK120000920}   \n",
       "6   {}  8.0     1.0                  ndh-cyoB   59.0    {ECK120000920}   \n",
       "\n",
       "   genetic                                   genetic features  \\\n",
       "0      1.0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "1      1.0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "2      1.0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "3      1.0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "4      1.0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "5      1.0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "6      1.0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...  1435      2.0   \n",
       "1  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   812      8.0   \n",
       "2  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...  1051      3.0   \n",
       "3  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   840      1.0   \n",
       "4  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   852      0.0   \n",
       "5  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   893      1.0   \n",
       "6  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   918      1.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "0                       sdhA   0.0       1.0       {}         0.0   \n",
       "1                       sdhA   0.0       1.0       {}         0.0   \n",
       "2                       sdhA   0.0       1.0       {}         0.0   \n",
       "3                       sdhA   0.0       1.0       {}         0.0   \n",
       "4                       sdhA   0.0       1.0       {}         0.0   \n",
       "5                       sdhA   0.0       1.0       {}         0.0   \n",
       "6                       sdhA   0.0       1.0       {}         0.0   \n",
       "\n",
       "              range     sample  tech_rep terminator  \n",
       "0  (756515, 756515)  3 412 2 1       1.0         {}  \n",
       "1  (756213, 756213)  3 164 8 1       1.0         {}  \n",
       "2  (757037, 757037)   4 50 3 1       1.0         {}  \n",
       "3  (756099, 756099)   5 59 1 1       1.0         {}  \n",
       "4  (756150, 756150)   6 44 0 1       1.0         {}  \n",
       "5  (756886, 756886)   7 56 1 1       1.0         {}  \n",
       "6  (756968, 756968)   8 59 1 1       1.0         {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_muts = pd.DataFrame()\n",
    "for _, r in all_muts_df.iterrows():  # Using genomic features ensures I get any large deletions\n",
    "    for f in r[\"genomic features\"]:\n",
    "        if f[\"name\"]==\"sdhA\":\n",
    "            meta_muts = meta_muts.append(r, ignore_index=True)\n",
    "            break\n",
    "display(len(meta_muts), meta_muts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'IS186': 1345,\n",
       " 'IS1': 768,\n",
       " 'IS911': 335,\n",
       " 'IS30': 1221,\n",
       " 'IS5': 1195,\n",
       " 'IS': 312,\n",
       " 'IS3': 1255,\n",
       " 'IS2': 1331,\n",
       " 'IS609': 1748,\n",
       " 'IS5L': 1195,\n",
       " 'IS150': 1443,\n",
       " 'IS4': 1426,\n",
       " 'IS600': 327}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "recs = [rec for rec in SeqIO.parse(\"./data/NC_000913_3.gb\", \"genbank\")]\n",
    "\n",
    "mob_sizes = dict()\n",
    "\n",
    "for rec in recs:\n",
    "    feats = [feat for feat in rec.features if feat.type == \"mobile_element\"]\n",
    "    for feat in feats:\n",
    "        for annot in feat.qualifiers[\"mobile_element_type\"]:\n",
    "            mob_name = annot.replace(\"insertion sequence:\",'')\n",
    "            if mob_name[-1].isdigit() == False:  # Removing annotations describing separate instances of a MOB.\n",
    "                mob_name = mob_name[:-1]\n",
    "            mob_size = len(feat.location)\n",
    "            if mob_name not in mob_sizes.keys():  # Assuming that all MOBs have the same length.\n",
    "                mob_sizes[mob_name] = mob_size\n",
    "mob_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from mutil.mut import get_mut_size, get_mob_size, is_genetic_mut, is_coding_mut, predict_mutation_effect_on_feature, get_SNP_aa_pos, get_DEL_INS_MOB_aa_start_pos, get_DEL_AA_range\n",
    "\n",
    "\n",
    "meta_muts[\"genetic\"] = meta_muts[\"Details\"].apply(is_genetic_mut)\n",
    "meta_muts[\"coding\"] = meta_muts[\"Details\"].apply(is_coding_mut)\n",
    "meta_muts[\"mutation size\"] = meta_muts.apply(get_mut_size, axis=1)  # get_mut_size currently doesn't return MOB mut sizes. Isn't something currently necessary.\n",
    "\n",
    "# TODO: integrate MOB size logic into get_mut_size\n",
    "meta_muts[\"mutation size\"] = meta_muts.apply(lambda r: get_mob_size(r[\"Sequence Change\"], mob_sizes) if r[\"Mutation Type\"] == \"MOB\" else r[\"mutation size\"], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    None\n",
       "1    None\n",
       "2    None\n",
       "3    None\n",
       "4    None\n",
       "5    None\n",
       "6    None\n",
       "dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _set_geno_feat_effect(mut):\n",
    "    for gf in mut[\"genomic features\"]:\n",
    "        gf[\"predicted mutation effect\"] = predict_mutation_effect_on_feature(mut, gf)\n",
    "\n",
    "meta_muts.apply(lambda m: _set_geno_feat_effect(m), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _is_gene_trunc(geno_feats, gene_name):\n",
    "    is_gene_trunc = False\n",
    "    for gf in geno_feats:\n",
    "        if (gf[\"name\"] == gene_name) & (gf['predicted mutation effect'] == 'truncation'):\n",
    "            \n",
    "            is_gene_trunc = True\n",
    "            break\n",
    "    return is_gene_trunc\n",
    "\n",
    "\n",
    "meta_muts[\"truncation\"] = meta_muts[\"genomic features\"].apply(lambda gfs: _is_gene_trunc(gfs, \"sdhA\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G203G (GGC→GGT)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756515.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>GLU</td>\n",
       "      <td>412.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>1435</td>\n",
       "      <td>2.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756515, 756515)</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(203, 203)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G103C (GGC→TGC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756213.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>164.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>812</td>\n",
       "      <td>8.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756213, 756213)</td>\n",
       "      <td>3 164 8 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(103, 103)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D377D (GAT→GAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>757037.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>4.0</td>\n",
       "      <td>True</td>\n",
       "      <td>TOL_hexamethylenediamine</td>\n",
       "      <td>50.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>1051</td>\n",
       "      <td>3.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(757037, 757037)</td>\n",
       "      <td>4 50 3 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(377, 377)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>W65R (TGG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756099.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>840</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756099, 756099)</td>\n",
       "      <td>5 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(65, 65)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q82* (CAG→TAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756150.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>44.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>852</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756150, 756150)</td>\n",
       "      <td>6 44 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(82, 82)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type  Position RBS Reference Seq Sequence Change  \\\n",
       "0  G203G (GGC→GGT)           SNP  756515.0  {}     NC_000913             C→T   \n",
       "1  G103C (GGC→TGC)           SNP  756213.0  {}           NaN             G→T   \n",
       "2  D377D (GAT→GAC)           SNP  757037.0  {}     NC_000913             T→C   \n",
       "3   W65R (TGG→CGG)           SNP  756099.0  {}     NC_000913             T→C   \n",
       "4   Q82* (CAG→TAG)           SNP  756150.0  {}     NC_000913             C→T   \n",
       "\n",
       "  TFBS  ale  coding                       exp  flask gene RegulonDB ID  \\\n",
       "0   {}  3.0    True                       GLU  412.0    {ECK120000920}   \n",
       "1   {}  3.0    True               SSW_GLU_XYL  164.0    {ECK120000920}   \n",
       "2   {}  4.0    True  TOL_hexamethylenediamine   50.0    {ECK120000920}   \n",
       "3   {}  5.0    True                  ndh-cyoB   59.0    {ECK120000920}   \n",
       "4   {}  6.0    True                  ndh-cyoB   44.0    {ECK120000920}   \n",
       "\n",
       "   genetic                                   genetic features  \\\n",
       "0     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "1     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "2     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "3     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "4     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...  1435      2.0   \n",
       "1  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   812      8.0   \n",
       "2  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...  1051      3.0   \n",
       "3  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   840      1.0   \n",
       "4  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   852      0.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "0                       sdhA   0.0       1.0       {}         0.0   \n",
       "1                       sdhA   0.0       1.0       {}         0.0   \n",
       "2                       sdhA   0.0       1.0       {}         0.0   \n",
       "3                       sdhA   0.0       1.0       {}         0.0   \n",
       "4                       sdhA   0.0       1.0       {}         0.0   \n",
       "\n",
       "              range     sample  tech_rep terminator  mutation size  \\\n",
       "0  (756515, 756515)  3 412 2 1       1.0         {}              1   \n",
       "1  (756213, 756213)  3 164 8 1       1.0         {}              1   \n",
       "2  (757037, 757037)   4 50 3 1       1.0         {}              1   \n",
       "3  (756099, 756099)   5 59 1 1       1.0         {}              1   \n",
       "4  (756150, 756150)   6 44 0 1       1.0         {}              1   \n",
       "\n",
       "   truncation    AA range  \n",
       "0       False  (203, 203)  \n",
       "1       False  (103, 103)  \n",
       "2       False  (377, 377)  \n",
       "3       False    (65, 65)  \n",
       "4        True    (82, 82)  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_SNP_AA_range(mut_details_str):\n",
    "    SNP_AA_range = ()\n",
    "    AA_pos = get_SNP_aa_pos(mut_details_str)\n",
    "    SNP_AA_range = (AA_pos, AA_pos)\n",
    "    return SNP_AA_range\n",
    "\n",
    "meta_muts[\"AA range\"] = None\n",
    "meta_muts[\"AA range\"] = meta_muts.apply(lambda r: _get_SNP_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"SNP\") & (r[\"coding\"]==1)) else r[\"AA range\"], axis=1)\n",
    "meta_muts[meta_muts[\"Mutation Type\"].isin([\"SNP\"])].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Details, Mutation Type, Position, RBS, Reference Seq, Sequence Change, TFBS, ale, coding, exp, flask, gene RegulonDB ID, genetic, genetic features, genomic features, index, isolate, mutation target annotation, oriC, presence, promoter, pseudogene, range, sample, tech_rep, terminator, mutation size, truncation, AA range]\n",
       "Index: []"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_INS_MOB_AA_range(mut_details_str):\n",
    "    INS_MOB_AA_range = ()\n",
    "    AA_pos = get_DEL_INS_MOB_aa_start_pos(mut_details_str)\n",
    "    INS_MOB_AA_range = (AA_pos, AA_pos)\n",
    "    return INS_MOB_AA_range\n",
    "\n",
    "\n",
    "meta_muts[\"AA range\"] = meta_muts.apply(lambda r: _get_INS_MOB_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] in [\"INS\", \"MOB\"]) & (r[\"coding\"]==1)) else r[\"AA range\"], axis=1)\n",
    "meta_muts[meta_muts[\"Mutation Type\"].isin([\"INS\", \"MOB\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Chain</td>\n",
       "      <td>1</td>\n",
       "      <td>588</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PRO_0000158652;Note=Succinate dehydrogenase...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>37</td>\n",
       "      <td>52</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>404</td>\n",
       "      <td>405</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Active site</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Proton acceptor;Ontology_term=ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>221</td>\n",
       "      <td>221</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Substrate binding site</td>\n",
       "      <td>242</td>\n",
       "      <td>242</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Substrate;Ontology_term=ECO:0000269;evide...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Substrate binding site</td>\n",
       "      <td>254</td>\n",
       "      <td>254</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Substrate;Ontology_term=ECO:0000269;evide...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Substrate binding site</td>\n",
       "      <td>354</td>\n",
       "      <td>354</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Substrate;Ontology_term=ECO:0000269;evide...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>388</td>\n",
       "      <td>388</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Substrate binding site</td>\n",
       "      <td>399</td>\n",
       "      <td>399</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Substrate;Ontology_term=ECO:0000269;evide...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Modified residue</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Tele-8alpha-FAD histidine;Ontology_term=E...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Modified residue</td>\n",
       "      <td>267</td>\n",
       "      <td>267</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=N6-acetyllysine;Ontology_term=ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>33</td>\n",
       "      <td>39</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>41</td>\n",
       "      <td>43</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>45</td>\n",
       "      <td>48</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>59</td>\n",
       "      <td>61</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>65</td>\n",
       "      <td>75</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Turn</td>\n",
       "      <td>76</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>82</td>\n",
       "      <td>101</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>112</td>\n",
       "      <td>114</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>119</td>\n",
       "      <td>121</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Turn</td>\n",
       "      <td>125</td>\n",
       "      <td>127</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>129</td>\n",
       "      <td>135</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>141</td>\n",
       "      <td>155</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>159</td>\n",
       "      <td>162</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>164</td>\n",
       "      <td>171</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>177</td>\n",
       "      <td>184</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Turn</td>\n",
       "      <td>185</td>\n",
       "      <td>187</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>190</td>\n",
       "      <td>200</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>206</td>\n",
       "      <td>208</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>209</td>\n",
       "      <td>214</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>221</td>\n",
       "      <td>228</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>233</td>\n",
       "      <td>235</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>239</td>\n",
       "      <td>246</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Turn</td>\n",
       "      <td>247</td>\n",
       "      <td>249</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>256</td>\n",
       "      <td>259</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>263</td>\n",
       "      <td>265</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>273</td>\n",
       "      <td>276</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Turn</td>\n",
       "      <td>278</td>\n",
       "      <td>280</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>281</td>\n",
       "      <td>283</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>286</td>\n",
       "      <td>298</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>305</td>\n",
       "      <td>307</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>309</td>\n",
       "      <td>314</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>316</td>\n",
       "      <td>318</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>319</td>\n",
       "      <td>326</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>328</td>\n",
       "      <td>337</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Turn</td>\n",
       "      <td>342</td>\n",
       "      <td>344</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>347</td>\n",
       "      <td>356</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>359</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>367</td>\n",
       "      <td>371</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>373</td>\n",
       "      <td>375</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>377</td>\n",
       "      <td>385</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>387</td>\n",
       "      <td>389</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>393</td>\n",
       "      <td>395</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>403</td>\n",
       "      <td>417</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>419</td>\n",
       "      <td>426</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>434</td>\n",
       "      <td>439</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>442</td>\n",
       "      <td>449</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>452</td>\n",
       "      <td>454</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>456</td>\n",
       "      <td>470</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>471</td>\n",
       "      <td>475</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>477</td>\n",
       "      <td>494</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>504</td>\n",
       "      <td>506</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>508</td>\n",
       "      <td>532</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>542</td>\n",
       "      <td>545</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Helix</td>\n",
       "      <td>550</td>\n",
       "      <td>553</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>554</td>\n",
       "      <td>560</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Turn</td>\n",
       "      <td>561</td>\n",
       "      <td>564</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>565</td>\n",
       "      <td>569</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Beta strand</td>\n",
       "      <td>576</td>\n",
       "      <td>578</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Ontology_term=ECO:0000244;evidence=ECO:0000244...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAD binding domain</td>\n",
       "      <td>9</td>\n",
       "      <td>405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Catalytic domain</td>\n",
       "      <td>236</td>\n",
       "      <td>356</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAD-dependent pyridine nucleotide reductase</td>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FAD-dependent pyridine nucleotide reductase</td>\n",
       "      <td>368</td>\n",
       "      <td>390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname     source                                      feature  start  \\\n",
       "0   P0AC41  UniProtKB                                        Chain      1   \n",
       "1   P0AC41  UniProtKB                             FAD binding site     14   \n",
       "2   P0AC41  UniProtKB                             FAD binding site     37   \n",
       "3   P0AC41  UniProtKB                             FAD binding site    404   \n",
       "4   P0AC41  UniProtKB                                  Active site    286   \n",
       "5   P0AC41  UniProtKB                             FAD binding site    221   \n",
       "6   P0AC41  UniProtKB                       Substrate binding site    242   \n",
       "7   P0AC41  UniProtKB                       Substrate binding site    254   \n",
       "8   P0AC41  UniProtKB                       Substrate binding site    354   \n",
       "9   P0AC41  UniProtKB                             FAD binding site    388   \n",
       "10  P0AC41  UniProtKB                       Substrate binding site    399   \n",
       "11  P0AC41  UniProtKB                             Modified residue     45   \n",
       "12  P0AC41  UniProtKB                             Modified residue    267   \n",
       "13  P0AC41  UniProtKB                                  Beta strand      5   \n",
       "14  P0AC41  UniProtKB                                        Helix     17   \n",
       "15  P0AC41  UniProtKB                                  Beta strand     33   \n",
       "16  P0AC41  UniProtKB                                        Helix     41   \n",
       "17  P0AC41  UniProtKB                                        Helix     45   \n",
       "18  P0AC41  UniProtKB                                  Beta strand     59   \n",
       "19  P0AC41  UniProtKB                                        Helix     65   \n",
       "20  P0AC41  UniProtKB                                         Turn     76   \n",
       "21  P0AC41  UniProtKB                                        Helix     82   \n",
       "22  P0AC41  UniProtKB                                  Beta strand    112   \n",
       "23  P0AC41  UniProtKB                                  Beta strand    119   \n",
       "24  P0AC41  UniProtKB                                         Turn    125   \n",
       "25  P0AC41  UniProtKB                                  Beta strand    129   \n",
       "26  P0AC41  UniProtKB                                        Helix    141   \n",
       "27  P0AC41  UniProtKB                                  Beta strand    159   \n",
       "28  P0AC41  UniProtKB                                  Beta strand    164   \n",
       "29  P0AC41  UniProtKB                                  Beta strand    177   \n",
       "30  P0AC41  UniProtKB                                         Turn    185   \n",
       "31  P0AC41  UniProtKB                                  Beta strand    190   \n",
       "32  P0AC41  UniProtKB                                        Helix    206   \n",
       "33  P0AC41  UniProtKB                                  Beta strand    209   \n",
       "34  P0AC41  UniProtKB                                        Helix    221   \n",
       "35  P0AC41  UniProtKB                                  Beta strand    233   \n",
       "36  P0AC41  UniProtKB                                  Beta strand    239   \n",
       "37  P0AC41  UniProtKB                                         Turn    247   \n",
       "38  P0AC41  UniProtKB                                        Helix    256   \n",
       "39  P0AC41  UniProtKB                                  Beta strand    263   \n",
       "40  P0AC41  UniProtKB                                        Helix    273   \n",
       "41  P0AC41  UniProtKB                                         Turn    278   \n",
       "42  P0AC41  UniProtKB                                        Helix    281   \n",
       "43  P0AC41  UniProtKB                                        Helix    286   \n",
       "44  P0AC41  UniProtKB                                  Beta strand    305   \n",
       "45  P0AC41  UniProtKB                                  Beta strand    309   \n",
       "46  P0AC41  UniProtKB                                        Helix    316   \n",
       "47  P0AC41  UniProtKB                                        Helix    319   \n",
       "48  P0AC41  UniProtKB                                        Helix    328   \n",
       "49  P0AC41  UniProtKB                                         Turn    342   \n",
       "50  P0AC41  UniProtKB                                  Beta strand    347   \n",
       "51  P0AC41  UniProtKB                                  Beta strand    359   \n",
       "52  P0AC41  UniProtKB                                  Beta strand    367   \n",
       "53  P0AC41  UniProtKB                                  Beta strand    373   \n",
       "54  P0AC41  UniProtKB                                  Beta strand    377   \n",
       "55  P0AC41  UniProtKB                                        Helix    387   \n",
       "56  P0AC41  UniProtKB                                  Beta strand    393   \n",
       "57  P0AC41  UniProtKB                                        Helix    403   \n",
       "58  P0AC41  UniProtKB                                        Helix    419   \n",
       "59  P0AC41  UniProtKB                                        Helix    434   \n",
       "60  P0AC41  UniProtKB                                        Helix    442   \n",
       "61  P0AC41  UniProtKB                                  Beta strand    452   \n",
       "62  P0AC41  UniProtKB                                        Helix    456   \n",
       "63  P0AC41  UniProtKB                                  Beta strand    471   \n",
       "64  P0AC41  UniProtKB                                        Helix    477   \n",
       "65  P0AC41  UniProtKB                                  Beta strand    504   \n",
       "66  P0AC41  UniProtKB                                        Helix    508   \n",
       "67  P0AC41  UniProtKB                                  Beta strand    542   \n",
       "68  P0AC41  UniProtKB                                        Helix    550   \n",
       "69  P0AC41  UniProtKB                                  Beta strand    554   \n",
       "70  P0AC41  UniProtKB                                         Turn    561   \n",
       "71  P0AC41  UniProtKB                                  Beta strand    565   \n",
       "72  P0AC41  UniProtKB                                  Beta strand    576   \n",
       "73     NaN        NaN                           FAD binding domain      9   \n",
       "74     NaN        NaN                             Catalytic domain    236   \n",
       "75     NaN        NaN  FAD-dependent pyridine nucleotide reductase     10   \n",
       "76     NaN        NaN  FAD-dependent pyridine nucleotide reductase    368   \n",
       "\n",
       "    end score strand frame                                          attribute  \\\n",
       "0   588     .      .     .  ID=PRO_0000158652;Note=Succinate dehydrogenase...   \n",
       "1    19     .      .     .  Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...   \n",
       "2    52     .      .     .  Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...   \n",
       "3   405     .      .     .  Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...   \n",
       "4   286     .      .     .  Note=Proton acceptor;Ontology_term=ECO:0000269...   \n",
       "5   221     .      .     .  Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...   \n",
       "6   242     .      .     .  Note=Substrate;Ontology_term=ECO:0000269;evide...   \n",
       "7   254     .      .     .  Note=Substrate;Ontology_term=ECO:0000269;evide...   \n",
       "8   354     .      .     .  Note=Substrate;Ontology_term=ECO:0000269;evide...   \n",
       "9   388     .      .     .  Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...   \n",
       "10  399     .      .     .  Note=Substrate;Ontology_term=ECO:0000269;evide...   \n",
       "11   45     .      .     .  Note=Tele-8alpha-FAD histidine;Ontology_term=E...   \n",
       "12  267     .      .     .  Note=N6-acetyllysine;Ontology_term=ECO:0000269...   \n",
       "13   13     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "14   28     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "15   39     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "16   43     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "17   48     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "18   61     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "19   75     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "20   78     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "21  101     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "22  114     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "23  121     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "24  127     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "25  135     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "26  155     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "27  162     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "28  171     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "29  184     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "30  187     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "31  200     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "32  208     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "33  214     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "34  228     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "35  235     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "36  246     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "37  249     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "38  259     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "39  265     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "40  276     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "41  280     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "42  283     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "43  298     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "44  307     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "45  314     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "46  318     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "47  326     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "48  337     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "49  344     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "50  356     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "51  362     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "52  371     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "53  375     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "54  385     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "55  389     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "56  395     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "57  417     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "58  426     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "59  439     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "60  449     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "61  454     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "62  470     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "63  475     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "64  494     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "65  506     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "66  532     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "67  545     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "68  553     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "69  560     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "70  564     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "71  569     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "72  578     .      .     .  Ontology_term=ECO:0000244;evidence=ECO:0000244...   \n",
       "73  405   NaN    NaN   NaN                                                NaN   \n",
       "74  356   NaN    NaN   NaN                                                NaN   \n",
       "75   29   NaN    NaN   NaN                                                NaN   \n",
       "76  390   NaN    NaN   NaN                                                NaN   \n",
       "\n",
       "     9  \n",
       "0  NaN  \n",
       "1  NaN  \n",
       "2  NaN  \n",
       "3  NaN  \n",
       "4  NaN  \n",
       "5  NaN  \n",
       "6  NaN  \n",
       "7  NaN  \n",
       "8  NaN  \n",
       "9  NaN  \n",
       "10 NaN  \n",
       "11 NaN  \n",
       "12 NaN  \n",
       "13 NaN  \n",
       "14 NaN  \n",
       "15 NaN  \n",
       "16 NaN  \n",
       "17 NaN  \n",
       "18 NaN  \n",
       "19 NaN  \n",
       "20 NaN  \n",
       "21 NaN  \n",
       "22 NaN  \n",
       "23 NaN  \n",
       "24 NaN  \n",
       "25 NaN  \n",
       "26 NaN  \n",
       "27 NaN  \n",
       "28 NaN  \n",
       "29 NaN  \n",
       "30 NaN  \n",
       "31 NaN  \n",
       "32 NaN  \n",
       "33 NaN  \n",
       "34 NaN  \n",
       "35 NaN  \n",
       "36 NaN  \n",
       "37 NaN  \n",
       "38 NaN  \n",
       "39 NaN  \n",
       "40 NaN  \n",
       "41 NaN  \n",
       "42 NaN  \n",
       "43 NaN  \n",
       "44 NaN  \n",
       "45 NaN  \n",
       "46 NaN  \n",
       "47 NaN  \n",
       "48 NaN  \n",
       "49 NaN  \n",
       "50 NaN  \n",
       "51 NaN  \n",
       "52 NaN  \n",
       "53 NaN  \n",
       "54 NaN  \n",
       "55 NaN  \n",
       "56 NaN  \n",
       "57 NaN  \n",
       "58 NaN  \n",
       "59 NaN  \n",
       "60 NaN  \n",
       "61 NaN  \n",
       "62 NaN  \n",
       "63 NaN  \n",
       "64 NaN  \n",
       "65 NaN  \n",
       "66 NaN  \n",
       "67 NaN  \n",
       "68 NaN  \n",
       "69 NaN  \n",
       "70 NaN  \n",
       "71 NaN  \n",
       "72 NaN  \n",
       "73 NaN  \n",
       "74 NaN  \n",
       "75 NaN  \n",
       "76 NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from mutil.gff import get_df_from_csv\n",
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "    \n",
    "gff_df = get_df_from_csv(\"./data/gff/sdhA.gff\")\n",
    "\n",
    "def update_gff_feature_description(feature, attribute):\n",
    "    updated_feature = feature\n",
    "    if \"Note=Substrate\" in attribute:\n",
    "        updated_feature = \"Substrate binding site\"\n",
    "    if \"Note=FAD\" in attribute:\n",
    "        updated_feature = \"FAD binding site\"\n",
    "    return updated_feature\n",
    "\n",
    "\n",
    "gff_df = gff_df[~(gff_df[\"feature\"].isin([\"Sequence conflict\", \"Mutagenesis\", \"Initiator methionine\"]))]  # Don't care about these\n",
    "gff_df[\"feature\"] = gff_df.apply(lambda r: update_gff_feature_description(r[\"feature\"], r[\"attribute\"]), axis=1)\n",
    "\n",
    "# Getting the below from https://www.ebi.ac.uk/interpro/protein/UniProt/P0AC41/\n",
    "\n",
    "gff_df = gff_df.append({\"feature\": \"FAD binding domain\", \"start\": 9, \"end\": 405}, ignore_index=True)\n",
    "gff_df = gff_df.append({\"feature\": \"Catalytic domain\", \"start\": 236, \"end\": 356}, ignore_index=True)\n",
    "gff_df = gff_df.append({\"feature\": \"FAD-dependent pyridine nucleotide reductase\", \"start\": 10, \"end\": 29}, ignore_index=True)\n",
    "gff_df = gff_df.append({\"feature\": \"FAD-dependent pyridine nucleotide reductase\", \"start\": 368, \"end\": 390}, ignore_index=True)\n",
    "\n",
    "gff_df.sort_values(by=\"start\")\n",
    "gff_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>prot_a</th>\n",
       "      <th>prot_b</th>\n",
       "      <th>evidence</th>\n",
       "      <th>chain</th>\n",
       "      <th>aa_wt</th>\n",
       "      <th>aa_mt</th>\n",
       "      <th>pdb_pos</th>\n",
       "      <th>uniprot_pos</th>\n",
       "      <th>mut</th>\n",
       "      <th>dG_wt</th>\n",
       "      <th>dG_mt</th>\n",
       "      <th>ddG</th>\n",
       "      <th>dG_wt_sd</th>\n",
       "      <th>dG_mt_sd</th>\n",
       "      <th>ddG_sd</th>\n",
       "      <th>NC_000913v3_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>A</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>DA32A</td>\n",
       "      <td>-31.58340</td>\n",
       "      <td>-30.63080</td>\n",
       "      <td>0.95260</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>C</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>DA32C</td>\n",
       "      <td>-31.63082</td>\n",
       "      <td>-31.27712</td>\n",
       "      <td>0.35370</td>\n",
       "      <td>0.022999</td>\n",
       "      <td>0.001285</td>\n",
       "      <td>0.022574</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>DA32D</td>\n",
       "      <td>-31.58852</td>\n",
       "      <td>-31.60058</td>\n",
       "      <td>-0.01206</td>\n",
       "      <td>0.011449</td>\n",
       "      <td>0.014490</td>\n",
       "      <td>0.003041</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>E</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>DA32E</td>\n",
       "      <td>-31.59360</td>\n",
       "      <td>-31.23336</td>\n",
       "      <td>0.36024</td>\n",
       "      <td>0.013214</td>\n",
       "      <td>0.013768</td>\n",
       "      <td>0.022002</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>F</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>DA32F</td>\n",
       "      <td>-31.59192</td>\n",
       "      <td>-31.13414</td>\n",
       "      <td>0.45778</td>\n",
       "      <td>0.015038</td>\n",
       "      <td>0.005938</td>\n",
       "      <td>0.019263</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2615</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>S</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>DB111S</td>\n",
       "      <td>-26.65122</td>\n",
       "      <td>-25.04938</td>\n",
       "      <td>1.60184</td>\n",
       "      <td>0.242473</td>\n",
       "      <td>0.259450</td>\n",
       "      <td>0.193316</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2616</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>DB111T</td>\n",
       "      <td>-26.59676</td>\n",
       "      <td>-24.10744</td>\n",
       "      <td>2.48932</td>\n",
       "      <td>0.078278</td>\n",
       "      <td>0.152139</td>\n",
       "      <td>0.213593</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2617</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>V</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>DB111V</td>\n",
       "      <td>-26.68806</td>\n",
       "      <td>-24.87832</td>\n",
       "      <td>1.80974</td>\n",
       "      <td>0.274525</td>\n",
       "      <td>0.422271</td>\n",
       "      <td>0.248554</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2618</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>W</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>DB111W</td>\n",
       "      <td>-26.55596</td>\n",
       "      <td>-25.18112</td>\n",
       "      <td>1.37484</td>\n",
       "      <td>0.035017</td>\n",
       "      <td>0.119287</td>\n",
       "      <td>0.095098</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619</th>\n",
       "      <td>P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb</td>\n",
       "      <td>P07014</td>\n",
       "      <td>P0AC41</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>D</td>\n",
       "      <td>Y</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>DB111Y</td>\n",
       "      <td>-26.73896</td>\n",
       "      <td>-25.28546</td>\n",
       "      <td>1.45350</td>\n",
       "      <td>0.340212</td>\n",
       "      <td>0.337151</td>\n",
       "      <td>0.105942</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2620 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          pdb  prot_a  prot_b evidence chain  \\\n",
       "0     P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     A   \n",
       "1     P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     A   \n",
       "2     P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     A   \n",
       "3     P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     A   \n",
       "4     P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     A   \n",
       "...                                       ...     ...     ...      ...   ...   \n",
       "2615  P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     B   \n",
       "2616  P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     B   \n",
       "2617  P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     B   \n",
       "2618  P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     B   \n",
       "2619  P07014-P0AC41-EXP-2wdq-pdb2-J-0-I-0.pdb  P07014  P0AC41      EXP     B   \n",
       "\n",
       "     aa_wt aa_mt  pdb_pos  uniprot_pos     mut     dG_wt     dG_mt      ddG  \\\n",
       "0        D     A       32           32   DA32A -31.58340 -30.63080  0.95260   \n",
       "1        D     C       32           32   DA32C -31.63082 -31.27712  0.35370   \n",
       "2        D     D       32           32   DA32D -31.58852 -31.60058 -0.01206   \n",
       "3        D     E       32           32   DA32E -31.59360 -31.23336  0.36024   \n",
       "4        D     F       32           32   DA32F -31.59192 -31.13414  0.45778   \n",
       "...    ...   ...      ...          ...     ...       ...       ...      ...   \n",
       "2615     D     S      111          111  DB111S -26.65122 -25.04938  1.60184   \n",
       "2616     D     T      111          111  DB111T -26.59676 -24.10744  2.48932   \n",
       "2617     D     V      111          111  DB111V -26.68806 -24.87832  1.80974   \n",
       "2618     D     W      111          111  DB111W -26.55596 -25.18112  1.37484   \n",
       "2619     D     Y      111          111  DB111Y -26.73896 -25.28546  1.45350   \n",
       "\n",
       "      dG_wt_sd  dG_mt_sd    ddG_sd  NC_000913v3_pos  \n",
       "0     0.000000  0.000000  0.000000               33  \n",
       "1     0.022999  0.001285  0.022574               33  \n",
       "2     0.011449  0.014490  0.003041               33  \n",
       "3     0.013214  0.013768  0.022002               33  \n",
       "4     0.015038  0.005938  0.019263               33  \n",
       "...        ...       ...       ...              ...  \n",
       "2615  0.242473  0.259450  0.193316              112  \n",
       "2616  0.078278  0.152139  0.213593              112  \n",
       "2617  0.274525  0.422271  0.248554              112  \n",
       "2618  0.035017  0.119287  0.095098              112  \n",
       "2619  0.340212  0.337151  0.105942              112  \n",
       "\n",
       "[2620 rows x 17 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: change \"all_glpK_interface_df\" to something else\n",
    "\n",
    "# getting interface regions from mutfunc's precalculated outputs\n",
    "all_glpK_interface_df = pd.read_csv('./data/mutfunc_full/sdhA_interfaces.tab', sep='\\t', header=None)\n",
    "all_glpK_interface_df.columns = [\"pdb\",\"prot_a\",\"prot_b\",\"evidence\",\"chain\",\"aa_wt\",\"aa_mt\",\"pdb_pos\",\"uniprot_pos\",\"mut\",\"dG_wt\",\"dG_mt\",\"ddG\",\"dG_wt_sd\",\"dG_mt_sd\", \"ddG_sd\"]\n",
    "all_glpK_interface_df[\"NC_000913v3_pos\"] = all_glpK_interface_df[\"uniprot_pos\"] + 1  # Adding 1 to account for initiator methionine (start codon) within NC_000913v3 AA ref\n",
    "all_glpK_interface_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['P0AC41'], dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = all_glpK_interface_df.copy()\n",
    "df = df.drop(columns=[\"aa_wt\", \"aa_mt\", \"mut\", \"dG_wt\", \"dG_mt\", \"ddG\", \"dG_wt_sd\", \"dG_mt_sd\", \"ddG_sd\"]).drop_duplicates()\n",
    "df.prot_b.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['P07014'], dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.prot_a.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mutil.gff import group_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[12, 13, 14],\n",
       " [33],\n",
       " [35],\n",
       " [37],\n",
       " [40, 41],\n",
       " [43, 44, 45, 46],\n",
       " [48, 49],\n",
       " [51, 52, 53, 54, 55, 56, 57, 58],\n",
       " [60, 61, 62, 63],\n",
       " [76, 77],\n",
       " [79],\n",
       " [94],\n",
       " [98],\n",
       " [101,\n",
       "  102,\n",
       "  103,\n",
       "  104,\n",
       "  105,\n",
       "  106,\n",
       "  107,\n",
       "  108,\n",
       "  109,\n",
       "  110,\n",
       "  111,\n",
       "  112,\n",
       "  113,\n",
       "  114,\n",
       "  115,\n",
       "  116],\n",
       " [120, 121, 122, 123, 124],\n",
       " [130],\n",
       " [132, 133, 134, 135, 136, 137, 138, 139],\n",
       " [141],\n",
       " [144, 145],\n",
       " [148, 149, 150, 151, 152, 153],\n",
       " [155, 156],\n",
       " [164],\n",
       " [181, 182],\n",
       " [187],\n",
       " [208],\n",
       " [213, 214, 215, 216, 217, 218, 219],\n",
       " [250, 251, 252, 253],\n",
       " [333, 334],\n",
       " [337, 338],\n",
       " [340],\n",
       " [456],\n",
       " [458],\n",
       " [461, 462],\n",
       " [501, 502],\n",
       " [504, 505, 506, 507, 508, 509, 510],\n",
       " [513],\n",
       " [516]]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: change across the NB \"GlpK_GlpK\" to \"SdhA_SdhB\"\n",
    "\n",
    "GlpK_GlpK_interface_df = all_glpK_interface_df[((all_glpK_interface_df[\"prot_a\"]==\"P07014\") & (all_glpK_interface_df[\"prot_b\"]==\"P0AC41\"))]\n",
    "GlpK_GlpK_interface_AAs = GlpK_GlpK_interface_df[\"NC_000913v3_pos\"].unique()\n",
    "\n",
    "# Outputting ranges for manual input into R\n",
    "GlpK_GlpK_interface_AAs.sort()\n",
    "GlpK_GlpK_interface_AA_ranges = group_sequences(GlpK_GlpK_interface_AAs)\n",
    "GlpK_GlpK_interface_AA_ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "for rng_l in GlpK_GlpK_interface_AA_ranges:\n",
    "    gff_df = gff_df.append({\"feature\": \"SdhB interface\", \"start\": rng_l[0], \"end\": rng_l[-1], \"color\": \"#4169E1\"}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "      <th>subgenic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G203G (GGC→GGT)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756515.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>GLU</td>\n",
       "      <td>412.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>1435</td>\n",
       "      <td>2.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756515, 756515)</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(203, 203)</td>\n",
       "      <td>{FAD binding domain}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G103C (GGC→TGC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756213.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>164.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>812</td>\n",
       "      <td>8.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756213, 756213)</td>\n",
       "      <td>3 164 8 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(103, 103)</td>\n",
       "      <td>{FAD binding domain, SdhB interface}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D377D (GAT→GAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>757037.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>4.0</td>\n",
       "      <td>True</td>\n",
       "      <td>TOL_hexamethylenediamine</td>\n",
       "      <td>50.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>1051</td>\n",
       "      <td>3.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(757037, 757037)</td>\n",
       "      <td>4 50 3 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(377, 377)</td>\n",
       "      <td>{Beta strand, FAD binding domain, FAD-dependen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>W65R (TGG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756099.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>840</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756099, 756099)</td>\n",
       "      <td>5 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(65, 65)</td>\n",
       "      <td>{FAD binding domain, Helix}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q82* (CAG→TAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756150.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>44.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>852</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756150, 756150)</td>\n",
       "      <td>6 44 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(82, 82)</td>\n",
       "      <td>{FAD binding domain, Helix}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type  Position RBS Reference Seq Sequence Change  \\\n",
       "0  G203G (GGC→GGT)           SNP  756515.0  {}     NC_000913             C→T   \n",
       "1  G103C (GGC→TGC)           SNP  756213.0  {}           NaN             G→T   \n",
       "2  D377D (GAT→GAC)           SNP  757037.0  {}     NC_000913             T→C   \n",
       "3   W65R (TGG→CGG)           SNP  756099.0  {}     NC_000913             T→C   \n",
       "4   Q82* (CAG→TAG)           SNP  756150.0  {}     NC_000913             C→T   \n",
       "\n",
       "  TFBS  ale  coding                       exp  flask gene RegulonDB ID  \\\n",
       "0   {}  3.0    True                       GLU  412.0    {ECK120000920}   \n",
       "1   {}  3.0    True               SSW_GLU_XYL  164.0    {ECK120000920}   \n",
       "2   {}  4.0    True  TOL_hexamethylenediamine   50.0    {ECK120000920}   \n",
       "3   {}  5.0    True                  ndh-cyoB   59.0    {ECK120000920}   \n",
       "4   {}  6.0    True                  ndh-cyoB   44.0    {ECK120000920}   \n",
       "\n",
       "   genetic                                   genetic features  \\\n",
       "0     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "1     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "2     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "3     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "4     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...  1435      2.0   \n",
       "1  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   812      8.0   \n",
       "2  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...  1051      3.0   \n",
       "3  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   840      1.0   \n",
       "4  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   852      0.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "0                       sdhA   0.0       1.0       {}         0.0   \n",
       "1                       sdhA   0.0       1.0       {}         0.0   \n",
       "2                       sdhA   0.0       1.0       {}         0.0   \n",
       "3                       sdhA   0.0       1.0       {}         0.0   \n",
       "4                       sdhA   0.0       1.0       {}         0.0   \n",
       "\n",
       "              range     sample  tech_rep terminator  mutation size  \\\n",
       "0  (756515, 756515)  3 412 2 1       1.0         {}              1   \n",
       "1  (756213, 756213)  3 164 8 1       1.0         {}              1   \n",
       "2  (757037, 757037)   4 50 3 1       1.0         {}              1   \n",
       "3  (756099, 756099)   5 59 1 1       1.0         {}              1   \n",
       "4  (756150, 756150)   6 44 0 1       1.0         {}              1   \n",
       "\n",
       "   truncation    AA range                                  subgenic features  \n",
       "0       False  (203, 203)                               {FAD binding domain}  \n",
       "1       False  (103, 103)               {FAD binding domain, SdhB interface}  \n",
       "2       False  (377, 377)  {Beta strand, FAD binding domain, FAD-dependen...  \n",
       "3       False    (65, 65)                        {FAD binding domain, Helix}  \n",
       "4        True    (82, 82)                        {FAD binding domain, Helix}  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_mut_subgenic_feats(mut_AA_pos_range, gff_df):\n",
    "    mut_subgenic_feats = set()\n",
    "    mut_AA_pos_set = set(range(mut_AA_pos_range[0], mut_AA_pos_range[1] + 1))  # +1 so that \"range\" function will count last position\n",
    "    for _, r in gff_df.iterrows():\n",
    "        if r[\"feature\"] != \"Chain\":\n",
    "            feat_AA_pos_set = set(range(r[\"start\"], r[\"end\"] + 1))  # +1 so that \"range\" function will count last position\n",
    "            if len(mut_AA_pos_set & feat_AA_pos_set) > 0:\n",
    "                mut_subgenic_feats.add(r[\"feature\"])\n",
    "    if len(mut_subgenic_feats) == 0:\n",
    "        mut_subgenic_feats.add(\"unannotated region\")\n",
    "    return mut_subgenic_feats\n",
    "\n",
    "\n",
    "meta_muts[\"subgenic features\"] = meta_muts.apply(lambda r: _get_mut_subgenic_feats(r[\"AA range\"], gff_df), axis=1)\n",
    "meta_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P0AC41 L327R\n",
      "P0AC41 W65R\n",
      "P0AC41 D377D\n",
      "P0AC41 H354Q\n",
      "P0AC41 G103C\n",
      "P0AC41 G203G\n"
     ]
    }
   ],
   "source": [
    "# TODO: put this in mutil since I end up using it in every gene-based mut-needle-plot\n",
    "# For mutfunc with methionine being considered as AA 1\n",
    "mut_set = set()  # to only get unique\n",
    "for _, r in meta_muts.iterrows():\n",
    "    if (r[\"Mutation Type\"] == \"SNP\") and ('*' not in r[\"Details\"]):\n",
    "        s = r[\"Details\"]\n",
    "        out_str = \"P0AC41 \" + s[:s.find(\" (\")]\n",
    "        mut_set.add(out_str)\n",
    "\n",
    "for s in mut_set:\n",
    "    print(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "      <th>subgenic features</th>\n",
       "      <th>AA change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G203G (GGC→GGT)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756515.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>GLU</td>\n",
       "      <td>412.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>1435</td>\n",
       "      <td>2.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756515, 756515)</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(203, 203)</td>\n",
       "      <td>{FAD binding domain}</td>\n",
       "      <td>G203G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G103C (GGC→TGC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756213.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>164.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>812</td>\n",
       "      <td>8.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756213, 756213)</td>\n",
       "      <td>3 164 8 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(103, 103)</td>\n",
       "      <td>{FAD binding domain, SdhB interface}</td>\n",
       "      <td>G103C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D377D (GAT→GAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>757037.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>4.0</td>\n",
       "      <td>True</td>\n",
       "      <td>TOL_hexamethylenediamine</td>\n",
       "      <td>50.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>1051</td>\n",
       "      <td>3.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(757037, 757037)</td>\n",
       "      <td>4 50 3 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(377, 377)</td>\n",
       "      <td>{Beta strand, FAD binding domain, FAD-dependen...</td>\n",
       "      <td>D377D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>W65R (TGG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756099.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>840</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756099, 756099)</td>\n",
       "      <td>5 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(65, 65)</td>\n",
       "      <td>{FAD binding domain, Helix}</td>\n",
       "      <td>W65R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q82* (CAG→TAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756150.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>44.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>852</td>\n",
       "      <td>0.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756150, 756150)</td>\n",
       "      <td>6 44 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(82, 82)</td>\n",
       "      <td>{FAD binding domain, Helix}</td>\n",
       "      <td>Q82*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type  Position RBS Reference Seq Sequence Change  \\\n",
       "0  G203G (GGC→GGT)           SNP  756515.0  {}     NC_000913             C→T   \n",
       "1  G103C (GGC→TGC)           SNP  756213.0  {}           NaN             G→T   \n",
       "2  D377D (GAT→GAC)           SNP  757037.0  {}     NC_000913             T→C   \n",
       "3   W65R (TGG→CGG)           SNP  756099.0  {}     NC_000913             T→C   \n",
       "4   Q82* (CAG→TAG)           SNP  756150.0  {}     NC_000913             C→T   \n",
       "\n",
       "  TFBS  ale  coding                       exp  flask gene RegulonDB ID  \\\n",
       "0   {}  3.0    True                       GLU  412.0    {ECK120000920}   \n",
       "1   {}  3.0    True               SSW_GLU_XYL  164.0    {ECK120000920}   \n",
       "2   {}  4.0    True  TOL_hexamethylenediamine   50.0    {ECK120000920}   \n",
       "3   {}  5.0    True                  ndh-cyoB   59.0    {ECK120000920}   \n",
       "4   {}  6.0    True                  ndh-cyoB   44.0    {ECK120000920}   \n",
       "\n",
       "   genetic                                   genetic features  \\\n",
       "0     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "1     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "2     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "3     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "4     True  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "0  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...  1435      2.0   \n",
       "1  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   812      8.0   \n",
       "2  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...  1051      3.0   \n",
       "3  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   840      1.0   \n",
       "4  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   852      0.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "0                       sdhA   0.0       1.0       {}         0.0   \n",
       "1                       sdhA   0.0       1.0       {}         0.0   \n",
       "2                       sdhA   0.0       1.0       {}         0.0   \n",
       "3                       sdhA   0.0       1.0       {}         0.0   \n",
       "4                       sdhA   0.0       1.0       {}         0.0   \n",
       "\n",
       "              range     sample  tech_rep terminator  mutation size  \\\n",
       "0  (756515, 756515)  3 412 2 1       1.0         {}              1   \n",
       "1  (756213, 756213)  3 164 8 1       1.0         {}              1   \n",
       "2  (757037, 757037)   4 50 3 1       1.0         {}              1   \n",
       "3  (756099, 756099)   5 59 1 1       1.0         {}              1   \n",
       "4  (756150, 756150)   6 44 0 1       1.0         {}              1   \n",
       "\n",
       "   truncation    AA range                                  subgenic features  \\\n",
       "0       False  (203, 203)                               {FAD binding domain}   \n",
       "1       False  (103, 103)               {FAD binding domain, SdhB interface}   \n",
       "2       False  (377, 377)  {Beta strand, FAD binding domain, FAD-dependen...   \n",
       "3       False    (65, 65)                        {FAD binding domain, Helix}   \n",
       "4        True    (82, 82)                        {FAD binding domain, Helix}   \n",
       "\n",
       "  AA change  \n",
       "0     G203G  \n",
       "1     G103C  \n",
       "2     D377D  \n",
       "3      W65R  \n",
       "4      Q82*  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_muts[\"AA change\"] = meta_muts.apply(lambda r: r[\"Details\"][:r[\"Details\"].find(\"(\")].strip() if r[\"Mutation Type\"]==\"SNP\" else \"\", axis=1)\n",
    "meta_muts[meta_muts[\"Mutation Type\"]==\"SNP\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "      <th>subgenic features</th>\n",
       "      <th>AA change</th>\n",
       "      <th>conservation (SIFT)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G103C (GGC→TGC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756213.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>164.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>812</td>\n",
       "      <td>8.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756213, 756213)</td>\n",
       "      <td>3 164 8 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(103, 103)</td>\n",
       "      <td>{FAD binding domain, SdhB interface}</td>\n",
       "      <td>G103C</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>W65R (TGG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756099.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>840</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756099, 756099)</td>\n",
       "      <td>5 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(65, 65)</td>\n",
       "      <td>{FAD binding domain, Helix}</td>\n",
       "      <td>W65R</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>L327R (CTG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756886.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→G</td>\n",
       "      <td>{}</td>\n",
       "      <td>7.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>56.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>893</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756886, 756886)</td>\n",
       "      <td>7 56 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(327, 327)</td>\n",
       "      <td>{FAD binding domain, Catalytic domain}</td>\n",
       "      <td>L327R</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H354Q (CAC→CAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756968.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→G</td>\n",
       "      <td>{}</td>\n",
       "      <td>8.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756968, 756968)</td>\n",
       "      <td>8 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(354, 354)</td>\n",
       "      <td>{Beta strand, FAD binding domain, Catalytic do...</td>\n",
       "      <td>H354Q</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type  Position RBS Reference Seq Sequence Change  \\\n",
       "1  G103C (GGC→TGC)           SNP  756213.0  {}           NaN             G→T   \n",
       "3   W65R (TGG→CGG)           SNP  756099.0  {}     NC_000913             T→C   \n",
       "5  L327R (CTG→CGG)           SNP  756886.0  {}     NC_000913             T→G   \n",
       "6  H354Q (CAC→CAG)           SNP  756968.0  {}     NC_000913             C→G   \n",
       "\n",
       "  TFBS  ale  coding          exp  flask gene RegulonDB ID  genetic  \\\n",
       "1   {}  3.0    True  SSW_GLU_XYL  164.0    {ECK120000920}     True   \n",
       "3   {}  5.0    True     ndh-cyoB   59.0    {ECK120000920}     True   \n",
       "5   {}  7.0    True     ndh-cyoB   56.0    {ECK120000920}     True   \n",
       "6   {}  8.0    True     ndh-cyoB   59.0    {ECK120000920}     True   \n",
       "\n",
       "                                    genetic features  \\\n",
       "1  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "3  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "5  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "6  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "1  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   812      8.0   \n",
       "3  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   840      1.0   \n",
       "5  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   893      1.0   \n",
       "6  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   918      1.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "1                       sdhA   0.0       1.0       {}         0.0   \n",
       "3                       sdhA   0.0       1.0       {}         0.0   \n",
       "5                       sdhA   0.0       1.0       {}         0.0   \n",
       "6                       sdhA   0.0       1.0       {}         0.0   \n",
       "\n",
       "              range     sample  tech_rep terminator  mutation size  \\\n",
       "1  (756213, 756213)  3 164 8 1       1.0         {}              1   \n",
       "3  (756099, 756099)   5 59 1 1       1.0         {}              1   \n",
       "5  (756886, 756886)   7 56 1 1       1.0         {}              1   \n",
       "6  (756968, 756968)   8 59 1 1       1.0         {}              1   \n",
       "\n",
       "   truncation    AA range                                  subgenic features  \\\n",
       "1       False  (103, 103)               {FAD binding domain, SdhB interface}   \n",
       "3       False    (65, 65)                        {FAD binding domain, Helix}   \n",
       "5       False  (327, 327)             {FAD binding domain, Catalytic domain}   \n",
       "6       False  (354, 354)  {Beta strand, FAD binding domain, Catalytic do...   \n",
       "\n",
       "  AA change  conservation (SIFT)  \n",
       "1     G103C                 True  \n",
       "3      W65R                 True  \n",
       "5     L327R                 True  \n",
       "6     H354Q                 True  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_conservation_df = pd.read_csv(\"./data/mutfunc/sdhA/conservation.tab\", sep='\\t', comment='#')\n",
    "mut_conservation_df[\"AA change\"] = mut_conservation_df.apply(lambda r: r[\"refaa\"] + str(r[\"posaa\"]) + r[\"altaa\"], axis=1)\n",
    "meta_muts[\"conservation (SIFT)\"] = meta_muts.apply(lambda r: True if r[\"AA change\"] in mut_conservation_df[\"AA change\"].values else False, axis=1)\n",
    "meta_muts[meta_muts[\"conservation (SIFT)\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "      <th>subgenic features</th>\n",
       "      <th>AA change</th>\n",
       "      <th>conservation (SIFT)</th>\n",
       "      <th>stability (ΔΔGpred)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G103C (GGC→TGC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756213.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>164.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>812</td>\n",
       "      <td>8.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756213, 756213)</td>\n",
       "      <td>3 164 8 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(103, 103)</td>\n",
       "      <td>{FAD binding domain, SdhB interface}</td>\n",
       "      <td>G103C</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>W65R (TGG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756099.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>840</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756099, 756099)</td>\n",
       "      <td>5 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(65, 65)</td>\n",
       "      <td>{FAD binding domain, Helix}</td>\n",
       "      <td>W65R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>L327R (CTG→CGG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756886.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→G</td>\n",
       "      <td>{}</td>\n",
       "      <td>7.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>56.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>893</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756886, 756886)</td>\n",
       "      <td>7 56 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(327, 327)</td>\n",
       "      <td>{FAD binding domain, Catalytic domain}</td>\n",
       "      <td>L327R</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H354Q (CAC→CAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>756968.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→G</td>\n",
       "      <td>{}</td>\n",
       "      <td>8.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>59.0</td>\n",
       "      <td>{ECK120000920}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>[{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...</td>\n",
       "      <td>918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sdhA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(756968, 756968)</td>\n",
       "      <td>8 59 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(354, 354)</td>\n",
       "      <td>{Beta strand, FAD binding domain, Catalytic do...</td>\n",
       "      <td>H354Q</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type  Position RBS Reference Seq Sequence Change  \\\n",
       "1  G103C (GGC→TGC)           SNP  756213.0  {}           NaN             G→T   \n",
       "3   W65R (TGG→CGG)           SNP  756099.0  {}     NC_000913             T→C   \n",
       "5  L327R (CTG→CGG)           SNP  756886.0  {}     NC_000913             T→G   \n",
       "6  H354Q (CAC→CAG)           SNP  756968.0  {}     NC_000913             C→G   \n",
       "\n",
       "  TFBS  ale  coding          exp  flask gene RegulonDB ID  genetic  \\\n",
       "1   {}  3.0    True  SSW_GLU_XYL  164.0    {ECK120000920}     True   \n",
       "3   {}  5.0    True     ndh-cyoB   59.0    {ECK120000920}     True   \n",
       "5   {}  7.0    True     ndh-cyoB   56.0    {ECK120000920}     True   \n",
       "6   {}  8.0    True     ndh-cyoB   59.0    {ECK120000920}     True   \n",
       "\n",
       "                                    genetic features  \\\n",
       "1  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "3  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "5  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "6  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "1  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   812      8.0   \n",
       "3  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   840      1.0   \n",
       "5  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   893      1.0   \n",
       "6  [{'name': 'sdhA', 'RegulonDB ID': 'ECK12000092...   918      1.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "1                       sdhA   0.0       1.0       {}         0.0   \n",
       "3                       sdhA   0.0       1.0       {}         0.0   \n",
       "5                       sdhA   0.0       1.0       {}         0.0   \n",
       "6                       sdhA   0.0       1.0       {}         0.0   \n",
       "\n",
       "              range     sample  tech_rep terminator  mutation size  \\\n",
       "1  (756213, 756213)  3 164 8 1       1.0         {}              1   \n",
       "3  (756099, 756099)   5 59 1 1       1.0         {}              1   \n",
       "5  (756886, 756886)   7 56 1 1       1.0         {}              1   \n",
       "6  (756968, 756968)   8 59 1 1       1.0         {}              1   \n",
       "\n",
       "   truncation    AA range                                  subgenic features  \\\n",
       "1       False  (103, 103)               {FAD binding domain, SdhB interface}   \n",
       "3       False    (65, 65)                        {FAD binding domain, Helix}   \n",
       "5       False  (327, 327)             {FAD binding domain, Catalytic domain}   \n",
       "6       False  (354, 354)  {Beta strand, FAD binding domain, Catalytic do...   \n",
       "\n",
       "  AA change  conservation (SIFT)  stability (ΔΔGpred)  \n",
       "1     G103C                 True                 True  \n",
       "3      W65R                 True                 True  \n",
       "5     L327R                 True                 True  \n",
       "6     H354Q                 True                 True  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parsing mutfunc output\n",
    "mut_stability_df = pd.read_csv(\"./data/mutfunc/sdhA/stability.tab\", sep='\\t', comment='#')\n",
    "mut_stability_df[\"AA change\"] = mut_stability_df.apply(lambda r: r[\"refaa\"] + str(r[\"posaa\"]) + r[\"altaa\"], axis=1)\n",
    "meta_muts[\"stability (ΔΔGpred)\"] = meta_muts.apply(lambda r: True if r[\"AA change\"] in mut_stability_df[\"AA change\"].values else False, axis=1)\n",
    "meta_muts[meta_muts[\"stability (ΔΔGpred)\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coding disruption</th>\n",
       "      <th>functional disruption</th>\n",
       "      <th>structural disruption</th>\n",
       "      <th>unknown effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Beta strand</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Substrate binding site</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAD-dependent pyridine nucleotide reductase</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Catalytic domain</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SdhB interface</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAD binding domain</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Helix</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             coding disruption  \\\n",
       "Beta strand                                                  0   \n",
       "Substrate binding site                                       0   \n",
       "FAD-dependent pyridine nucleotide reductase                  0   \n",
       "Catalytic domain                                             0   \n",
       "SdhB interface                                               0   \n",
       "FAD binding domain                                           1   \n",
       "Helix                                                        1   \n",
       "\n",
       "                                             functional disruption  \\\n",
       "Beta strand                                                      1   \n",
       "Substrate binding site                                           1   \n",
       "FAD-dependent pyridine nucleotide reductase                      0   \n",
       "Catalytic domain                                                 2   \n",
       "SdhB interface                                                   1   \n",
       "FAD binding domain                                               4   \n",
       "Helix                                                            1   \n",
       "\n",
       "                                             structural disruption  \\\n",
       "Beta strand                                                      1   \n",
       "Substrate binding site                                           1   \n",
       "FAD-dependent pyridine nucleotide reductase                      0   \n",
       "Catalytic domain                                                 2   \n",
       "SdhB interface                                                   1   \n",
       "FAD binding domain                                               4   \n",
       "Helix                                                            1   \n",
       "\n",
       "                                             unknown effect  \n",
       "Beta strand                                               1  \n",
       "Substrate binding site                                    0  \n",
       "FAD-dependent pyridine nucleotide reductase               1  \n",
       "Catalytic domain                                          0  \n",
       "SdhB interface                                            0  \n",
       "FAD binding domain                                        2  \n",
       "Helix                                                     0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: need to combine a few of the features\n",
    "\n",
    "feat_set = set(gff_df[\"feature\"])\n",
    "feat_set |= {\"unannotated region\"}\n",
    "feat_mut_cnt_df = pd.DataFrame(0,\n",
    "                               columns=[\"coding disruption\", \"functional disruption\", \"structural disruption\", \"unknown effect\"],\n",
    "                               index=feat_set)\n",
    "feat_mut_cnt_df\n",
    "\n",
    "for _, r in meta_muts.iterrows():\n",
    "    for f in r[\"subgenic features\"]:\n",
    "        effect_predicted = False\n",
    "        if r[\"truncation\"]:\n",
    "            feat_mut_cnt_df.at[f, \"coding disruption\"] += 1\n",
    "            effect_predicted = True\n",
    "        else:\n",
    "            if r[\"conservation (SIFT)\"]:\n",
    "                feat_mut_cnt_df.at[f, \"functional disruption\"] += 1\n",
    "                effect_predicted = True\n",
    "            if r[\"stability (ΔΔGpred)\"]:\n",
    "                feat_mut_cnt_df.at[f, \"structural disruption\"] += 1\n",
    "                effect_predicted = True\n",
    "        if not effect_predicted:\n",
    "            feat_mut_cnt_df.at[f, \"unknown effect\"] += 1\n",
    "# feat_mut_cnt_df = feat_mut_cnt_df[~(feat_mut_cnt_df.index.isin([\"Turn\", \"Helix\", \"Beta strand\"]))]\n",
    "feat_mut_cnt_df = feat_mut_cnt_df[(feat_mut_cnt_df.T != 0).any(axis=0)]  # remove rows that are all zeros\n",
    "feat_mut_cnt_df = feat_mut_cnt_df.loc[:, (feat_mut_cnt_df != 0).any(axis=0)]  # remove columns that are all zeros\n",
    "feat_mut_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GLU', 'SSW_GLU_XYL', 'TOL_hexamethylenediamine', 'ndh-cyoB'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_muts.exp.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA position</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>disrupt coding seq</th>\n",
       "      <th>disrupt func</th>\n",
       "      <th>disrupt struct</th>\n",
       "      <th>study</th>\n",
       "      <th>study color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>203</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GLU</td>\n",
       "      <td>#fdc086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>103</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>#FF9A00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>377</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TOL_hexamethylenediamine</td>\n",
       "      <td>#3875DB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>65</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>82</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>327</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>354</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  AA position mutation type disrupt coding seq disrupt func disrupt struct  \\\n",
       "0         203           SNP                  0            0              0   \n",
       "1         103           SNP                  0            1              1   \n",
       "2         377           SNP                  0            0              0   \n",
       "3          65           SNP                  0            1              1   \n",
       "4          82           SNP                  1            0              0   \n",
       "5         327           SNP                  0            1              1   \n",
       "6         354           SNP                  0            1              1   \n",
       "\n",
       "                      study study color  \n",
       "0                       GLU     #fdc086  \n",
       "1               SSW_GLU_XYL     #FF9A00  \n",
       "2  TOL_hexamethylenediamine     #3875DB  \n",
       "3                  ndh-cyoB     #beaed4  \n",
       "4                  ndh-cyoB     #beaed4  \n",
       "5                  ndh-cyoB     #beaed4  \n",
       "6                  ndh-cyoB     #beaed4  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_id_d = {\n",
    "#     \"ndh\": \"ndh\",\n",
    "    \"ndh-cyoB\": \"ndh-cyoB\",\n",
    "    'GLU': 'GLU',\n",
    "    \"SSW_GLU_XYL\": \"SSW_GLU_XYL\",\n",
    "    \"TOL_hexamethylenediamine\": \"TOL_hexamethylenediamine\"\n",
    "}\n",
    "\n",
    "study_color_d = {\n",
    "#     \"ndh\": \"#7fc97f\",\n",
    "    \"ndh-cyoB\": \"#beaed4\",\n",
    "    \"SSW_GLU_XYL\": \"#FF9A00\",\n",
    "    \"GLU\": \"#fdc086\",\n",
    "    \"TOL_hexamethylenediamine\": \"#3875DB\",\n",
    "}\n",
    "\n",
    "\n",
    "# This is done specifically to expand the count of rows that describe the full range of AA per mutations.\n",
    "AA_rng_exp_df = pd.DataFrame(columns=[\"AA position\", \"mutation type\", \"disrupt coding seq\", \"disrupt func\",\n",
    "                                      \"disrupt struct\"\n",
    "                                      ])\n",
    "for _, m in meta_muts.iterrows():\n",
    "    AA_positions = set(range(m[\"AA range\"][0], m[\"AA range\"][1] + 1))\n",
    "\n",
    "    disrupt_coding = 0\n",
    "    if m[\"truncation\"]:\n",
    "        disrupt_coding = 1\n",
    "        AA_positions = {m[\"AA range\"][0]}\n",
    "\n",
    "    disrupt_func = 0\n",
    "    if m[\"conservation (SIFT)\"]:\n",
    "        disrupt_func = 1\n",
    "\n",
    "    disrupt_struct = 0\n",
    "    if \"stability (ΔΔGpred)\" in m.index and m[\"stability (ΔΔGpred)\"]:\n",
    "        disrupt_struct = 1\n",
    "\n",
    "    for p in AA_positions:\n",
    "        doi = exp_id_d[m[\"exp\"]]\n",
    "        AA_rng_exp_df = AA_rng_exp_df.append(\n",
    "            {\"AA position\": p,\n",
    "             \"mutation type\": m[\"Mutation Type\"],\n",
    "             \"disrupt coding seq\": disrupt_coding,\n",
    "             \"disrupt func\": disrupt_func,\n",
    "             \"disrupt struct\": disrupt_struct,\n",
    "             \"study\": doi,\n",
    "             \"study color\": study_color_d[doi]\n",
    "            },\n",
    "            ignore_index=True)\n",
    "AA_rng_exp_df[\"mutation type\"] = AA_rng_exp_df[\"mutation type\"].apply(lambda mt: \"CNV\" if mt == \"AMP\" else mt)\n",
    "AA_rng_exp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA position</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>disrupt coding seq</th>\n",
       "      <th>disrupt func</th>\n",
       "      <th>disrupt struct</th>\n",
       "      <th>study</th>\n",
       "      <th>study color</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>65</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>82</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>103</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>#FF9A00</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>203</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GLU</td>\n",
       "      <td>#fdc086</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>327</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>354</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>377</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TOL_hexamethylenediamine</td>\n",
       "      <td>#3875DB</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 377</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  AA position mutation type disrupt coding seq disrupt func disrupt struct  \\\n",
       "0          65           SNP                  0            1              1   \n",
       "1          82           SNP                  1            0              0   \n",
       "2         103           SNP                  0            1              1   \n",
       "3         203           SNP                  0            0              0   \n",
       "4         327           SNP                  0            1              1   \n",
       "5         354           SNP                  0            1              1   \n",
       "6         377           SNP                  0            0              0   \n",
       "\n",
       "                      study study color  mutation count        name  \n",
       "0                  ndh-cyoB     #beaed4               1   SNP AA 65  \n",
       "1                  ndh-cyoB     #beaed4               1   SNP AA 82  \n",
       "2               SSW_GLU_XYL     #FF9A00               1  SNP AA 103  \n",
       "3                       GLU     #fdc086               1  SNP AA 203  \n",
       "4                  ndh-cyoB     #beaed4               1  SNP AA 327  \n",
       "5                  ndh-cyoB     #beaed4               1  SNP AA 354  \n",
       "6  TOL_hexamethylenediamine     #3875DB               1  SNP AA 377  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Just want to count the mutation rows\n",
    "aa_mut_cnt_df = pd.DataFrame()\n",
    "for t, gdf in AA_rng_exp_df.groupby(list(AA_rng_exp_df.columns)):\n",
    "    c = len(gdf)\n",
    "    df = gdf.copy()\n",
    "    df = df.drop_duplicates()\n",
    "    df[\"mutation count\"] = c\n",
    "    aa_mut_cnt_df = aa_mut_cnt_df.append(df, ignore_index=True)\n",
    "aa_mut_cnt_df[\"name\"] = aa_mut_cnt_df.apply(lambda r: r[\"mutation type\"] + \" AA \" + str(r[\"AA position\"]), axis=1)\n",
    "aa_mut_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA position</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>disrupt coding seq</th>\n",
       "      <th>disrupt func</th>\n",
       "      <th>disrupt struct</th>\n",
       "      <th>study</th>\n",
       "      <th>study color</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>65</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 65</td>\n",
       "      <td>#A52A2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>82</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 82</td>\n",
       "      <td>red</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>103</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_GLU_XYL</td>\n",
       "      <td>#FF9A00</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 103</td>\n",
       "      <td>#A52A2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>203</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GLU</td>\n",
       "      <td>#fdc086</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 203</td>\n",
       "      <td>#ABABAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>327</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 327</td>\n",
       "      <td>#A52A2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>354</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cyoB</td>\n",
       "      <td>#beaed4</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 354</td>\n",
       "      <td>#A52A2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>377</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TOL_hexamethylenediamine</td>\n",
       "      <td>#3875DB</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 377</td>\n",
       "      <td>#ABABAB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  AA position mutation type disrupt coding seq disrupt func disrupt struct  \\\n",
       "0          65           SNP                  0            1              1   \n",
       "1          82           SNP                  1            0              0   \n",
       "2         103           SNP                  0            1              1   \n",
       "3         203           SNP                  0            0              0   \n",
       "4         327           SNP                  0            1              1   \n",
       "5         354           SNP                  0            1              1   \n",
       "6         377           SNP                  0            0              0   \n",
       "\n",
       "                      study study color  mutation count        name    color  \n",
       "0                  ndh-cyoB     #beaed4               1   SNP AA 65  #A52A2A  \n",
       "1                  ndh-cyoB     #beaed4               1   SNP AA 82      red  \n",
       "2               SSW_GLU_XYL     #FF9A00               1  SNP AA 103  #A52A2A  \n",
       "3                       GLU     #fdc086               1  SNP AA 203  #ABABAB  \n",
       "4                  ndh-cyoB     #beaed4               1  SNP AA 327  #A52A2A  \n",
       "5                  ndh-cyoB     #beaed4               1  SNP AA 354  #A52A2A  \n",
       "6  TOL_hexamethylenediamine     #3875DB               1  SNP AA 377  #ABABAB  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_mut_color(r):\n",
    "    mut_color = \"grey\"\n",
    "    if r[\"disrupt coding seq\"] == 1:  # red has priority over all other colors\n",
    "        mut_color = 'red'\n",
    "    else:\n",
    "        if (r[\"disrupt func\"] == 1) & (r[\"disrupt struct\"] == 1):\n",
    "            mut_color = \"#A52A2A\"\n",
    "        elif (r[\"disrupt func\"] == 1) & (r[\"disrupt struct\"] == 0):\n",
    "            mut_color = \"#FFA500\"\n",
    "        elif (r[\"disrupt func\"] == 0) & (r[\"disrupt struct\"] == 1):\n",
    "            mut_color = \"#A020F0\"\n",
    "        else:\n",
    "            mut_color = '#ABABAB'  # unknown\n",
    "    return mut_color\n",
    "\n",
    "\n",
    "aa_mut_cnt_df[\"color\"] = aa_mut_cnt_df.apply(lambda r: get_mut_color(r), axis=1)\n",
    "aa_mut_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_mut_cnt_df.to_csv(\"./sdhA_aa_muts.csv\", index=False)  # For R and other IPYNBs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ndh-cyoB', 'SSW_GLU_XYL', 'GLU', 'TOL_hexamethylenediamine'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_mut_cnt_df.study.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>9</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Chain</td>\n",
       "      <td>1</td>\n",
       "      <td>588</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PRO_0000158652;Note=Succinate dehydrogenase...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#00C200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>37</td>\n",
       "      <td>52</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#00C200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>404</td>\n",
       "      <td>405</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=FAD;Ontology_term=ECO:0000269,ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#00C200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P0AC41</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Active site</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Proton acceptor;Ontology_term=ECO:0000269...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#FFC125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     source           feature  start  end score strand frame  \\\n",
       "0  P0AC41  UniProtKB             Chain      1  588     .      .     .   \n",
       "1  P0AC41  UniProtKB  FAD binding site     14   19     .      .     .   \n",
       "2  P0AC41  UniProtKB  FAD binding site     37   52     .      .     .   \n",
       "3  P0AC41  UniProtKB  FAD binding site    404  405     .      .     .   \n",
       "4  P0AC41  UniProtKB       Active site    286  286     .      .     .   \n",
       "\n",
       "                                           attribute   9    color  \n",
       "0  ID=PRO_0000158652;Note=Succinate dehydrogenase... NaN    white  \n",
       "1  Note=FAD;Ontology_term=ECO:0000269,ECO:0000269... NaN  #00C200  \n",
       "2  Note=FAD;Ontology_term=ECO:0000269,ECO:0000269... NaN  #00C200  \n",
       "3  Note=FAD;Ontology_term=ECO:0000269,ECO:0000269... NaN  #00C200  \n",
       "4  Note=Proton acceptor;Ontology_term=ECO:0000269... NaN  #FFC125  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feat_to_color_d = {\n",
    "    \"Chain\": \"white\",  # don't need to find intermediate grey chains if the overlap is clean.\n",
    "    \"Helix\": \"#937860\",\n",
    "    \"Turn\": \"#f58231\",\n",
    "    \"Beta strand\": \"#8C8C8C\",\n",
    "    'Substrate binding site':'#911EB4',\n",
    "    'Active site':'#FFC125',\n",
    "    'FAD binding site':'#00C200',\n",
    "    'Modified residue':'#2AB0C1',\n",
    "    'SdhB interface': \"#4169E1\",\n",
    "    'FAD binding domain': \"#E4BCFF\",\n",
    "    'Catalytic domain': '#FFD8AE',\n",
    "    'FAD-dependent pyridine nucleotide reductase': \"#A5FFC1\"\n",
    "}\n",
    "\n",
    "\n",
    "gff_df[\"color\"] = gff_df.apply(lambda r: feat_to_color_d[r[\"feature\"]], axis=1)\n",
    "gff_df.to_csv(\"./sdhA_aa_feats.csv\", index=False)  # For R\n",
    "gff_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chain</td>\n",
       "      <td>0</td>\n",
       "      <td>587</td>\n",
       "      <td>white</td>\n",
       "      <td>588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FAD binding domain</td>\n",
       "      <td>8</td>\n",
       "      <td>404</td>\n",
       "      <td>#E4BCFF</td>\n",
       "      <td>397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Catalytic domain</td>\n",
       "      <td>235</td>\n",
       "      <td>355</td>\n",
       "      <td>#FFD8AE</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FAD-dependent pyridine nucleotide reductase</td>\n",
       "      <td>367</td>\n",
       "      <td>389</td>\n",
       "      <td>#A5FFC1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAD-dependent pyridine nucleotide reductase</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>#A5FFC1</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>36</td>\n",
       "      <td>51</td>\n",
       "      <td>#00C200</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>100</td>\n",
       "      <td>115</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>131</td>\n",
       "      <td>138</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>50</td>\n",
       "      <td>57</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>503</td>\n",
       "      <td>509</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>212</td>\n",
       "      <td>218</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>#00C200</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>147</td>\n",
       "      <td>152</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>119</td>\n",
       "      <td>123</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>59</td>\n",
       "      <td>62</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>42</td>\n",
       "      <td>45</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>249</td>\n",
       "      <td>252</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>332</td>\n",
       "      <td>333</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>154</td>\n",
       "      <td>155</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>403</td>\n",
       "      <td>404</td>\n",
       "      <td>#00C200</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>500</td>\n",
       "      <td>501</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>180</td>\n",
       "      <td>181</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>460</td>\n",
       "      <td>461</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>75</td>\n",
       "      <td>76</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>47</td>\n",
       "      <td>48</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>39</td>\n",
       "      <td>40</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>336</td>\n",
       "      <td>337</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>143</td>\n",
       "      <td>144</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>457</td>\n",
       "      <td>457</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>455</td>\n",
       "      <td>455</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>339</td>\n",
       "      <td>339</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>512</td>\n",
       "      <td>512</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>163</td>\n",
       "      <td>163</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>207</td>\n",
       "      <td>207</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>140</td>\n",
       "      <td>140</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Substrate binding site</td>\n",
       "      <td>398</td>\n",
       "      <td>398</td>\n",
       "      <td>#911EB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Active site</td>\n",
       "      <td>285</td>\n",
       "      <td>285</td>\n",
       "      <td>#FFC125</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>220</td>\n",
       "      <td>220</td>\n",
       "      <td>#00C200</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Substrate binding site</td>\n",
       "      <td>241</td>\n",
       "      <td>241</td>\n",
       "      <td>#911EB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Substrate binding site</td>\n",
       "      <td>253</td>\n",
       "      <td>253</td>\n",
       "      <td>#911EB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Substrate binding site</td>\n",
       "      <td>353</td>\n",
       "      <td>353</td>\n",
       "      <td>#911EB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>FAD binding site</td>\n",
       "      <td>387</td>\n",
       "      <td>387</td>\n",
       "      <td>#00C200</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Modified residue</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>#2AB0C1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>129</td>\n",
       "      <td>129</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Modified residue</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "      <td>#2AB0C1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>SdhB interface</td>\n",
       "      <td>515</td>\n",
       "      <td>515</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        feature  start  end    color  size\n",
       "0                                         Chain      0  587    white   588\n",
       "1                            FAD binding domain      8  404  #E4BCFF   397\n",
       "2                              Catalytic domain    235  355  #FFD8AE   121\n",
       "3   FAD-dependent pyridine nucleotide reductase    367  389  #A5FFC1    23\n",
       "4   FAD-dependent pyridine nucleotide reductase      9   28  #A5FFC1    20\n",
       "5                              FAD binding site     36   51  #00C200    16\n",
       "6                                SdhB interface    100  115  #4169E1    16\n",
       "7                                SdhB interface    131  138  #4169E1     8\n",
       "8                                SdhB interface     50   57  #4169E1     8\n",
       "9                                SdhB interface    503  509  #4169E1     7\n",
       "10                               SdhB interface    212  218  #4169E1     7\n",
       "11                             FAD binding site     13   18  #00C200     6\n",
       "12                               SdhB interface    147  152  #4169E1     6\n",
       "13                               SdhB interface    119  123  #4169E1     5\n",
       "14                               SdhB interface     59   62  #4169E1     4\n",
       "15                               SdhB interface     42   45  #4169E1     4\n",
       "16                               SdhB interface    249  252  #4169E1     4\n",
       "17                               SdhB interface     11   13  #4169E1     3\n",
       "18                               SdhB interface    332  333  #4169E1     2\n",
       "19                               SdhB interface    154  155  #4169E1     2\n",
       "20                             FAD binding site    403  404  #00C200     2\n",
       "21                               SdhB interface    500  501  #4169E1     2\n",
       "22                               SdhB interface    180  181  #4169E1     2\n",
       "23                               SdhB interface    460  461  #4169E1     2\n",
       "24                               SdhB interface     75   76  #4169E1     2\n",
       "25                               SdhB interface     47   48  #4169E1     2\n",
       "26                               SdhB interface     39   40  #4169E1     2\n",
       "27                               SdhB interface    336  337  #4169E1     2\n",
       "28                               SdhB interface    143  144  #4169E1     2\n",
       "29                               SdhB interface    457  457  #4169E1     1\n",
       "30                               SdhB interface    455  455  #4169E1     1\n",
       "31                               SdhB interface    186  186  #4169E1     1\n",
       "32                               SdhB interface    339  339  #4169E1     1\n",
       "33                               SdhB interface    512  512  #4169E1     1\n",
       "34                               SdhB interface    163  163  #4169E1     1\n",
       "35                               SdhB interface    207  207  #4169E1     1\n",
       "36                               SdhB interface     78   78  #4169E1     1\n",
       "37                               SdhB interface    140  140  #4169E1     1\n",
       "38                       Substrate binding site    398  398  #911EB4     1\n",
       "39                                  Active site    285  285  #FFC125     1\n",
       "40                             FAD binding site    220  220  #00C200     1\n",
       "41                       Substrate binding site    241  241  #911EB4     1\n",
       "42                       Substrate binding site    253  253  #911EB4     1\n",
       "43                       Substrate binding site    353  353  #911EB4     1\n",
       "44                             FAD binding site    387  387  #00C200     1\n",
       "45                             Modified residue     44   44  #2AB0C1     1\n",
       "46                               SdhB interface    129  129  #4169E1     1\n",
       "47                             Modified residue    266  266  #2AB0C1     1\n",
       "48                               SdhB interface     32   32  #4169E1     1\n",
       "49                               SdhB interface     34   34  #4169E1     1\n",
       "50                               SdhB interface     36   36  #4169E1     1\n",
       "51                               SdhB interface     93   93  #4169E1     1\n",
       "52                               SdhB interface     97   97  #4169E1     1\n",
       "53                               SdhB interface    515  515  #4169E1     1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# output feature data for NGL\n",
    "ngl_feat_df = gff_df[~(gff_df.feature.isin([\"Turn\", \"Helix\", \"Beta strand\"]))].copy()\n",
    "ngl_feat_df[\"size\"] = ngl_feat_df.apply(lambda r: r[\"end\"] - r[\"start\"] + 1, axis=1)\n",
    "ngl_feat_df = ngl_feat_df[[\"feature\", \"start\", \"end\", \"color\", \"size\"]]\n",
    "ngl_feat_df = ngl_feat_df.sort_values(\"size\", ascending=False)\n",
    "ngl_feat_df = ngl_feat_df.reset_index(drop=True)  # json explicitly uses the indexes\n",
    "\n",
    "# Need to so that the positions defined by AAs now match structural residues\n",
    "ngl_feat_df[\"start\"] = ngl_feat_df[\"start\"] - 1\n",
    "ngl_feat_df[\"end\"] = ngl_feat_df[\"end\"] - 1\n",
    "display(ngl_feat_df)\n",
    "with open('SdhA_feats.json', 'w') as outfile:   # for NGL\n",
    "    outfile.write(ngl_feat_df.to_json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA position</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>name</th>\n",
       "      <th>color</th>\n",
       "      <th>residue pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 65</td>\n",
       "      <td>#A52A2A</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 82</td>\n",
       "      <td>red</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>103</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 103</td>\n",
       "      <td>#A52A2A</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>203</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 203</td>\n",
       "      <td>#ABABAB</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>327</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 327</td>\n",
       "      <td>#A52A2A</td>\n",
       "      <td>326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>354</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 354</td>\n",
       "      <td>#A52A2A</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>377</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 377</td>\n",
       "      <td>#ABABAB</td>\n",
       "      <td>376</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  AA position  mutation count mutation type        name    color residue pos\n",
       "0          65               1           SNP   SNP AA 65  #A52A2A          64\n",
       "1          82               1           SNP   SNP AA 82      red          81\n",
       "2         103               1           SNP  SNP AA 103  #A52A2A         102\n",
       "3         203               1           SNP  SNP AA 203  #ABABAB         202\n",
       "4         327               1           SNP  SNP AA 327  #A52A2A         326\n",
       "5         354               1           SNP  SNP AA 354  #A52A2A         353\n",
       "6         377               1           SNP  SNP AA 377  #ABABAB         376"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "out_mut_df = aa_mut_cnt_df[[\"AA position\", \"mutation count\", \"mutation type\", \"name\", \"color\"]].copy()\n",
    "out_mut_df[\"residue pos\"] = out_mut_df[\"AA position\"] - 1\n",
    "display(out_mut_df)\n",
    "with open('sdhA_aa_muts.json', 'w') as outfile:  # for NGL\n",
    "    outfile.write(out_mut_df.to_json())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
