{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('display.max_rows', 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "585"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1392</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13957</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>M599L (ATG→TTG)</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(13957, 13957)</td>\n",
       "      <td>{ECK120000235}</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1428</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28175</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>W295R (TGG→AGG)</td>\n",
       "      <td>rihC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(28175, 28175)</td>\n",
       "      <td>{ECK120001070}</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1393</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101342</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>T193I (ACC→ATC)</td>\n",
       "      <td>murC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(101342, 101342)</td>\n",
       "      <td>{ECK120000612}</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'murC', 'RegulonDB ID': 'ECK12000061...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1394</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145691</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A204V (GCC→GTC)</td>\n",
       "      <td>yadE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(145691, 145691)</td>\n",
       "      <td>{ECK120001687}</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1429</td>\n",
       "      <td>GLU</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171072</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→G</td>\n",
       "      <td>K166K (AAA→AAG)</td>\n",
       "      <td>fhuD</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>3 412 2 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(171072, 171072)</td>\n",
       "      <td>{ECK120000299}</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0  1392  GLU    3    412        2         1       1.0     13957           SNP   \n",
       "1  1428  GLU    3    412        2         1       1.0     28175           SNP   \n",
       "2  1393  GLU    3    412        2         1       1.0    101342           SNP   \n",
       "3  1394  GLU    3    412        2         1       1.0    145691           SNP   \n",
       "4  1429  GLU    3    412        2         1       1.0    171072           SNP   \n",
       "\n",
       "  Sequence Change          Details mutation target annotation Reference Seq  \\\n",
       "0             A→T  M599L (ATG→TTG)                       dnaK     NC_000913   \n",
       "1             T→A  W295R (TGG→AGG)                       rihC     NC_000913   \n",
       "2             C→T  T193I (ACC→ATC)                       murC     NC_000913   \n",
       "3             C→T  A204V (GCC→GTC)                       yadE     NC_000913   \n",
       "4             A→G  K166K (AAA→AAG)                       fhuD     NC_000913   \n",
       "\n",
       "      sample  coding             range gene RegulonDB ID  \\\n",
       "0  3 412 2 1    True    (13957, 13957)    {ECK120000235}   \n",
       "1  3 412 2 1    True    (28175, 28175)    {ECK120001070}   \n",
       "2  3 412 2 1    True  (101342, 101342)    {ECK120000612}   \n",
       "3  3 412 2 1    True  (145691, 145691)    {ECK120001687}   \n",
       "4  3 412 2 1    True  (171072, 171072)    {ECK120000299}   \n",
       "\n",
       "                                    genetic features   oriC  pseudogene TFBS  \\\n",
       "0  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  False       False   {}   \n",
       "1  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  False       False   {}   \n",
       "2  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  False       False   {}   \n",
       "3  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  False       False   {}   \n",
       "4  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  False       False   {}   \n",
       "\n",
       "  promoter RBS terminator  genetic  \\\n",
       "0       {}  {}         {}     True   \n",
       "1       {}  {}         {}     True   \n",
       "2       {}  {}         {}     True   \n",
       "3       {}  {}         {}     True   \n",
       "4       {}  {}         {}     True   \n",
       "\n",
       "                                    genomic features  \n",
       "0  [{'name': 'dnaK', 'RegulonDB ID': 'ECK12000023...  \n",
       "1  [{'name': 'rihC', 'RegulonDB ID': 'ECK12000107...  \n",
       "2  [{'name': 'murC', 'RegulonDB ID': 'ECK12000061...  \n",
       "3  [{'name': 'yadE', 'RegulonDB ID': 'ECK12000168...  \n",
       "4  [{'name': 'fhuD', 'RegulonDB ID': 'ECK12000029...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./data/2_2_df.pkl\")\n",
    "display(len(all_muts_df), all_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1M (ATG→ATA) †</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633743.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>30.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>362</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633743, 4633743)</td>\n",
       "      <td>9 30 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E38K (GAG→AAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633634.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>32.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633634, 4633634)</td>\n",
       "      <td>10 32 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>63.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>11 63 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>60.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>12 60 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type   Position RBS Reference Seq Sequence Change  \\\n",
       "0  M1M (ATG→ATA) †           SNP  4633743.0  {}     NC_000913             C→T   \n",
       "1   E38K (GAG→AAG)           SNP  4633634.0  {}     NC_000913             C→T   \n",
       "2   C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "3   C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "\n",
       "  TFBS   ale  coding            exp  flask gene RegulonDB ID  genetic  \\\n",
       "0   {}   9.0     1.0  ndh-cydB-appC   30.0    {ECK120002327}      1.0   \n",
       "1   {}  10.0     1.0  ndh-cydB-appC   32.0    {ECK120002327}      1.0   \n",
       "2   {}  11.0     1.0  ndh-cydB-appC   63.0    {ECK120002327}      1.0   \n",
       "3   {}  12.0     1.0  ndh-cydB-appC   60.0    {ECK120002327}      1.0   \n",
       "\n",
       "                                    genetic features  \\\n",
       "0  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "1  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "0  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   362      0.0   \n",
       "1  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...    22      0.0   \n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   190      0.0   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   320      1.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "0                       yjjX   0.0       1.0       {}         0.0   \n",
       "1                       yjjX   0.0       1.0       {}         0.0   \n",
       "2                       yjjX   0.0       1.0       {}         0.0   \n",
       "3                       yjjX   0.0       1.0       {}         0.0   \n",
       "\n",
       "                range     sample  tech_rep terminator  \n",
       "0  (4633743, 4633743)   9 30 0 1       1.0         {}  \n",
       "1  (4633634, 4633634)  10 32 0 1       1.0         {}  \n",
       "2  (4633657, 4633657)  11 63 0 1       1.0         {}  \n",
       "3  (4633657, 4633657)  12 60 1 1       1.0         {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_muts = pd.DataFrame()\n",
    "for _, r in all_muts_df.iterrows():  # Using genomic features ensures I get any large deletions\n",
    "    for f in r[\"genomic features\"]:\n",
    "        if f[\"name\"]==\"yjjX\":\n",
    "            meta_muts = meta_muts.append(r, ignore_index=True)\n",
    "            break\n",
    "display(len(meta_muts), meta_muts.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'IS186': 1345,\n",
       " 'IS1': 768,\n",
       " 'IS911': 335,\n",
       " 'IS30': 1221,\n",
       " 'IS5': 1195,\n",
       " 'IS': 312,\n",
       " 'IS3': 1255,\n",
       " 'IS2': 1331,\n",
       " 'IS609': 1748,\n",
       " 'IS5L': 1195,\n",
       " 'IS150': 1443,\n",
       " 'IS4': 1426,\n",
       " 'IS600': 327}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "recs = [rec for rec in SeqIO.parse(\"./data/NC_000913_3.gb\", \"genbank\")]\n",
    "\n",
    "mob_sizes = dict()\n",
    "\n",
    "for rec in recs:\n",
    "    feats = [feat for feat in rec.features if feat.type == \"mobile_element\"]\n",
    "    for feat in feats:\n",
    "        for annot in feat.qualifiers[\"mobile_element_type\"]:\n",
    "            mob_name = annot.replace(\"insertion sequence:\",'')\n",
    "            if mob_name[-1].isdigit() == False:  # Removing annotations describing separate instances of a MOB.\n",
    "                mob_name = mob_name[:-1]\n",
    "            mob_size = len(feat.location)\n",
    "            if mob_name not in mob_sizes.keys():  # Assuming that all MOBs have the same length.\n",
    "                mob_sizes[mob_name] = mob_size\n",
    "mob_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from mutil.mut import get_mut_size, get_mob_size, is_genetic_mut, is_coding_mut, predict_mutation_effect_on_feature, get_SNP_aa_pos, get_DEL_INS_MOB_aa_start_pos, get_DEL_AA_range\n",
    "\n",
    "\n",
    "meta_muts[\"genetic\"] = meta_muts[\"Details\"].apply(is_genetic_mut)\n",
    "meta_muts[\"coding\"] = meta_muts[\"Details\"].apply(is_coding_mut)\n",
    "meta_muts[\"mutation size\"] = meta_muts.apply(get_mut_size, axis=1)  # get_mut_size currently doesn't return MOB mut sizes. Isn't something currently necessary.\n",
    "\n",
    "# TODO: integrate MOB size logic into get_mut_size\n",
    "meta_muts[\"mutation size\"] = meta_muts.apply(lambda r: get_mob_size(r[\"Sequence Change\"], mob_sizes) if r[\"Mutation Type\"] == \"MOB\" else r[\"mutation size\"], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    None\n",
       "1    None\n",
       "2    None\n",
       "3    None\n",
       "dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _set_geno_feat_effect(mut):\n",
    "    for gf in mut[\"genomic features\"]:\n",
    "        gf[\"predicted mutation effect\"] = predict_mutation_effect_on_feature(mut, gf)\n",
    "\n",
    "meta_muts.apply(lambda m: _set_geno_feat_effect(m), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _is_gene_trunc(geno_feats, gene_name):\n",
    "    is_gene_trunc = False\n",
    "    for gf in geno_feats:\n",
    "        if (gf[\"name\"] == gene_name) & (gf['predicted mutation effect'] == 'truncation'):\n",
    "            \n",
    "            is_gene_trunc = True\n",
    "            break\n",
    "    return is_gene_trunc\n",
    "\n",
    "\n",
    "meta_muts[\"truncation\"] = meta_muts[\"genomic features\"].apply(lambda gfs: _is_gene_trunc(gfs, \"yjjX\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1M (ATG→ATA) †</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633743.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>9.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>30.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>362</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633743, 4633743)</td>\n",
       "      <td>9 30 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E38K (GAG→AAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633634.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>10.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>32.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633634, 4633634)</td>\n",
       "      <td>10 32 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(38, 38)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>11.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>63.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>11 63 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(30, 30)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>12.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>60.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>12 60 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(30, 30)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type   Position RBS Reference Seq Sequence Change  \\\n",
       "0  M1M (ATG→ATA) †           SNP  4633743.0  {}     NC_000913             C→T   \n",
       "1   E38K (GAG→AAG)           SNP  4633634.0  {}     NC_000913             C→T   \n",
       "2   C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "3   C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "\n",
       "  TFBS   ale  coding            exp  flask gene RegulonDB ID  genetic  \\\n",
       "0   {}   9.0    True  ndh-cydB-appC   30.0    {ECK120002327}     True   \n",
       "1   {}  10.0    True  ndh-cydB-appC   32.0    {ECK120002327}     True   \n",
       "2   {}  11.0    True  ndh-cydB-appC   63.0    {ECK120002327}     True   \n",
       "3   {}  12.0    True  ndh-cydB-appC   60.0    {ECK120002327}     True   \n",
       "\n",
       "                                    genetic features  \\\n",
       "0  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "1  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "0  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   362      0.0   \n",
       "1  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...    22      0.0   \n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   190      0.0   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   320      1.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "0                       yjjX   0.0       1.0       {}         0.0   \n",
       "1                       yjjX   0.0       1.0       {}         0.0   \n",
       "2                       yjjX   0.0       1.0       {}         0.0   \n",
       "3                       yjjX   0.0       1.0       {}         0.0   \n",
       "\n",
       "                range     sample  tech_rep terminator  mutation size  \\\n",
       "0  (4633743, 4633743)   9 30 0 1       1.0         {}              1   \n",
       "1  (4633634, 4633634)  10 32 0 1       1.0         {}              1   \n",
       "2  (4633657, 4633657)  11 63 0 1       1.0         {}              1   \n",
       "3  (4633657, 4633657)  12 60 1 1       1.0         {}              1   \n",
       "\n",
       "   truncation  AA range  \n",
       "0        True    (1, 1)  \n",
       "1       False  (38, 38)  \n",
       "2       False  (30, 30)  \n",
       "3       False  (30, 30)  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_SNP_AA_range(mut_details_str):\n",
    "    SNP_AA_range = ()\n",
    "    AA_pos = get_SNP_aa_pos(mut_details_str)\n",
    "    SNP_AA_range = (AA_pos, AA_pos)\n",
    "    return SNP_AA_range\n",
    "\n",
    "meta_muts[\"AA range\"] = None\n",
    "meta_muts[\"AA range\"] = meta_muts.apply(lambda r: _get_SNP_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"SNP\") & (r[\"coding\"]==1)) else r[\"AA range\"], axis=1)\n",
    "meta_muts[meta_muts[\"Mutation Type\"].isin([\"SNP\"])].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Chain</td>\n",
       "      <td>1</td>\n",
       "      <td>170</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PRO_0000156337;Note=Inosine/xanthosine trip...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Substrate binding region</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Substrate binding;Ontology_term=ECO:00002...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Substrate binding region</td>\n",
       "      <td>68</td>\n",
       "      <td>69</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Substrate binding;Ontology_term=ECO:00002...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Manganese or magnesium mental binding site</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Manganese or magnesium;Ontology_term=ECO:...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Manganese or magnesium mental binding site</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Manganese or magnesium;Ontology_term=ECO:...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     source                                     feature  start  end  \\\n",
       "0  P39411  UniProtKB                                       Chain      1  170   \n",
       "1  P39411  UniProtKB                    Substrate binding region      8   13   \n",
       "2  P39411  UniProtKB                    Substrate binding region     68   69   \n",
       "3  P39411  UniProtKB  Manganese or magnesium mental binding site     38   38   \n",
       "4  P39411  UniProtKB  Manganese or magnesium mental binding site     68   68   \n",
       "\n",
       "  score strand frame                                          attribute   9  \n",
       "0     .      .     .  ID=PRO_0000156337;Note=Inosine/xanthosine trip... NaN  \n",
       "1     .      .     .  Note=Substrate binding;Ontology_term=ECO:00002... NaN  \n",
       "2     .      .     .  Note=Substrate binding;Ontology_term=ECO:00002... NaN  \n",
       "3     .      .     .  Note=Manganese or magnesium;Ontology_term=ECO:... NaN  \n",
       "4     .      .     .  Note=Manganese or magnesium;Ontology_term=ECO:... NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from mutil.gff import get_df_from_csv\n",
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "    \n",
    "gff_df = get_df_from_csv(\"./data/gff/yjjX.gff\")\n",
    "\n",
    "def update_gff_feature_description(feature, attribute):\n",
    "    updated_feature = feature\n",
    "    if \"Note=Substrate binding\" in attribute:\n",
    "        updated_feature = \"Substrate binding region\"\n",
    "    if \"Note=Manganese or magnesium\" in attribute:\n",
    "        updated_feature = \"Manganese or magnesium mental binding site\"\n",
    "    return updated_feature\n",
    "\n",
    "\n",
    "gff_df = gff_df[~(gff_df[\"feature\"].isin([\"Sequence conflict\", \"Mutagenesis\", \"Initiator methionine\", \"Helix\", \"Turn\", \"Beta strand\"]))]  # Don't care about these\n",
    "gff_df[\"feature\"] = gff_df.apply(lambda r: update_gff_feature_description(r[\"feature\"], r[\"attribute\"]), axis=1)\n",
    "\n",
    "# Additional annotations from Pfam (on EcoCyc)\n",
    "# gff_df = gff_df.append({\"feature\": \"NTPase_I-T Pfam domain\", \"start\": 7, \"end\": 166}, ignore_index=True)\n",
    "\n",
    "gff_df.sort_values(by=\"start\")\n",
    "gff_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>prot_a</th>\n",
       "      <th>prot_b</th>\n",
       "      <th>evidence</th>\n",
       "      <th>chain</th>\n",
       "      <th>aa_wt</th>\n",
       "      <th>aa_mt</th>\n",
       "      <th>pdb_pos</th>\n",
       "      <th>uniprot_pos</th>\n",
       "      <th>mut</th>\n",
       "      <th>dG_wt</th>\n",
       "      <th>dG_mt</th>\n",
       "      <th>ddG</th>\n",
       "      <th>dG_wt_sd</th>\n",
       "      <th>dG_mt_sd</th>\n",
       "      <th>ddG_sd</th>\n",
       "      <th>NC_000913v3_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>W</td>\n",
       "      <td>A</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>WA90A</td>\n",
       "      <td>4.186050</td>\n",
       "      <td>6.821188</td>\n",
       "      <td>2.635138</td>\n",
       "      <td>0.400612</td>\n",
       "      <td>0.897989</td>\n",
       "      <td>0.632237</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>WA90C</td>\n",
       "      <td>4.586894</td>\n",
       "      <td>7.469766</td>\n",
       "      <td>2.882872</td>\n",
       "      <td>0.034782</td>\n",
       "      <td>0.716512</td>\n",
       "      <td>0.682362</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>W</td>\n",
       "      <td>D</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>WA90D</td>\n",
       "      <td>4.191682</td>\n",
       "      <td>8.043704</td>\n",
       "      <td>3.852022</td>\n",
       "      <td>0.286227</td>\n",
       "      <td>0.995652</td>\n",
       "      <td>0.806663</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>W</td>\n",
       "      <td>E</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>WA90E</td>\n",
       "      <td>4.316510</td>\n",
       "      <td>7.292560</td>\n",
       "      <td>2.976050</td>\n",
       "      <td>0.228080</td>\n",
       "      <td>0.448962</td>\n",
       "      <td>0.251118</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>A</td>\n",
       "      <td>W</td>\n",
       "      <td>F</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>WA90F</td>\n",
       "      <td>4.399760</td>\n",
       "      <td>4.369536</td>\n",
       "      <td>-0.030224</td>\n",
       "      <td>0.405339</td>\n",
       "      <td>1.240901</td>\n",
       "      <td>1.500742</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1655</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "      <td>S</td>\n",
       "      <td>169</td>\n",
       "      <td>169</td>\n",
       "      <td>HB169S</td>\n",
       "      <td>5.195986</td>\n",
       "      <td>5.483526</td>\n",
       "      <td>0.287540</td>\n",
       "      <td>0.061543</td>\n",
       "      <td>0.053406</td>\n",
       "      <td>0.067100</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1656</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "      <td>T</td>\n",
       "      <td>169</td>\n",
       "      <td>169</td>\n",
       "      <td>HB169T</td>\n",
       "      <td>5.210294</td>\n",
       "      <td>5.444138</td>\n",
       "      <td>0.233844</td>\n",
       "      <td>0.053618</td>\n",
       "      <td>0.042732</td>\n",
       "      <td>0.067445</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1657</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "      <td>V</td>\n",
       "      <td>169</td>\n",
       "      <td>169</td>\n",
       "      <td>HB169V</td>\n",
       "      <td>5.221500</td>\n",
       "      <td>4.231318</td>\n",
       "      <td>-0.990182</td>\n",
       "      <td>0.028443</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.028500</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "      <td>W</td>\n",
       "      <td>169</td>\n",
       "      <td>169</td>\n",
       "      <td>HB169W</td>\n",
       "      <td>5.178334</td>\n",
       "      <td>4.848346</td>\n",
       "      <td>-0.329988</td>\n",
       "      <td>0.099939</td>\n",
       "      <td>0.120134</td>\n",
       "      <td>0.146561</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1659</th>\n",
       "      <td>P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb</td>\n",
       "      <td>P39411</td>\n",
       "      <td>P39411</td>\n",
       "      <td>EXP</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>169</td>\n",
       "      <td>169</td>\n",
       "      <td>HB169Y</td>\n",
       "      <td>5.224274</td>\n",
       "      <td>4.733868</td>\n",
       "      <td>-0.490406</td>\n",
       "      <td>0.022279</td>\n",
       "      <td>0.374764</td>\n",
       "      <td>0.363093</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1660 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          pdb  prot_a  prot_b evidence chain  \\\n",
       "0     P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     A   \n",
       "1     P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     A   \n",
       "2     P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     A   \n",
       "3     P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     A   \n",
       "4     P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     A   \n",
       "...                                       ...     ...     ...      ...   ...   \n",
       "1655  P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     B   \n",
       "1656  P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     B   \n",
       "1657  P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     B   \n",
       "1658  P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     B   \n",
       "1659  P39411-P39411-EXP-1u5w-pdb3-F-0-E-0.pdb  P39411  P39411      EXP     B   \n",
       "\n",
       "     aa_wt aa_mt  pdb_pos  uniprot_pos     mut     dG_wt     dG_mt       ddG  \\\n",
       "0        W     A       90           90   WA90A  4.186050  6.821188  2.635138   \n",
       "1        W     C       90           90   WA90C  4.586894  7.469766  2.882872   \n",
       "2        W     D       90           90   WA90D  4.191682  8.043704  3.852022   \n",
       "3        W     E       90           90   WA90E  4.316510  7.292560  2.976050   \n",
       "4        W     F       90           90   WA90F  4.399760  4.369536 -0.030224   \n",
       "...    ...   ...      ...          ...     ...       ...       ...       ...   \n",
       "1655     H     S      169          169  HB169S  5.195986  5.483526  0.287540   \n",
       "1656     H     T      169          169  HB169T  5.210294  5.444138  0.233844   \n",
       "1657     H     V      169          169  HB169V  5.221500  4.231318 -0.990182   \n",
       "1658     H     W      169          169  HB169W  5.178334  4.848346 -0.329988   \n",
       "1659     H     Y      169          169  HB169Y  5.224274  4.733868 -0.490406   \n",
       "\n",
       "      dG_wt_sd  dG_mt_sd    ddG_sd  NC_000913v3_pos  \n",
       "0     0.400612  0.897989  0.632237               91  \n",
       "1     0.034782  0.716512  0.682362               91  \n",
       "2     0.286227  0.995652  0.806663               91  \n",
       "3     0.228080  0.448962  0.251118               91  \n",
       "4     0.405339  1.240901  1.500742               91  \n",
       "...        ...       ...       ...              ...  \n",
       "1655  0.061543  0.053406  0.067100              170  \n",
       "1656  0.053618  0.042732  0.067445              170  \n",
       "1657  0.028443  0.000090  0.028500              170  \n",
       "1658  0.099939  0.120134  0.146561              170  \n",
       "1659  0.022279  0.374764  0.363093              170  \n",
       "\n",
       "[1660 rows x 17 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: change \"all_glpK_interface_df\" to something else\n",
    "\n",
    "# getting interface regions from mutfunc's precalculated outputs\n",
    "all_glpK_interface_df = pd.read_csv('./data/mutfunc_full/yjjX_interfaces.tab', sep='\\t', header=None)\n",
    "all_glpK_interface_df.columns = [\"pdb\",\"prot_a\",\"prot_b\",\"evidence\",\"chain\",\"aa_wt\",\"aa_mt\",\"pdb_pos\",\"uniprot_pos\",\"mut\",\"dG_wt\",\"dG_mt\",\"ddG\",\"dG_wt_sd\",\"dG_mt_sd\", \"ddG_sd\"]\n",
    "all_glpK_interface_df[\"NC_000913v3_pos\"] = all_glpK_interface_df[\"uniprot_pos\"] + 1  # Adding 1 to account for initiator methionine (start codon) within NC_000913v3 AA ref\n",
    "all_glpK_interface_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['P39411'], dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "array(['P39411'], dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = all_glpK_interface_df.copy()\n",
    "df = df.drop(columns=[\"aa_wt\", \"aa_mt\", \"mut\", \"dG_wt\", \"dG_mt\", \"ddG\", \"dG_wt_sd\", \"dG_mt_sd\", \"ddG_sd\"]).drop_duplicates()\n",
    "display(df.prot_a.unique(), df.prot_b.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mutil.gff import group_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[27, 28],\n",
       " [56],\n",
       " [84],\n",
       " [86, 87],\n",
       " [91],\n",
       " [101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114],\n",
       " [131],\n",
       " [144],\n",
       " [146, 147, 148, 149],\n",
       " [151, 152],\n",
       " [156, 157],\n",
       " [159, 160, 161],\n",
       " [163, 164, 165],\n",
       " [167, 168],\n",
       " [170, 171],\n",
       " [173, 174],\n",
       " [176]]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: change across the NB \"GlpK_GlpK\" to something else\n",
    "\n",
    "GlpK_GlpK_interface_df = all_glpK_interface_df[((all_glpK_interface_df[\"prot_a\"]==\"P39411\") & (all_glpK_interface_df[\"prot_b\"]==\"P39411\"))]\n",
    "GlpK_GlpK_interface_AAs = GlpK_GlpK_interface_df[\"NC_000913v3_pos\"].unique()\n",
    "\n",
    "# Outputting ranges for manual input into R\n",
    "GlpK_GlpK_interface_AAs.sort()\n",
    "GlpK_GlpK_interface_AA_ranges = group_sequences(GlpK_GlpK_interface_AAs)\n",
    "GlpK_GlpK_interface_AA_ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for rng_l in GlpK_GlpK_interface_AA_ranges:\n",
    "    gff_df = gff_df.append({\"feature\": \"YjjX subunit interface\", \"start\": rng_l[0], \"end\": rng_l[-1], \"color\": \"#4169E1\"}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 1)\n",
      "(38, 38)\n",
      "(30, 30)\n",
      "(30, 30)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "      <th>subgenic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1M (ATG→ATA) †</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633743.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>9.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>30.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>362</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633743, 4633743)</td>\n",
       "      <td>9 30 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>{unannotated region}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E38K (GAG→AAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633634.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>10.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>32.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633634, 4633634)</td>\n",
       "      <td>10 32 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(38, 38)</td>\n",
       "      <td>{Manganese or magnesium mental binding site}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>11.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>63.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>11 63 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(30, 30)</td>\n",
       "      <td>{unannotated region}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>12.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>60.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>12 60 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(30, 30)</td>\n",
       "      <td>{unannotated region}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type   Position RBS Reference Seq Sequence Change  \\\n",
       "0  M1M (ATG→ATA) †           SNP  4633743.0  {}     NC_000913             C→T   \n",
       "1   E38K (GAG→AAG)           SNP  4633634.0  {}     NC_000913             C→T   \n",
       "2   C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "3   C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "\n",
       "  TFBS   ale  coding            exp  flask gene RegulonDB ID  genetic  \\\n",
       "0   {}   9.0    True  ndh-cydB-appC   30.0    {ECK120002327}     True   \n",
       "1   {}  10.0    True  ndh-cydB-appC   32.0    {ECK120002327}     True   \n",
       "2   {}  11.0    True  ndh-cydB-appC   63.0    {ECK120002327}     True   \n",
       "3   {}  12.0    True  ndh-cydB-appC   60.0    {ECK120002327}     True   \n",
       "\n",
       "                                    genetic features  \\\n",
       "0  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "1  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "0  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   362      0.0   \n",
       "1  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...    22      0.0   \n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   190      0.0   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   320      1.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "0                       yjjX   0.0       1.0       {}         0.0   \n",
       "1                       yjjX   0.0       1.0       {}         0.0   \n",
       "2                       yjjX   0.0       1.0       {}         0.0   \n",
       "3                       yjjX   0.0       1.0       {}         0.0   \n",
       "\n",
       "                range     sample  tech_rep terminator  mutation size  \\\n",
       "0  (4633743, 4633743)   9 30 0 1       1.0         {}              1   \n",
       "1  (4633634, 4633634)  10 32 0 1       1.0         {}              1   \n",
       "2  (4633657, 4633657)  11 63 0 1       1.0         {}              1   \n",
       "3  (4633657, 4633657)  12 60 1 1       1.0         {}              1   \n",
       "\n",
       "   truncation  AA range                             subgenic features  \n",
       "0        True    (1, 1)                          {unannotated region}  \n",
       "1       False  (38, 38)  {Manganese or magnesium mental binding site}  \n",
       "2       False  (30, 30)                          {unannotated region}  \n",
       "3       False  (30, 30)                          {unannotated region}  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_mut_subgenic_feats(mut_AA_pos_range, gff_df):\n",
    "    mut_subgenic_feats = set()\n",
    "    print(mut_AA_pos_range)\n",
    "    mut_AA_pos_set = set(range(mut_AA_pos_range[0], mut_AA_pos_range[1] + 1))  # +1 so that \"range\" function will count last position\n",
    "    for _, r in gff_df.iterrows():\n",
    "        if r[\"feature\"] != \"Chain\":\n",
    "            feat_AA_pos_set = set(range(r[\"start\"], r[\"end\"] + 1))  # +1 so that \"range\" function will count last position\n",
    "            if len(mut_AA_pos_set & feat_AA_pos_set) > 0:\n",
    "                mut_subgenic_feats.add(r[\"feature\"])\n",
    "    if len(mut_subgenic_feats) == 0:\n",
    "        mut_subgenic_feats.add(\"unannotated region\")\n",
    "    return mut_subgenic_feats\n",
    "\n",
    "\n",
    "meta_muts[\"subgenic features\"] = meta_muts.apply(lambda r: _get_mut_subgenic_feats(r[\"AA range\"], gff_df), axis=1)\n",
    "meta_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P39411 M1M\n",
      "P39411 C30Y\n",
      "P39411 E38K\n"
     ]
    }
   ],
   "source": [
    "# TODO: put this in mutil since I end up using it in every gene-based mut-needle-plot\n",
    "# For mutfunc with methionine being considered as AA 1\n",
    "mut_set = set()  # to only get unique\n",
    "for _, r in meta_muts.iterrows():\n",
    "    if (r[\"Mutation Type\"] == \"SNP\") and ('*' not in r[\"Details\"]):\n",
    "        s = r[\"Details\"]\n",
    "        out_str = \"P39411 \" + s[:s.find(\" (\")]\n",
    "        mut_set.add(out_str)\n",
    "\n",
    "for s in mut_set:\n",
    "    print(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "      <th>subgenic features</th>\n",
       "      <th>AA change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1M (ATG→ATA) †</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633743.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>9.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>30.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>362</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633743, 4633743)</td>\n",
       "      <td>9 30 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1, 1)</td>\n",
       "      <td>{unannotated region}</td>\n",
       "      <td>M1M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E38K (GAG→AAG)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633634.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>10.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>32.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633634, 4633634)</td>\n",
       "      <td>10 32 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(38, 38)</td>\n",
       "      <td>{Manganese or magnesium mental binding site}</td>\n",
       "      <td>E38K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>11.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>63.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>11 63 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(30, 30)</td>\n",
       "      <td>{unannotated region}</td>\n",
       "      <td>C30Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>12.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>60.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>12 60 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(30, 30)</td>\n",
       "      <td>{unannotated region}</td>\n",
       "      <td>C30Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Details Mutation Type   Position RBS Reference Seq Sequence Change  \\\n",
       "0  M1M (ATG→ATA) †           SNP  4633743.0  {}     NC_000913             C→T   \n",
       "1   E38K (GAG→AAG)           SNP  4633634.0  {}     NC_000913             C→T   \n",
       "2   C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "3   C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "\n",
       "  TFBS   ale  coding            exp  flask gene RegulonDB ID  genetic  \\\n",
       "0   {}   9.0    True  ndh-cydB-appC   30.0    {ECK120002327}     True   \n",
       "1   {}  10.0    True  ndh-cydB-appC   32.0    {ECK120002327}     True   \n",
       "2   {}  11.0    True  ndh-cydB-appC   63.0    {ECK120002327}     True   \n",
       "3   {}  12.0    True  ndh-cydB-appC   60.0    {ECK120002327}     True   \n",
       "\n",
       "                                    genetic features  \\\n",
       "0  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "1  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "0  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   362      0.0   \n",
       "1  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...    22      0.0   \n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   190      0.0   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   320      1.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "0                       yjjX   0.0       1.0       {}         0.0   \n",
       "1                       yjjX   0.0       1.0       {}         0.0   \n",
       "2                       yjjX   0.0       1.0       {}         0.0   \n",
       "3                       yjjX   0.0       1.0       {}         0.0   \n",
       "\n",
       "                range     sample  tech_rep terminator  mutation size  \\\n",
       "0  (4633743, 4633743)   9 30 0 1       1.0         {}              1   \n",
       "1  (4633634, 4633634)  10 32 0 1       1.0         {}              1   \n",
       "2  (4633657, 4633657)  11 63 0 1       1.0         {}              1   \n",
       "3  (4633657, 4633657)  12 60 1 1       1.0         {}              1   \n",
       "\n",
       "   truncation  AA range                             subgenic features  \\\n",
       "0        True    (1, 1)                          {unannotated region}   \n",
       "1       False  (38, 38)  {Manganese or magnesium mental binding site}   \n",
       "2       False  (30, 30)                          {unannotated region}   \n",
       "3       False  (30, 30)                          {unannotated region}   \n",
       "\n",
       "  AA change  \n",
       "0       M1M  \n",
       "1      E38K  \n",
       "2      C30Y  \n",
       "3      C30Y  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_muts[\"AA change\"] = meta_muts.apply(lambda r: r[\"Details\"][:r[\"Details\"].find(\"(\")].strip() if r[\"Mutation Type\"]==\"SNP\" else \"\", axis=1)\n",
    "meta_muts[meta_muts[\"Mutation Type\"]==\"SNP\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mut_conservation_df = pd.read_csv(\"./data/mutfunc/yjjX/conservation.tab\", sep='\\t', comment='#')\n",
    "# mut_conservation_df[\"AA change\"] = mut_conservation_df.apply(lambda r: r[\"refaa\"] + str(r[\"posaa\"]) + r[\"altaa\"], axis=1)\n",
    "# meta_muts[\"conservation (SIFT)\"] = meta_muts.apply(lambda r: True if r[\"AA change\"] in mut_conservation_df[\"AA change\"].values else False, axis=1)\n",
    "# meta_muts[meta_muts[\"conservation (SIFT)\"]].head()\n",
    "meta_muts[\"conservation (SIFT)\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>ale</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>oriC</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>sample</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>terminator</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>truncation</th>\n",
       "      <th>AA range</th>\n",
       "      <th>subgenic features</th>\n",
       "      <th>AA change</th>\n",
       "      <th>conservation (SIFT)</th>\n",
       "      <th>stability (ΔΔGpred)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>11.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>63.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>11 63 0 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(30, 30)</td>\n",
       "      <td>{unannotated region}</td>\n",
       "      <td>C30Y</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C30Y (TGC→TAC)</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4633657.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>12.0</td>\n",
       "      <td>True</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>60.0</td>\n",
       "      <td>{ECK120002327}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>[{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...</td>\n",
       "      <td>320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4633657, 4633657)</td>\n",
       "      <td>12 60 1 1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(30, 30)</td>\n",
       "      <td>{unannotated region}</td>\n",
       "      <td>C30Y</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Details Mutation Type   Position RBS Reference Seq Sequence Change  \\\n",
       "2  C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "3  C30Y (TGC→TAC)           SNP  4633657.0  {}     NC_000913             C→T   \n",
       "\n",
       "  TFBS   ale  coding            exp  flask gene RegulonDB ID  genetic  \\\n",
       "2   {}  11.0    True  ndh-cydB-appC   63.0    {ECK120002327}     True   \n",
       "3   {}  12.0    True  ndh-cydB-appC   60.0    {ECK120002327}     True   \n",
       "\n",
       "                                    genetic features  \\\n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   \n",
       "\n",
       "                                    genomic features index  isolate  \\\n",
       "2  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   190      0.0   \n",
       "3  [{'name': 'yjjX', 'RegulonDB ID': 'ECK12000232...   320      1.0   \n",
       "\n",
       "  mutation target annotation  oriC  presence promoter  pseudogene  \\\n",
       "2                       yjjX   0.0       1.0       {}         0.0   \n",
       "3                       yjjX   0.0       1.0       {}         0.0   \n",
       "\n",
       "                range     sample  tech_rep terminator  mutation size  \\\n",
       "2  (4633657, 4633657)  11 63 0 1       1.0         {}              1   \n",
       "3  (4633657, 4633657)  12 60 1 1       1.0         {}              1   \n",
       "\n",
       "   truncation  AA range     subgenic features AA change  conservation (SIFT)  \\\n",
       "2       False  (30, 30)  {unannotated region}      C30Y                False   \n",
       "3       False  (30, 30)  {unannotated region}      C30Y                False   \n",
       "\n",
       "   stability (ΔΔGpred)  \n",
       "2                 True  \n",
       "3                 True  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_stability_df = pd.read_csv(\"./data/mutfunc/yjjX/stability.tab\", sep='\\t', comment='#')\n",
    "mut_stability_df[\"AA change\"] = mut_stability_df.apply(lambda r: r[\"refaa\"] + str(r[\"posaa\"]) + r[\"altaa\"], axis=1)\n",
    "meta_muts[\"stability (ΔΔGpred)\"] = meta_muts.apply(lambda r: True if r[\"AA change\"] in mut_stability_df[\"AA change\"].values else False, axis=1)\n",
    "meta_muts[meta_muts[\"stability (ΔΔGpred)\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coding disruption</th>\n",
       "      <th>structural disruption</th>\n",
       "      <th>unknown effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Manganese or magnesium mental binding site</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unannotated region</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            coding disruption  \\\n",
       "Manganese or magnesium mental binding site                  0   \n",
       "unannotated region                                          1   \n",
       "\n",
       "                                            structural disruption  \\\n",
       "Manganese or magnesium mental binding site                      0   \n",
       "unannotated region                                              2   \n",
       "\n",
       "                                            unknown effect  \n",
       "Manganese or magnesium mental binding site               1  \n",
       "unannotated region                                       0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: need to combine a few of the features\n",
    "\n",
    "feat_set = set(gff_df[\"feature\"])\n",
    "feat_set |= {\"unannotated region\"}\n",
    "feat_mut_cnt_df = pd.DataFrame(0,\n",
    "                               columns=[\"coding disruption\", \"functional disruption\", \"structural disruption\", \"unknown effect\"],\n",
    "                               index=feat_set)\n",
    "feat_mut_cnt_df\n",
    "\n",
    "for _, r in meta_muts.iterrows():\n",
    "    for f in r[\"subgenic features\"]:\n",
    "        effect_predicted = False\n",
    "        if r[\"truncation\"]:\n",
    "            feat_mut_cnt_df.at[f, \"coding disruption\"] += 1\n",
    "            effect_predicted = True\n",
    "        else:\n",
    "            if r[\"conservation (SIFT)\"]:\n",
    "                feat_mut_cnt_df.at[f, \"functional disruption\"] += 1\n",
    "                effect_predicted = True\n",
    "            if r[\"stability (ΔΔGpred)\"]:\n",
    "                feat_mut_cnt_df.at[f, \"structural disruption\"] += 1\n",
    "                effect_predicted = True\n",
    "        if not effect_predicted:\n",
    "            feat_mut_cnt_df.at[f, \"unknown effect\"] += 1\n",
    "# feat_mut_cnt_df = feat_mut_cnt_df[~(feat_mut_cnt_df.index.isin([\"Turn\", \"Helix\", \"Beta strand\"]))]\n",
    "feat_mut_cnt_df = feat_mut_cnt_df[(feat_mut_cnt_df.T != 0).any(axis=0)]  # remove rows that are all zeros\n",
    "feat_mut_cnt_df = feat_mut_cnt_df.loc[:, (feat_mut_cnt_df != 0).any(axis=0)]  # remove columns that are all zeros\n",
    "feat_mut_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ndh-cydB-appC'], dtype=object)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_muts.exp.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA position</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>disrupt coding seq</th>\n",
       "      <th>disrupt func</th>\n",
       "      <th>disrupt struct</th>\n",
       "      <th>study</th>\n",
       "      <th>study color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>38</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  AA position mutation type disrupt coding seq disrupt func disrupt struct  \\\n",
       "0           1           SNP                  1            0              0   \n",
       "1          38           SNP                  0            0              0   \n",
       "2          30           SNP                  0            0              1   \n",
       "3          30           SNP                  0            0              1   \n",
       "\n",
       "           study study color  \n",
       "0  ndh-cydB-appC     #4878D0  \n",
       "1  ndh-cydB-appC     #4878D0  \n",
       "2  ndh-cydB-appC     #4878D0  \n",
       "3  ndh-cydB-appC     #4878D0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_id_d = {\n",
    "    \"ndh-cydB-appC\": \"ndh-cydB-appC\",\n",
    "}\n",
    "\n",
    "study_color_d = {\n",
    "    \"ndh-cydB-appC\": \"#4878D0\",\n",
    "}\n",
    "\n",
    "\n",
    "# This is done specifically to expand the count of rows that describe the full range of AA per mutations.\n",
    "AA_rng_exp_df = pd.DataFrame(columns=[\"AA position\", \"mutation type\", \"disrupt coding seq\", \"disrupt func\",\n",
    "                                      \"disrupt struct\"\n",
    "                                      ])\n",
    "for _, m in meta_muts.iterrows():\n",
    "    AA_positions = set(range(m[\"AA range\"][0], m[\"AA range\"][1] + 1))\n",
    "\n",
    "    disrupt_coding = 0\n",
    "    if m[\"truncation\"]:\n",
    "        disrupt_coding = 1\n",
    "        AA_positions = {m[\"AA range\"][0]}\n",
    "\n",
    "    disrupt_func = 0\n",
    "    if m[\"conservation (SIFT)\"]:\n",
    "        disrupt_func = 1\n",
    "\n",
    "    disrupt_struct = 0\n",
    "    if \"stability (ΔΔGpred)\" in m.index and m[\"stability (ΔΔGpred)\"]:\n",
    "        disrupt_struct = 1\n",
    "\n",
    "    for p in AA_positions:\n",
    "        doi = exp_id_d[m[\"exp\"]]\n",
    "        AA_rng_exp_df = AA_rng_exp_df.append(\n",
    "            {\"AA position\": p,\n",
    "             \"mutation type\": m[\"Mutation Type\"],\n",
    "             \"disrupt coding seq\": disrupt_coding,\n",
    "             \"disrupt func\": disrupt_func,\n",
    "             \"disrupt struct\": disrupt_struct,\n",
    "             \"study\": doi,\n",
    "             \"study color\": study_color_d[doi]\n",
    "            },\n",
    "            ignore_index=True)\n",
    "AA_rng_exp_df[\"mutation type\"] = AA_rng_exp_df[\"mutation type\"].apply(lambda mt: \"CNV\" if mt == \"AMP\" else mt)\n",
    "AA_rng_exp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA position</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>disrupt coding seq</th>\n",
       "      <th>disrupt func</th>\n",
       "      <th>disrupt struct</th>\n",
       "      <th>study</th>\n",
       "      <th>study color</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "      <td>2</td>\n",
       "      <td>SNP AA 30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  AA position mutation type disrupt coding seq disrupt func disrupt struct  \\\n",
       "0           1           SNP                  1            0              0   \n",
       "1          30           SNP                  0            0              1   \n",
       "2          38           SNP                  0            0              0   \n",
       "\n",
       "           study study color  mutation count       name  \n",
       "0  ndh-cydB-appC     #4878D0               1   SNP AA 1  \n",
       "1  ndh-cydB-appC     #4878D0               2  SNP AA 30  \n",
       "2  ndh-cydB-appC     #4878D0               1  SNP AA 38  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Just want to count the mutation rows\n",
    "aa_mut_cnt_df = pd.DataFrame()\n",
    "for t, gdf in AA_rng_exp_df.groupby(list(AA_rng_exp_df.columns)):\n",
    "    c = len(gdf)\n",
    "    df = gdf.copy()\n",
    "    df = df.drop_duplicates()\n",
    "    df[\"mutation count\"] = c\n",
    "    aa_mut_cnt_df = aa_mut_cnt_df.append(df, ignore_index=True)\n",
    "aa_mut_cnt_df[\"name\"] = aa_mut_cnt_df.apply(lambda r: r[\"mutation type\"] + \" AA \" + str(r[\"AA position\"]), axis=1)\n",
    "aa_mut_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA position</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>disrupt coding seq</th>\n",
       "      <th>disrupt func</th>\n",
       "      <th>disrupt struct</th>\n",
       "      <th>study</th>\n",
       "      <th>study color</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 1</td>\n",
       "      <td>red</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "      <td>2</td>\n",
       "      <td>SNP AA 30</td>\n",
       "      <td>#A020F0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ndh-cydB-appC</td>\n",
       "      <td>#4878D0</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP AA 38</td>\n",
       "      <td>#ABABAB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  AA position mutation type disrupt coding seq disrupt func disrupt struct  \\\n",
       "0           1           SNP                  1            0              0   \n",
       "1          30           SNP                  0            0              1   \n",
       "2          38           SNP                  0            0              0   \n",
       "\n",
       "           study study color  mutation count       name    color  \n",
       "0  ndh-cydB-appC     #4878D0               1   SNP AA 1      red  \n",
       "1  ndh-cydB-appC     #4878D0               2  SNP AA 30  #A020F0  \n",
       "2  ndh-cydB-appC     #4878D0               1  SNP AA 38  #ABABAB  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_mut_color(r):\n",
    "    mut_color = \"grey\"\n",
    "    if r[\"disrupt coding seq\"] == 1:  # red has priority over all other colors\n",
    "        mut_color = 'red'\n",
    "    else:\n",
    "        if (r[\"disrupt func\"] == 1) & (r[\"disrupt struct\"] == 1):\n",
    "            mut_color = \"#A52A2A\"\n",
    "        elif (r[\"disrupt func\"] == 1) & (r[\"disrupt struct\"] == 0):\n",
    "            mut_color = \"#FFA500\"\n",
    "        elif (r[\"disrupt func\"] == 0) & (r[\"disrupt struct\"] == 1):\n",
    "            mut_color = \"#A020F0\"\n",
    "        else:\n",
    "            mut_color = '#ABABAB'  # unknown\n",
    "    return mut_color\n",
    "\n",
    "\n",
    "aa_mut_cnt_df[\"color\"] = aa_mut_cnt_df.apply(lambda r: get_mut_color(r), axis=1)\n",
    "aa_mut_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_mut_cnt_df.to_csv(\"./yjjX_aa_muts.csv\", index=False)  # For R and other IPYNBs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>9</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Chain</td>\n",
       "      <td>1</td>\n",
       "      <td>170</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PRO_0000156337;Note=Inosine/xanthosine trip...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Substrate binding region</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Substrate binding;Ontology_term=ECO:00002...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#911EB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Substrate binding region</td>\n",
       "      <td>68</td>\n",
       "      <td>69</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Substrate binding;Ontology_term=ECO:00002...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#911EB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Manganese or magnesium mental binding site</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Manganese or magnesium;Ontology_term=ECO:...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#4169E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P39411</td>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>Manganese or magnesium mental binding site</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Note=Manganese or magnesium;Ontology_term=ECO:...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#4169E1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     source                                     feature  start  end  \\\n",
       "0  P39411  UniProtKB                                       Chain      1  170   \n",
       "1  P39411  UniProtKB                    Substrate binding region      8   13   \n",
       "2  P39411  UniProtKB                    Substrate binding region     68   69   \n",
       "3  P39411  UniProtKB  Manganese or magnesium mental binding site     38   38   \n",
       "4  P39411  UniProtKB  Manganese or magnesium mental binding site     68   68   \n",
       "\n",
       "  score strand frame                                          attribute   9  \\\n",
       "0     .      .     .  ID=PRO_0000156337;Note=Inosine/xanthosine trip... NaN   \n",
       "1     .      .     .  Note=Substrate binding;Ontology_term=ECO:00002... NaN   \n",
       "2     .      .     .  Note=Substrate binding;Ontology_term=ECO:00002... NaN   \n",
       "3     .      .     .  Note=Manganese or magnesium;Ontology_term=ECO:... NaN   \n",
       "4     .      .     .  Note=Manganese or magnesium;Ontology_term=ECO:... NaN   \n",
       "\n",
       "     color  \n",
       "0    white  \n",
       "1  #911EB4  \n",
       "2  #911EB4  \n",
       "3  #4169E1  \n",
       "4  #4169E1  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feat_to_color_d = {\n",
    "    \"Chain\": \"white\",  # don't need to find intermediate grey chains if the overlap is clean.\n",
    "    \"Helix\": \"#937860\",\n",
    "    \"Turn\": \"#f58231\",\n",
    "    \"Beta strand\": \"#8C8C8C\",\n",
    "    'Substrate binding region':'#911EB4',\n",
    "    'Manganese or magnesium mental binding site':'#4169E1',\n",
    "    'YjjX subunit interface':'#469990',\n",
    "    'NTPase_I-T Pfam domain':'#FCAD74',\n",
    "}\n",
    "\n",
    "\n",
    "gff_df[\"color\"] = gff_df.apply(lambda r: feat_to_color_d[r[\"feature\"]], axis=1)\n",
    "gff_df.to_csv(\"./yjjX_aa_feats.csv\", index=False)  # For R\n",
    "gff_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chain</td>\n",
       "      <td>0</td>\n",
       "      <td>169</td>\n",
       "      <td>white</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>100</td>\n",
       "      <td>113</td>\n",
       "      <td>#469990</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Substrate binding region</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>#911EB4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>145</td>\n",
       "      <td>148</td>\n",
       "      <td>#469990</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>162</td>\n",
       "      <td>164</td>\n",
       "      <td>#469990</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>158</td>\n",
       "      <td>160</td>\n",
       "      <td>#469990</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>150</td>\n",
       "      <td>151</td>\n",
       "      <td>#469990</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Substrate binding region</td>\n",
       "      <td>67</td>\n",
       "      <td>68</td>\n",
       "      <td>#911EB4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>172</td>\n",
       "      <td>173</td>\n",
       "      <td>#469990</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>169</td>\n",
       "      <td>170</td>\n",
       "      <td>#469990</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>26</td>\n",
       "      <td>27</td>\n",
       "      <td>#469990</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>166</td>\n",
       "      <td>167</td>\n",
       "      <td>#469990</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>85</td>\n",
       "      <td>86</td>\n",
       "      <td>#469990</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>155</td>\n",
       "      <td>156</td>\n",
       "      <td>#469990</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>130</td>\n",
       "      <td>130</td>\n",
       "      <td>#469990</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>#469990</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>#469990</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>#469990</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>#469990</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Manganese or magnesium mental binding site</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Manganese or magnesium mental binding site</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>#4169E1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>YjjX subunit interface</td>\n",
       "      <td>175</td>\n",
       "      <td>175</td>\n",
       "      <td>#469990</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       feature  start  end    color  size\n",
       "0                                        Chain      0  169    white   170\n",
       "1                       YjjX subunit interface    100  113  #469990    14\n",
       "2                     Substrate binding region      7   12  #911EB4     6\n",
       "3                       YjjX subunit interface    145  148  #469990     4\n",
       "4                       YjjX subunit interface    162  164  #469990     3\n",
       "5                       YjjX subunit interface    158  160  #469990     3\n",
       "6                       YjjX subunit interface    150  151  #469990     2\n",
       "7                     Substrate binding region     67   68  #911EB4     2\n",
       "8                       YjjX subunit interface    172  173  #469990     2\n",
       "9                       YjjX subunit interface    169  170  #469990     2\n",
       "10                      YjjX subunit interface     26   27  #469990     2\n",
       "11                      YjjX subunit interface    166  167  #469990     2\n",
       "12                      YjjX subunit interface     85   86  #469990     2\n",
       "13                      YjjX subunit interface    155  156  #469990     2\n",
       "14                      YjjX subunit interface    130  130  #469990     1\n",
       "15                      YjjX subunit interface    143  143  #469990     1\n",
       "16                      YjjX subunit interface     90   90  #469990     1\n",
       "17                      YjjX subunit interface     83   83  #469990     1\n",
       "18                      YjjX subunit interface     55   55  #469990     1\n",
       "19  Manganese or magnesium mental binding site     67   67  #4169E1     1\n",
       "20  Manganese or magnesium mental binding site     37   37  #4169E1     1\n",
       "21                      YjjX subunit interface    175  175  #469990     1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# output feature data for NGL\n",
    "ngl_feat_df = gff_df[~(gff_df.feature.isin([\"Turn\", \"Helix\", \"Beta strand\"]))].copy()\n",
    "ngl_feat_df[\"size\"] = ngl_feat_df.apply(lambda r: r[\"end\"] - r[\"start\"] + 1, axis=1)\n",
    "ngl_feat_df = ngl_feat_df[[\"feature\", \"start\", \"end\", \"color\", \"size\"]]\n",
    "ngl_feat_df = ngl_feat_df.sort_values(\"size\", ascending=False)\n",
    "ngl_feat_df = ngl_feat_df.reset_index(drop=True)  # json explicitly uses the indexes\n",
    "\n",
    "# Need to so that the positions defined by AAs now match structural residues\n",
    "ngl_feat_df[\"start\"] = ngl_feat_df[\"start\"] - 1\n",
    "ngl_feat_df[\"end\"] = ngl_feat_df[\"end\"] - 1\n",
    "display(ngl_feat_df)\n",
    "with open('YjjX_feats.json', 'w') as outfile:   # for NGL\n",
    "    outfile.write(ngl_feat_df.to_json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA position</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>name</th>\n",
       "      <th>color</th>\n",
       "      <th>residue pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 1</td>\n",
       "      <td>red</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 30</td>\n",
       "      <td>#A020F0</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP AA 38</td>\n",
       "      <td>#ABABAB</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  AA position  mutation count mutation type       name    color residue pos\n",
       "0           1               1           SNP   SNP AA 1      red           0\n",
       "1          30               2           SNP  SNP AA 30  #A020F0          29\n",
       "2          38               1           SNP  SNP AA 38  #ABABAB          37"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "out_mut_df = aa_mut_cnt_df[[\"AA position\", \"mutation count\", \"mutation type\", \"name\", \"color\"]].copy()\n",
    "out_mut_df[\"residue pos\"] = out_mut_df[\"AA position\"] - 1\n",
    "display(out_mut_df)\n",
    "with open('YjjX_aa_muts.json', 'w') as outfile:  # for NGL\n",
    "    outfile.write(out_mut_df.to_json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
